From 7a0723ca280d1a3c38041e9ed7197c8de370164e Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Mon, 7 Oct 2019 21:41:17 +0100
Subject: [PATCH] Add MaterialJsonAdapter


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index d9aa81ce..1adbac67 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -1,6 +1,7 @@
 package org.bukkit;
 
 import com.google.common.collect.Maps;
+import com.google.gson.annotations.JsonAdapter;
 import org.apache.commons.lang.ArrayUtils;
 import org.apache.commons.lang.Validate;
 import org.bukkit.block.data.Ageable;
@@ -84,6 +85,7 @@ import java.util.function.Consumer;
  * An enum of all material IDs accepted by the official server and client
  */
 @SuppressWarnings({"DeprecatedIsStillUsed", "deprecation"}) // Paper
+@JsonAdapter(MaterialJsonAdapter.class)
 public class Material implements Keyed {
     // Enum compatibility
     private static final TreeSet<Material> MATERIALS = new TreeSet<>((a, b) -> a.name().compareTo(b.name));
diff --git a/src/main/java/org/bukkit/MaterialJsonAdapter.java b/src/main/java/org/bukkit/MaterialJsonAdapter.java
new file mode 100644
index 00000000..45068ae1
--- /dev/null
+++ b/src/main/java/org/bukkit/MaterialJsonAdapter.java
@@ -0,0 +1,20 @@
+package org.bukkit;
+
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+import java.io.IOException;
+
+public class MaterialJsonAdapter extends TypeAdapter<Material> {
+
+    @Override
+    public void write(JsonWriter out, Material material) throws IOException {
+        out.value(material.name());
+    }
+
+    @Override
+    public Material read(JsonReader in) throws IOException {
+        return Material.valueOf(in.nextString());
+    }
+}
-- 
2.23.0

