From 3ecb55018c0f88ec7e80a3f12b34d6fbecfcd765 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Wed, 30 Oct 2019 01:31:14 +0000
Subject: [PATCH] Add fake OP permission level API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 8d8d8957..ee627ac3 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1979,6 +1979,23 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     void resetCooldown();
     // Paper end
 
+    // Papyrus start
+
+    /**
+     * Returns the fake OP permission level of this player, or -1 if none has been set.
+     * @return fake OP permission level or -1 if none has been set
+     */
+    int getFakeOpPermissionLevel();
+
+    /**
+     * Sets the fake OP permission level of this player. This permission level will be sent to the client which will
+     * modify some aspects of its behavior, such as command block opening or some F3 commands, but will not change
+     * anything server-side. Set it to -1 to unset fake permission level and send real permission level instead.
+     * @param level fake OP permission level (1-4) or -1 to set none
+     */
+    void setFakeOpPermissionLevel(int level);
+    // Papyrus end
+
     // Spigot start
     public class Spigot extends Entity.Spigot
     {
-- 
2.23.0

