From be522b6f1b803204daaa38b286484de88d773d7a Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Wed, 30 Oct 2019 01:31:14 +0000
Subject: [PATCH] Add Player canAlwaysOpenCommandBlocks API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index ea5257cc9..b2ff703ae 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1795,6 +1795,23 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     Firework boostElytra(@NotNull ItemStack firework);
     // Paper end
 
+    // Papyrus start
+    /**
+     * Gets whether this player can always open command blocks and see their content
+     * @return whether this player can always open command blocks
+     */
+    boolean canAlwaysOpenCommandBlocks();
+
+    /**
+     * Sets whether this player can always open command blocks and see their content.
+     * This setting has some minor client side-effects, such as some F3 commands, but will not change
+     * anything server-side unrelated to command blocks opening.
+     * @see <a href="https://github.com/TheKinrar/Papyrus/issues/2">Papyrus#2</a>
+     * @param canAlwaysOpenCommandBlocks whether this player can always open command blocks
+     */
+    void setCanAlwaysOpenCommandBlocks(boolean canAlwaysOpenCommandBlocks);
+    // Papyrus end
+
     // Spigot start
     public class Spigot extends Entity.Spigot {
 
-- 
2.30.0

