From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Mon, 4 Nov 2019 19:42:56 +0000
Subject: [PATCH] Hybrid mode API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 5f7208196684d9c8373df28b7cfb5f9e21baa41e..e1cfcde2f53bca782cff127b6b995bf87c393549 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1939,6 +1939,17 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Papyrus start
+    /**
+     * Gets whether the Server is in hybrid mode or not.
+     *
+     * @return true if the server is in hybrid mode, false otherwise
+     */
+    public static boolean getHybridMode() {
+        return server.getHybridMode();
+    }
+    // Papyrus end
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index f3e27d2d02a9407bb1b091b8c1125ad5abf99e55..b88f4b201b6d971465ce659619fb89ef0b60e093 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1699,4 +1699,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     com.destroystokyo.paper.entity.ai.MobGoals getMobGoals();
     // Paper end
+
+    // Papyrus start
+    /**
+     * Gets whether the Server is in hybrid mode or not.
+     *
+     * @return true if the server is in hybrid mode, false otherwise
+     */
+    boolean getHybridMode();
+    // Papyrus end
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 9473084f12d31f986a2aaeec25f6b32afcb30075..e980f5c94c62271d47a9ed5be66f6027fa7afdbc 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1951,6 +1951,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param canAlwaysOpenCommandBlocks whether this player can always open command blocks
      */
     void setCanAlwaysOpenCommandBlocks(boolean canAlwaysOpenCommandBlocks);
+
+    /**
+     * Returns whether this player is authenticated with the server.
+     * That is, if the player either used /login to authenticate,
+     * registered for the first time using /register,
+     * was auto-logged in (online ID or auth timeout),
+     * or was force-authenticated with the API.
+     * @return whether this player is authenticated
+     */
+    boolean isAuthenticated();
     // Papyrus end
 
     // Spigot start
diff --git a/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java b/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
index 16b5fd279b0cb926900247618bcdb381a93f5a35..86f1a6b2f8bce5e6fd516a8b99d4047f945301cc 100644
--- a/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
@@ -209,6 +209,10 @@ public class EntityPotionEffectEvent extends EntityEvent implements Cancellable
          * When an illusion illager makes himself disappear.
          */
         ILLUSION,
+        /**
+         * When a user is requested to login or register. No event will be triggered for this cause.
+         */
+        HYBRID_AUTH,
         /**
          * When all effects are removed due to a bucket of milk.
          */
