From 183e2469d7c78cf679f2ed9c5814f740e7585931 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Thu, 21 Nov 2019 11:58:39 +0000
Subject: [PATCH] Add Block#getCenter


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index a14a7583..3e0f08a9 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -227,6 +227,13 @@ public interface Block extends Metadatable {
     @NotNull
     Location getLocation();
 
+    /**
+     * Gets the Location of the center of the block
+     * @return Location of center of block
+     */
+    @NotNull
+    Location getCenter();
+
     /**
      * Stores the location of the block in the provided Location object.
      * <p>
-- 
2.25.1

