From ffdd66c1da998bb9733c6c6d60d864903a50230c Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sun, 29 Dec 2019 22:06:28 +0000
Subject: [PATCH] Add Skeleton#getAttackDelay and setAttackDelay


diff --git a/src/main/java/org/bukkit/entity/Skeleton.java b/src/main/java/org/bukkit/entity/Skeleton.java
index 1c367f78e..f48a20925 100644
--- a/src/main/java/org/bukkit/entity/Skeleton.java
+++ b/src/main/java/org/bukkit/entity/Skeleton.java
@@ -27,6 +27,10 @@ public interface Skeleton extends Monster, RangedEntity { // Paper
     @Contract("_ -> fail")
     public void setSkeletonType(SkeletonType type);
 
+    public int getAttackDelay();
+
+    public void setAttackDelay(int delay);
+
     /*
      * @deprecated classes are different types
      */
-- 
2.25.0

