From eb7cf1cbc085579b3da43f596c4a83803283b3e6 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Wed, 15 Jan 2020 07:27:06 +0000
Subject: [PATCH] Differenciate papyrus plugins in /plugins


diff --git a/src/main/java/org/bukkit/command/defaults/PluginsCommand.java b/src/main/java/org/bukkit/command/defaults/PluginsCommand.java
index fd74512a2..7a438ea96 100644
--- a/src/main/java/org/bukkit/command/defaults/PluginsCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/PluginsCommand.java
@@ -58,14 +58,20 @@ public class PluginsCommand extends BukkitCommand {
             }
 
 
-            pluginList.append(plugin.isEnabled() ? ChatColor.GREEN : ChatColor.RED);
-            // Paper start - Add an asterisk to legacy plugins (so admins are aware)
+            ChatColor color = ChatColor.GREEN;
             String pluginName = plugin.getDescription().getName();
+            if(!plugin.getDescription().isPapyrus()) {
+                pluginName += ".";
+                color = ChatColor.YELLOW;
+            }
+
             if (org.bukkit.UnsafeValues.isLegacyPlugin(plugin)) {
                 pluginName += "*";
+                color = ChatColor.DARK_GRAY;
             }
+
+            pluginList.append(plugin.isEnabled() ? color : ChatColor.RED);
             pluginList.append(pluginName);
-            // Paper end
         }
 
         return "(" + plugins.size() + "): " + pluginList.toString();
-- 
2.28.0

