From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Wed, 15 Jan 2020 07:27:06 +0000
Subject: [PATCH] Differenciate papyrus plugins in /plugins


diff --git a/src/main/java/org/bukkit/command/defaults/PluginsCommand.java b/src/main/java/org/bukkit/command/defaults/PluginsCommand.java
index 1aa58c59e1e8738bbdc77752885ff3b18b29de42..4a8805fc0a87a5a7cdc85377f5ea7614d816822f 100644
--- a/src/main/java/org/bukkit/command/defaults/PluginsCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/PluginsCommand.java
@@ -53,12 +53,23 @@ public class PluginsCommand extends BukkitCommand {
 
             Plugin plugin = entry.getValue();
 
-            pluginList.append(plugin.isEnabled() ? ChatColor.GREEN : ChatColor.RED);
-            // Paper start - Add an asterisk to legacy plugins (so admins are aware)
+            if (plugin.getDescription().getProvides().size() > 0) {
+                pluginList.append("(").append(String.join(", ", plugin.getDescription().getProvides())).append(")"); // Paper - Remove unnecessary space
+            }
+
+            ChatColor color = ChatColor.GREEN;
             String pluginName = plugin.getDescription().getName();
+            if(!plugin.getDescription().isPapyrus()) {
+                pluginName += ".";
+                color = ChatColor.YELLOW;
+            }
+
             if (org.bukkit.UnsafeValues.isLegacyPlugin(plugin)) {
                 pluginName += "*";
+                color = ChatColor.DARK_GRAY;
             }
+
+            pluginList.append(plugin.isEnabled() ? color : ChatColor.RED);
             pluginList.append(pluginName);
             // Paper end
 
