From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sun, 1 Mar 2020 01:17:13 +0000
Subject: [PATCH] More plugin-loading related APIs


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 1087abf086397d34abae04d82dd4d00280ae3674..55950bee47644a090f08ed3008d0ec4172419799 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -332,6 +332,13 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Papyrus start
+    @NotNull
+    public static File getPluginFolder() {
+        return server.getPluginFolder();
+    }
+    // Papyrus end
+
     /**
      * Gets the name of the update folder. The update folder is used to safely
      * update plugins at the right moment on a plugin load.
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 9e0b6e1be44e6379c0d6fda7482343352cbae34c..e365b6ac725af7a543483dce53d72a2d623d9bd6 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -277,6 +277,11 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     }
     // Paper end
 
+    // Papyrus start
+    @NotNull
+    public File getPluginFolder();
+    // Papyrus end
+
     /**
      * Gets the name of the update folder. The update folder is used to safely
      * update plugins at the right moment on a plugin load.
diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index 92b66aba6aedc9362cb9d7aa2e1b91f89d5dd061..8ba0c70f13efa0b9e34e05650029826129fb6046 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -21,6 +21,7 @@ import org.bukkit.command.defaults.PluginsCommand;
 import org.bukkit.command.defaults.ReloadCommand;
 import org.bukkit.command.defaults.VersionCommand;
 import org.bukkit.entity.Player;
+import org.bukkit.plugin.Plugin;
 import org.bukkit.util.StringUtil;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
@@ -45,6 +46,13 @@ public class SimpleCommandMap implements CommandMap {
         register("bukkit", new HelpCommand());
     }
 
+    // Papyrus start - unregisterAll
+    public void unregisterAll(@NotNull Plugin plugin) {
+        knownCommands.entrySet().removeIf(entry -> entry.getValue() instanceof PluginCommand
+                && ((PluginCommand) entry.getValue()).getPlugin().equals(plugin));
+    }
+    // Papyrus end
+
     /**
      * {@inheritDoc}
      */
diff --git a/src/main/java/org/bukkit/plugin/PluginLoader.java b/src/main/java/org/bukkit/plugin/PluginLoader.java
index 6ab9cd8213cbe35943748dcf42948d5fc048c84c..7d3acca0ac25ec3682df0e3f9079295705ac2e31 100644
--- a/src/main/java/org/bukkit/plugin/PluginLoader.java
+++ b/src/main/java/org/bukkit/plugin/PluginLoader.java
@@ -7,6 +7,7 @@ import java.util.regex.Pattern;
 import org.bukkit.event.Event;
 import org.bukkit.event.Listener;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * Represents a plugin loader, which handles direct access to specific types
@@ -28,6 +29,9 @@ public interface PluginLoader {
     @NotNull
     public Plugin loadPlugin(@NotNull File file) throws InvalidPluginException, UnknownDependencyException;
 
+    @NotNull
+    public Plugin loadPlugin(@NotNull File file, @Nullable File parent) throws InvalidPluginException, UnknownDependencyException;
+
     /**
      * Loads a PluginDescriptionFile from the specified file
      *
diff --git a/src/main/java/org/bukkit/plugin/PluginManager.java b/src/main/java/org/bukkit/plugin/PluginManager.java
index 86cc5025ad98f7a752c51713b7cd6a39d5136ecc..62d20b9d9660391d574cef358061d4cc2265763d 100644
--- a/src/main/java/org/bukkit/plugin/PluginManager.java
+++ b/src/main/java/org/bukkit/plugin/PluginManager.java
@@ -80,6 +80,9 @@ public interface PluginManager {
     @Nullable
     public Plugin loadPlugin(@NotNull File file) throws InvalidPluginException, InvalidDescriptionException, UnknownDependencyException;
 
+    @Nullable
+    public Plugin loadPlugin(@NotNull File file, @Nullable File parentFile) throws InvalidPluginException, InvalidDescriptionException, UnknownDependencyException;
+
     /**
      * Loads the plugins contained within the specified directory
      *
@@ -89,6 +92,12 @@ public interface PluginManager {
     @NotNull
     public Plugin[] loadPlugins(@NotNull File directory);
 
+    // Papyrus start - unload
+
+    public void unloadPlugin(@NotNull Plugin plugin);
+
+    // Papyrus end
+
     /**
      * Disables all the loaded plugins
      */
diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
index 0e25119564dfa9cb12f3c5dc5f653d7f2c147a9d..0dad789a906a6f805f12eea9a624eff9dba3194e 100644
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -382,6 +382,23 @@ public final class SimplePluginManager implements PluginManager {
         return result.toArray(new Plugin[result.size()]);
     }
 
+    // Papyrus start - unload
+
+    @Override
+    public void unloadPlugin(@NotNull Plugin plugin) {
+        if(plugin.isEnabled()) {
+            disablePlugin(plugin, true);
+        }
+
+        commandMap.unregisterAll(plugin);
+        plugin.getDescription().getPermissions().forEach(this::removePermission);
+
+        plugins.remove(plugin);
+        lookupNames.remove(plugin.getDescription().getName().toLowerCase(java.util.Locale.ENGLISH));
+    }
+
+    // Papyrus end
+
     /**
      * Loads the plugin in the specified file
      * <p>
@@ -397,6 +414,12 @@ public final class SimplePluginManager implements PluginManager {
     @Override
     @Nullable
     public synchronized Plugin loadPlugin(@NotNull File file) throws InvalidPluginException, UnknownDependencyException {
+        return loadPlugin(file, null);
+    }
+
+    @Override
+    @Nullable
+    public synchronized Plugin loadPlugin(@NotNull File file, @Nullable File parentFile) throws InvalidPluginException, UnknownDependencyException {
         Validate.notNull(file, "File cannot be null");
 
         checkUpdate(file);
@@ -411,7 +434,7 @@ public final class SimplePluginManager implements PluginManager {
             if (match.find()) {
                 PluginLoader loader = fileAssociations.get(filter);
 
-                result = loader.loadPlugin(file);
+                result = loader.loadPlugin(file, parentFile);
             }
         }
 
diff --git a/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java b/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
index b622cedeeee017f042bcf92485d81832030a8030..c5cbf5318eedaa28ee3b1b7d557993b2bf8d2701 100644
--- a/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
@@ -78,9 +78,14 @@ public final class JavaPluginLoader implements PluginLoader {
         this.libraryLoader = libraryLoader;
     }
 
+    @Override
+    public @NotNull Plugin loadPlugin(@NotNull File file) throws InvalidPluginException, UnknownDependencyException {
+        return loadPlugin(file, null);
+    }
+
     @Override
     @NotNull
-    public Plugin loadPlugin(@NotNull final File file) throws InvalidPluginException {
+    public Plugin loadPlugin(@NotNull final File file, @Nullable File parentFile) throws InvalidPluginException {
         Validate.notNull(file, "File cannot be null");
 
         if (!file.exists()) {
@@ -94,7 +99,7 @@ public final class JavaPluginLoader implements PluginLoader {
             throw new InvalidPluginException(ex);
         }
 
-        final File parentFile = ((SimplePluginManager) this.server.getPluginManager()).pluginsDirectory(); // Paper
+        if(parentFile == null) parentFile = ((SimplePluginManager) this.server.getPluginManager()).pluginsDirectory(); // Paper // Papyrus - only if not provided
         final File dataFolder = new File(parentFile, description.getName());
         @SuppressWarnings("deprecation")
         final File oldDataFolder = new File(parentFile, description.getRawName());
