From 3a7b99b1f917b2aa4cf68ad52c7950012c81afff Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Wed, 15 Jan 2020 07:27:06 +0000
Subject: [PATCH] Differenciate papyrus plugins in /plugins


diff --git a/src/main/java/org/bukkit/command/defaults/PluginsCommand.java b/src/main/java/org/bukkit/command/defaults/PluginsCommand.java
index 6cfd9f3c..46df48f2 100644
--- a/src/main/java/org/bukkit/command/defaults/PluginsCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/PluginsCommand.java
@@ -43,11 +43,19 @@ public class PluginsCommand extends BukkitCommand {
         for (Plugin plugin : Bukkit.getPluginManager().getPlugins()) {
             // Paper start - Add an asterisk to legacy plugins (so admins are aware)
             String pluginName = plugin.getDescription().getName();
+
+            ChatColor color = ChatColor.GREEN;
+            if(!plugin.getDescription().isPapyrus()) {
+                pluginName += ".";
+                color = ChatColor.YELLOW;
+            }
+
             if (org.bukkit.UnsafeValues.isLegacyPlugin(plugin)) {
                 pluginName += "*";
+                color = ChatColor.DARK_GRAY;
             }
 
-            plugins.put(pluginName, plugin.isEnabled() ? ChatColor.GREEN : ChatColor.RED);
+            plugins.put(pluginName, plugin.isEnabled() ? color : ChatColor.RED);
             // Paper end
         }
 
-- 
2.24.1

