From 4da07835ff63edecc062c0c4b0025711fb502810 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sat, 22 Feb 2020 20:16:47 +0000
Subject: [PATCH] Add alwaysOffline hybrid API


diff --git a/src/main/java/fr/thekinrar/papyrus/hybrid/HybridProfile.java b/src/main/java/fr/thekinrar/papyrus/hybrid/HybridProfile.java
index 68a93884..b6a12950 100644
--- a/src/main/java/fr/thekinrar/papyrus/hybrid/HybridProfile.java
+++ b/src/main/java/fr/thekinrar/papyrus/hybrid/HybridProfile.java
@@ -13,4 +13,7 @@ public interface HybridProfile {
     void setPassword(@NotNull String password);
 
     void stayLoggedIn(long millis);
+
+    boolean isAlwaysOffline();
+    void setAlwaysOffline(boolean alwaysOffline);
 }
-- 
2.25.1

