From 341d36a092d6cbc9dc089cf96a83fe03036bba37 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sat, 4 Jul 2020 21:59:23 +0200
Subject: [PATCH] Add isUpToDate to VersionFetcher


diff --git a/src/main/java/com/destroystokyo/paper/util/VersionFetcher.java b/src/main/java/com/destroystokyo/paper/util/VersionFetcher.java
index a736d7bc..7d354291 100644
--- a/src/main/java/com/destroystokyo/paper/util/VersionFetcher.java
+++ b/src/main/java/com/destroystokyo/paper/util/VersionFetcher.java
@@ -26,6 +26,8 @@ public interface VersionFetcher {
     @NotNull
     Component getVersionMessage(@NotNull String serverVersion);
 
+    boolean isUpToDate();
+
     class DummyVersionFetcher implements VersionFetcher {
 
         @Override
@@ -41,5 +43,10 @@ public interface VersionFetcher {
             new Throwable().printStackTrace();
             return Component.text("Unable to check for updates. No version provider set.", NamedTextColor.RED);
         }
+
+        @Override
+        public boolean isUpToDate() {
+            return false;
+        }
     }
 }
-- 
2.30.1

