From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sun, 14 Mar 2021 18:38:42 +0100
Subject: [PATCH] Add noWeight, noDeclaration, maxVectorLength to Minecart


diff --git a/src/main/java/org/bukkit/entity/Minecart.java b/src/main/java/org/bukkit/entity/Minecart.java
index 53b042f8ebbbf6ee77435b93d4e89371375cc515..708f2f7ff1f82a10cdca57dae683b9b828bb78c6 100644
--- a/src/main/java/org/bukkit/entity/Minecart.java
+++ b/src/main/java/org/bukkit/entity/Minecart.java
@@ -154,4 +154,18 @@ public interface Minecart extends Vehicle {
     @NotNull
     public Material getMinecartMaterial();
     // Paper end
+
+    // Papyrus start
+    public void setMaxVectorLength(double maxVectorLength);
+
+    public double getMaxVectorLength();
+
+    public void setNoDeceleration(boolean noDeceleration);
+
+    public boolean getNoDeceleration();
+
+    public void setNoWeight(boolean noWeight);
+
+    public boolean getNoWeight();
+    // Papyrus end
 }
