From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sun, 9 May 2021 22:08:40 +0200
Subject: [PATCH] Add unsafe PluginDescriptionFile API


diff --git a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
index 9933dcf5c9b1057faf6a8810f70ca65137923b4f..07cb435ddf49eeabb0c0abe8dbc7cb5f44b8ef34 100644
--- a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
+++ b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
@@ -250,7 +250,11 @@ public final class PluginDescriptionFile {
     private boolean papyrus = false;
 
     public PluginDescriptionFile(@NotNull final InputStream stream) throws InvalidDescriptionException {
-        loadMap(asMap(YAML.get().load(stream)));
+        this(stream, true);
+    }
+
+    public PluginDescriptionFile(@NotNull final InputStream stream, boolean safe) throws InvalidDescriptionException {
+        loadMap(asMap(YAML.get().load(stream)), safe);
     }
 
     /**
@@ -261,9 +265,15 @@ public final class PluginDescriptionFile {
      *     invalid
      */
     public PluginDescriptionFile(@NotNull final Reader reader) throws InvalidDescriptionException {
-        loadMap(asMap(YAML.get().load(reader)));
+        this(reader, true);
+    }
+
+    public PluginDescriptionFile(@NotNull final Reader reader, boolean safe) throws InvalidDescriptionException {
+        loadMap(asMap(YAML.get().load(reader)), safe);
     }
 
+
+
     /**
      * Creates a new PluginDescriptionFile with the given detailed
      *
@@ -981,7 +991,7 @@ public final class PluginDescriptionFile {
         YAML.get().dump(saveMap(), writer);
     }
 
-    private void loadMap(@NotNull Map<?, ?> map) throws InvalidDescriptionException {
+    private void loadMap(@NotNull Map<?, ?> map, boolean safe) throws InvalidDescriptionException {
         try {
             name = rawName = map.get("name").toString();
 
@@ -1007,12 +1017,15 @@ public final class PluginDescriptionFile {
 
         try {
             main = map.get("main").toString();
-            if (main.startsWith("org.bukkit.")) {
+            if(main.startsWith("org.bukkit.")) {
+                if(safe)
                 throw new InvalidDescriptionException("main may not be within the org.bukkit namespace");
             }
-        } catch (NullPointerException ex) {
+        } catch(NullPointerException ex) {
+            if(safe)
             throw new InvalidDescriptionException(ex, "main is not defined");
-        } catch (ClassCastException ex) {
+        } catch(ClassCastException ex) {
+            if(safe)
             throw new InvalidDescriptionException(ex, "main is of wrong type");
         }
 
