From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Thu, 21 Nov 2019 11:58:39 +0000
Subject: [PATCH] Add Block#getCenter


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 9930ebe7a23d306c602840fd43652fbdaba481b3..88175938698521949416260a518bbc1956b6633b 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -248,6 +248,13 @@ public interface Block extends Metadatable, net.kyori.adventure.translation.Tran
     @NotNull
     Location getLocation();
 
+    /**
+     * Gets the Location of the center of the block
+     * @return Location of center of block
+     */
+    @NotNull
+    Location getCenter();
+
     /**
      * Stores the location of the block in the provided Location object.
      * <p>
