From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Thu, 21 Nov 2019 11:58:39 +0000
Subject: [PATCH] Add Block#getCenter


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 390a2a2611df35a9ea6f1eb996b47e2aa4597ff0..4f795faa31cf3c4b28deafe1205ba642b1969ac0 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -249,6 +249,13 @@ public interface Block extends Metadatable, Translatable, net.kyori.adventure.tr
     @NotNull
     Location getLocation();
 
+    /**
+     * Gets the Location of the center of the block
+     * @return Location of center of block
+     */
+    @NotNull
+    Location getCenter();
+
     /**
      * Stores the location of the block in the provided Location object.
      * <p>
