From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Wed, 15 Jan 2020 02:38:47 +0000
Subject: [PATCH] Add papyrus option to PluginDescriptionFile


diff --git a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
index 801a323d9c328c1ab8e58c651cea9c5f3dfbcdca..857e9100ddacbf893815e1c6480770b4ceee5a42 100644
--- a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
+++ b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
@@ -254,6 +254,7 @@ public final class PluginDescriptionFile {
     private Set<PluginAwareness> awareness = ImmutableSet.of();
     private String apiVersion = null;
     private List<String> libraries = ImmutableList.of();
+    private boolean papyrus = false;
 
     public PluginDescriptionFile(@NotNull final InputStream stream) throws InvalidDescriptionException {
         loadMap(asMap(YAML.get().load(stream)));
@@ -964,6 +965,10 @@ public final class PluginDescriptionFile {
         return apiVersion;
     }
 
+    public boolean isPapyrus() {
+        return papyrus;
+    }
+
     /**
      * Gets the libraries this plugin requires. This is a preview feature.
      * <ul>
@@ -1173,6 +1178,14 @@ public final class PluginDescriptionFile {
         if (map.get("prefix") != null) {
             prefix = map.get("prefix").toString();
         }
+
+        if (map.get("papyrus") != null) {
+            try {
+                papyrus = (Boolean) map.get("papyrus");
+            } catch (ClassCastException ex) {
+                throw new InvalidDescriptionException(ex, "papyrus has wrong type");
+            }
+        }
     }
 
     @NotNull
@@ -1250,6 +1263,8 @@ public final class PluginDescriptionFile {
             map.put("prefix", prefix);
         }
 
+        map.put("papyrus", papyrus);
+
         return map;
     }
 
diff --git a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
index 81292899918c4dc880661ee628384cb840a6244f..bb99425e6629b64180880834fe5ac494d5e30cd9 100644
--- a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
@@ -19,7 +19,9 @@ import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
 import java.util.jar.Manifest;
 import java.util.logging.Level;
+
 import org.apache.commons.lang.Validate;
+import org.bukkit.Bukkit;
 import org.bukkit.plugin.InvalidPluginException;
 import org.bukkit.plugin.PluginDescriptionFile;
 import org.bukkit.plugin.SimplePluginManager;
