From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Amaury Rousseau <kinrar@kinrar.io>
Date: Sun, 19 Dec 2021 18:42:50 +0100
Subject: [PATCH] Add canSeePlayerEntities API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 88c4885569d2b8b22fce55601d50608ac8e9388c..635181e93991aad27b9104ac0ea948156fa58094 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1645,6 +1645,35 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @ApiStatus.Experimental
     public boolean canSee(@NotNull Entity entity);
 
+    /**
+     * Hides player entities from this player. This is different from {@link Player#hidePlayer(Plugin, Player)}
+     * as it will only hide the entity instead of hiding the player from list, commands, etc.
+     * @param plugin Plugin that wants to hide player entities
+     */
+    public void hidePlayerEntities(@NotNull Plugin plugin);
+
+    /**
+     * Shows player entities to this player.<br>
+     * See {@link Player#hidePlayerEntities(Plugin)} for details.
+     * @param plugin Plugin that wants to show player entities
+     */
+    public void showPlayerEntities(@NotNull Plugin plugin);
+
+    /**
+     * Checks to see if player entities have been hidden from
+     * this player using {@link Player#hidePlayerEntities(Plugin)}.
+     * @return true if player entities have been hidden
+     */
+    public boolean canSeePlayerEntities();
+
+    /**
+     * Checks to see if a player entity has been hidden from this player, either using
+     * {@link Player#hidePlayerEntities(Plugin)} or {@link Player#hidePlayer(Plugin, Player)}.
+     * @param player Player to check
+     * @return Whether the player entity have been hidden from this player
+     */
+    public boolean canSeePlayerEntity(@NotNull Player player);
+
     /**
      * Checks to see if this player is currently flying or not.
      *
