From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Wed, 30 Oct 2019 01:31:14 +0000
Subject: [PATCH] Add Player canAlwaysOpenCommandBlocks API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index afa245018192a63e5db8bc568ddea2790bf5b8a2..0c0b70a93460cd9f8070e0e0b249ec78e422f79f 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2153,6 +2153,23 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     Set<Player> getTrackedPlayers();
     // Paper end
 
+    // Papyrus start
+    /**
+     * Gets whether this player can always open command blocks and see their content
+     * @return whether this player can always open command blocks
+     */
+    boolean canAlwaysOpenCommandBlocks();
+
+    /**
+     * Sets whether this player can always open command blocks and see their content.
+     * This setting has some minor client side-effects, such as some F3 commands, but will not change
+     * anything server-side unrelated to command blocks opening.
+     * @see <a href="https://github.com/TheKinrar/Papyrus/issues/2">Papyrus#2</a>
+     * @param canAlwaysOpenCommandBlocks whether this player can always open command blocks
+     */
+    void setCanAlwaysOpenCommandBlocks(boolean canAlwaysOpenCommandBlocks);
+    // Papyrus end
+
     // Spigot start
     public class Spigot extends Entity.Spigot {
 
