From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Wed, 30 Oct 2019 01:31:14 +0000
Subject: [PATCH] Add Player canAlwaysOpenCommandBlocks API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 1b4b54741719252892a3a51491d49cd5423de499..cd0d4c0b42687494cf5d93e471e259ca17d2ffe2 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2367,6 +2367,23 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     void sendOpLevel(byte level);
     // Paper end
 
+    // Papyrus start
+    /**
+     * Gets whether this player can always open command blocks and see their content
+     * @return whether this player can always open command blocks
+     */
+    boolean canAlwaysOpenCommandBlocks();
+
+    /**
+     * Sets whether this player can always open command blocks and see their content.
+     * This setting has some minor client side-effects, such as some F3 commands, but will not change
+     * anything server-side unrelated to command blocks opening.
+     * @see <a href="https://github.com/TheKinrar/Papyrus/issues/2">Papyrus#2</a>
+     * @param canAlwaysOpenCommandBlocks whether this player can always open command blocks
+     */
+    void setCanAlwaysOpenCommandBlocks(boolean canAlwaysOpenCommandBlocks);
+    // Papyrus end
+
     // Spigot start
     public class Spigot extends Entity.Spigot {
 
