From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sun, 14 Mar 2021 18:38:42 +0100
Subject: [PATCH] Add noWeight, noDeclaration, maxVectorLength to Minecart


diff --git a/src/main/java/org/bukkit/entity/Minecart.java b/src/main/java/org/bukkit/entity/Minecart.java
index bc1e62759c995a1463ebcd10d73c7d502c9acc45..46a593ddbb4766e8d9354c7e8347c61c806cdea9 100644
--- a/src/main/java/org/bukkit/entity/Minecart.java
+++ b/src/main/java/org/bukkit/entity/Minecart.java
@@ -158,4 +158,18 @@ public interface Minecart extends Vehicle {
     @NotNull
     public Material getMinecartMaterial();
     // Paper end
+
+    // Papyrus start
+    public void setMaxVectorLength(double maxVectorLength);
+
+    public double getMaxVectorLength();
+
+    public void setNoDeceleration(boolean noDeceleration);
+
+    public boolean getNoDeceleration();
+
+    public void setNoWeight(boolean noWeight);
+
+    public boolean getNoWeight();
+    // Papyrus end
 }
