From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Fri, 25 Jun 2021 21:36:42 +0200
Subject: [PATCH] Add disableVanillaEntityCollisions API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 9e3cb75536ae260dc898ab9dafbc1d98398782bc..d4ae6e6c10212054ea886eb86e8841b9a87dbfa9 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1063,4 +1063,10 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     @NotNull String getScoreboardEntryName();
     // Paper end - entity scoreboard name
+
+    // Papyrus start
+    public void setIgnoresVanillaEntityCollisions(boolean ignores);
+
+    public boolean ignoresVanillaEntityCollisions();
+    // Papyrus end
 }
