From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Mon, 4 Nov 2019 19:42:56 +0000
Subject: [PATCH] Hybrid mode API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index d8666481f9a407403d0114ff02024fd3c50c27c4..abdd189edd4317f4ba88509e4a4d769f8138cbdc 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2364,6 +2364,17 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Papyrus start
+    /**
+     * Gets whether the Server is in hybrid mode or not.
+     *
+     * @return true if the server is in hybrid mode, false otherwise
+     */
+    public static boolean getHybridMode() {
+        return server.getHybridMode();
+    }
+    // Papyrus end
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 743a4091e06015b303d1eebc3b6bd393f98e4b9f..dfb373cc4f840eaa10ead2bd6f4ce64366d8159d 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2058,4 +2058,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer();
     // Paper end
+
+    // Papyrus start
+    /**
+     * Gets whether the Server is in hybrid mode or not.
+     *
+     * @return true if the server is in hybrid mode, false otherwise
+     */
+    boolean getHybridMode();
+    // Papyrus end
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 65876d94f5237be583beabed119f34ae8a6ec5f5..157596d844840358fe93549581e9891f26778990 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2568,6 +2568,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param canAlwaysOpenCommandBlocks whether this player can always open command blocks
      */
     void setCanAlwaysOpenCommandBlocks(boolean canAlwaysOpenCommandBlocks);
+
+    /**
+     * Returns whether this player is authenticated with the server.
+     * That is, if the player either used /login to authenticate,
+     * registered for the first time using /register,
+     * was auto-logged in (online ID or auth timeout),
+     * or was force-authenticated with the API.
+     * @return whether this player is authenticated
+     */
+    boolean isAuthenticated();
     // Papyrus end
 
     // Spigot start
diff --git a/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java b/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
index b915c3cab168a2bbdca740e662baaa708d425e41..955c0b6dcce2e61cf30adfb9d9f28e64a7da81f7 100644
--- a/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
@@ -213,6 +213,10 @@ public class EntityPotionEffectEvent extends EntityEvent implements Cancellable
          * When an illusion illager makes himself disappear.
          */
         ILLUSION,
+        /**
+         * When a user is requested to login or register. No event will be triggered for this cause.
+         */
+        HYBRID_AUTH,
         /**
          * When all effects are removed due to a bucket of milk.
          */
