From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Fri, 25 Jun 2021 21:36:42 +0200
Subject: [PATCH] Add disableVanillaEntityCollisions API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 33a6b7a27dc91552799c07a7aad9b3df31ad13f7..7a91b41ecfeb24aabdc74b9b3be60f740c9cd32b 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -951,4 +951,10 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     boolean wouldCollideUsing(@NotNull BoundingBox boundingBox);
     // Paper End - Collision API
+
+    // Papyrus start
+    public void setIgnoresVanillaEntityCollisions(boolean ignores);
+
+    public boolean ignoresVanillaEntityCollisions();
+    // Papyrus end
 }
