From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Fri, 15 Nov 2019 00:12:51 +0000
Subject: [PATCH] Add API for compatibility with legacy selectors in command
 blocks


diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index 460fda05a62b12db2edcfb7ea8b2a5dd8e4b110d..92b66aba6aedc9362cb9d7aa2e1b91f89d5dd061 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -156,7 +156,10 @@ public class SimpleCommandMap implements CommandMap {
         try {
             try (co.aikar.timings.Timing ignored = target.timings.startTiming()) { // Paper - use try with resources
             // Note: we don't return the result of target.execute as thats success / failure, we return handled (true) or not handled (false)
-            target.execute(sender, sentCommandLabel, Arrays.copyOfRange(args, 1, args.length));
+            if(sender instanceof BlockCommandSender)
+                parseSelectors(target, (BlockCommandSender) sender, sentCommandLabel, Arrays.copyOfRange(args, 1, args.length));
+            else
+                target.execute(sender, sentCommandLabel, Arrays.copyOfRange(args, 1, args.length));
             } // target.timings.stopTiming(); // Spigot // Paper
         } catch (CommandException ex) {
             server.getPluginManager().callEvent(new ServerExceptionEvent(new ServerCommandException(ex, target, sender, args))); // Paper
@@ -173,6 +176,13 @@ public class SimpleCommandMap implements CommandMap {
         return true;
     }
 
+    /**
+     * Parse legacy-style selectors - only for compatibility with command blocks from old versions using Bukkit commands
+     */
+    protected boolean parseSelectors(@NotNull Command command, @NotNull BlockCommandSender sender, @NotNull String label, @NotNull String[] args) {
+        return command.execute(sender, label, args);
+    }
+
     @Override
     public synchronized void clearCommands() {
         for (Map.Entry<String, Command> entry : knownCommands.entrySet()) {
