From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Thu, 21 Nov 2019 11:58:39 +0000
Subject: [PATCH] Add Block#getCenter


diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index ca488173a20bcf3427906f0a6548f06a97c9dc36..1b151001ff29dadc52d0207560598a1229592c78 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -238,6 +238,13 @@ public interface Block extends Metadatable {
     @NotNull
     Location getLocation();
 
+    /**
+     * Gets the Location of the center of the block
+     * @return Location of center of block
+     */
+    @NotNull
+    Location getCenter();
+
     /**
      * Stores the location of the block in the provided Location object.
      * <p>
