From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sun, 29 Dec 2019 22:06:28 +0000
Subject: [PATCH] Add Skeleton#getAttackDelay and setAttackDelay


diff --git a/src/main/java/org/bukkit/entity/Skeleton.java b/src/main/java/org/bukkit/entity/Skeleton.java
index 2ca7c040270fa1b0fe623a0a779879783201a2f1..f8d90cd09b31eef5254e5630d6d081e8a73f2bc5 100644
--- a/src/main/java/org/bukkit/entity/Skeleton.java
+++ b/src/main/java/org/bukkit/entity/Skeleton.java
@@ -41,6 +41,12 @@ public interface Skeleton extends AbstractSkeleton, com.destroystokyo.paper.enti
      */
     void setConversionTime(int time);
 
+    // Papyrus start
+    public int getAttackDelay();
+
+    public void setAttackDelay(int delay);
+    // Papyrus end
+
     /**
      * A legacy enum that defines the different variances of skeleton-like
      * entities on the server.
