From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Fri, 25 Jun 2021 21:36:42 +0200
Subject: [PATCH] Add disableVanillaEntityCollisions API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index c2c45a93887f99466a8ac4275355ac75c7f901c4..2ba0346d7a5cf2432966cf01bc71e35070929804 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -822,4 +822,10 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     boolean isInPowderedSnow();
     // Paper end
+
+    // Papyrus start
+    public void setIgnoresVanillaEntityCollisions(boolean ignores);
+
+    public boolean ignoresVanillaEntityCollisions();
+    // Papyrus end
 }
