From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Thu, 21 Nov 2019 11:59:46 +0000
Subject: [PATCH] Implement Block#getCenter


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index bec8e6b62dba2bd0e4e85a7d1fb51287384f1290..9e9d1e654c7ac3144acb1d6a8aa6d87d45fa6370 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -99,6 +99,11 @@ public class CraftBlock implements Block {
         return CraftLocation.toBukkit(this.position, this.getWorld());
     }
 
+    @Override
+    public Location getCenter() {
+        return new Location(getWorld(), position.getX() + 0.5, position.getY() + 0.5, position.getZ() + 0.5);
+    }
+
     @Override
     public Location getLocation(Location loc) {
         if (loc != null) {
