From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Wed, 30 Oct 2019 01:32:35 +0000
Subject: [PATCH] Implement Player canAlwaysOpenCommandBlocks


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
index e572088cad8b9e09b1d64f7971bacac2f10c5b17..83dbe2dcb6d7f6e98052cd56e72a8eee8482544c 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -398,8 +398,9 @@ public class ServerPlayerGameMode {
             BlockEntity tileentity = this.level.getBlockEntity(pos);
             Block block = iblockdata.getBlock();
 
-            if (block instanceof GameMasterBlock && !this.player.canUseGameMasterBlocks() && !(block instanceof net.minecraft.world.level.block.CommandBlock && (this.player.isCreative() && this.player.getBukkitEntity().hasPermission("minecraft.commandblock")))) { // Paper - command block permission
+            if (block instanceof GameMasterBlock && !this.player.canUseGameMasterBlocks() && !(block instanceof net.minecraft.world.level.block.CommandBlock && (this.player.getBukkitEntity().canAlwaysOpenCommandBlocks() || (this.player.isCreative() && this.player.getBukkitEntity().hasPermission("minecraft.commandblock"))))) { // Paper - command block permission
                 this.level.sendBlockUpdated(pos, iblockdata, iblockdata, 3);
+
                 return false;
             } else if (this.player.blockActionRestricted((Level) this.level, pos, this.gameModeForPlayer)) {
                 return false;
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 65657c009f6d5a5d5740e80f912a5893333c7085..67aa995a76eac48fbd8e48c4c4156307ef017218 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1000,6 +1000,9 @@ public abstract class PlayerList {
         GameProfile gameprofile = player.getGameProfile();
         int i = this.server.getProfilePermissions(gameprofile);
 
+        if(player.getBukkitEntity().canAlwaysOpenCommandBlocks() && i < 2)
+            i = 2;
+
         this.sendPlayerPermissionLevel(player, i);
     }
 
diff --git a/src/main/java/net/minecraft/world/item/GameMasterBlockItem.java b/src/main/java/net/minecraft/world/item/GameMasterBlockItem.java
index 8cfce2d85041849ef3943e56c204fefeb07f96ec..b82e4e8220b0dd4444e51aed4b10cc75825875d2 100644
--- a/src/main/java/net/minecraft/world/item/GameMasterBlockItem.java
+++ b/src/main/java/net/minecraft/world/item/GameMasterBlockItem.java
@@ -15,6 +15,9 @@ public class GameMasterBlockItem extends BlockItem {
     @Override
     protected BlockState getPlacementState(BlockPlaceContext context) {
         Player player = context.getPlayer();
-        return player != null && !player.canUseGameMasterBlocks() ? null : super.getPlacementState(context);
+        return player != null && !(
+                player.canUseGameMasterBlocks() ||
+                ((org.bukkit.craftbukkit.entity.CraftPlayer) player.getBukkitEntity()).canAlwaysOpenCommandBlocks()
+        ) ? null : super.getPlacementState(context);
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/CommandBlock.java b/src/main/java/net/minecraft/world/level/block/CommandBlock.java
index fd73e9100feac8c7bf9a5fee21a0ab2d502dc3e0..59fb3a3bfb764efc97f4ee13e2d9d3ad64beeff3 100644
--- a/src/main/java/net/minecraft/world/level/block/CommandBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CommandBlock.java
@@ -26,6 +26,7 @@ import net.minecraft.world.phys.BlockHitResult;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import org.bukkit.craftbukkit.entity.CraftPlayer;
 import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
 
 public class CommandBlock extends BaseEntityBlock implements GameMasterBlock {
@@ -129,7 +130,11 @@ public class CommandBlock extends BaseEntityBlock implements GameMasterBlock {
     public InteractionResult use(BlockState state, Level world, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
         BlockEntity tileentity = world.getBlockEntity(pos);
 
-        if (tileentity instanceof CommandBlockEntity && (player.canUseGameMasterBlocks() || (player.isCreative() && player.getBukkitEntity().hasPermission("minecraft.commandblock")))) { // Paper - command block permission
+        if (tileentity instanceof CommandBlockEntity && (
+                player.canUseGameMasterBlocks()
+                || ((CraftPlayer) player.getBukkitEntity()).canAlwaysOpenCommandBlocks()
+                || (player.isCreative() && player.getBukkitEntity().hasPermission("minecraft.commandblock"))
+        )) {
             player.openCommandBlock((CommandBlockEntity) tileentity);
             return InteractionResult.sidedSuccess(world.isClientSide);
         } else {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 2e41a1bc20f257f4f461a74623a3ffe2d3112fdd..0583a2e299862fb1b0a6c16751f5a5a6f73efc8c 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -152,6 +152,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private static final boolean DISABLE_CHANNEL_LIMIT = System.getProperty("paper.disableChannelLimit") != null; // Paper - add a flag to disable the channel limit
     private long lastSaveTime;
     // Paper end
+    private boolean canAlwaysOpenCommandBlocks = false; // Papyrus
 
     public CraftPlayer(CraftServer server, ServerPlayer entity) {
         super(server, entity);
@@ -2361,6 +2362,20 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
     // Paper end
 
+    // Papyrus start
+    @Override
+    public boolean canAlwaysOpenCommandBlocks() {
+        return canAlwaysOpenCommandBlocks;
+    }
+
+    @Override
+    public void setCanAlwaysOpenCommandBlocks(boolean canAlwaysOpenCommandBlocks) {
+        this.canAlwaysOpenCommandBlocks = canAlwaysOpenCommandBlocks;
+
+        net.minecraft.server.MinecraftServer.getServer().getPlayerList().sendPlayerPermissionLevel(getHandle());
+    }
+    // Papyrus end
+
     // Spigot start
     private final Player.Spigot spigot = new Player.Spigot()
     {
