From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Wed, 30 Oct 2019 01:32:35 +0000
Subject: [PATCH] Implement Player canAlwaysOpenCommandBlocks


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
index 3125af569ec2bb1cd613a9dd96c3a181d723006d..a49bd1bdc188f25ee0af53efbc3842eece527279 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -429,8 +429,9 @@ public class ServerPlayerGameMode {
             BlockEntity tileentity = this.level.getBlockEntity(pos);
             Block block = iblockdata.getBlock();
 
-            if (block instanceof GameMasterBlock && !this.player.canUseGameMasterBlocks() && !(block instanceof net.minecraft.world.level.block.CommandBlock && (this.player.isCreative() && this.player.getBukkitEntity().hasPermission("minecraft.commandblock")))) { // Paper - command block permission
+            if (block instanceof GameMasterBlock && !this.player.canUseGameMasterBlocks() && !(block instanceof net.minecraft.world.level.block.CommandBlock && (this.player.getBukkitEntity().canAlwaysOpenCommandBlocks() || (this.player.isCreative() && this.player.getBukkitEntity().hasPermission("minecraft.commandblock"))))) { // Paper - command block permission
                 this.level.sendBlockUpdated(pos, iblockdata, iblockdata, 3);
+
                 return false;
             } else if (this.player.blockActionRestricted(this.level, pos, this.gameModeForPlayer)) {
                 return false;
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index cd08f9b16c065be8f0eacaeba51d3e72d332daf9..7201f5b8f08d4be4592b75ed3afa35b863b1eba1 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1005,6 +1005,9 @@ public abstract class PlayerList {
         GameProfile gameprofile = player.getGameProfile();
         int i = this.server.getProfilePermissions(gameprofile);
 
+        if(player.getBukkitEntity().canAlwaysOpenCommandBlocks() && i < 2)
+            i = 2;
+
         this.sendPlayerPermissionLevel(player, i);
     }
 
diff --git a/src/main/java/net/minecraft/world/item/GameMasterBlockItem.java b/src/main/java/net/minecraft/world/item/GameMasterBlockItem.java
index 8cfce2d85041849ef3943e56c204fefeb07f96ec..b82e4e8220b0dd4444e51aed4b10cc75825875d2 100644
--- a/src/main/java/net/minecraft/world/item/GameMasterBlockItem.java
+++ b/src/main/java/net/minecraft/world/item/GameMasterBlockItem.java
@@ -15,6 +15,9 @@ public class GameMasterBlockItem extends BlockItem {
     @Override
     protected BlockState getPlacementState(BlockPlaceContext context) {
         Player player = context.getPlayer();
-        return player != null && !player.canUseGameMasterBlocks() ? null : super.getPlacementState(context);
+        return player != null && !(
+                player.canUseGameMasterBlocks() ||
+                ((org.bukkit.craftbukkit.entity.CraftPlayer) player.getBukkitEntity()).canAlwaysOpenCommandBlocks()
+        ) ? null : super.getPlacementState(context);
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/CommandBlock.java b/src/main/java/net/minecraft/world/level/block/CommandBlock.java
index 135f37f79d1463896e47679e1d13e123b78f97c9..ee60c08ce5fb78611fe583af215a29d11f46f033 100644
--- a/src/main/java/net/minecraft/world/level/block/CommandBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CommandBlock.java
@@ -27,6 +27,7 @@ import net.minecraft.world.phys.BlockHitResult;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import org.bukkit.craftbukkit.entity.CraftPlayer;
 import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
 
 public class CommandBlock extends BaseEntityBlock implements GameMasterBlock {
@@ -130,7 +131,11 @@ public class CommandBlock extends BaseEntityBlock implements GameMasterBlock {
     public InteractionResult use(BlockState state, Level world, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
         BlockEntity tileentity = world.getBlockEntity(pos);
 
-        if (tileentity instanceof CommandBlockEntity && (player.canUseGameMasterBlocks() || (player.isCreative() && player.getBukkitEntity().hasPermission("minecraft.commandblock")))) { // Paper - command block permission
+        if (tileentity instanceof CommandBlockEntity && (
+                player.canUseGameMasterBlocks()
+                || ((CraftPlayer) player.getBukkitEntity()).canAlwaysOpenCommandBlocks()
+                || (player.isCreative() && player.getBukkitEntity().hasPermission("minecraft.commandblock"))
+        )) {
             player.openCommandBlock((CommandBlockEntity) tileentity);
             return InteractionResult.sidedSuccess(world.isClientSide);
         } else {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 8cac6f49b23841414fbcc7f3a158a7342b382496..ddad3bc0c32a0a160b0c5409ea74a1b286ecc1db 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -160,6 +160,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private static final boolean DISABLE_CHANNEL_LIMIT = System.getProperty("paper.disableChannelLimit") != null; // Paper - add a flag to disable the channel limit
     private long lastSaveTime;
     // Paper end
+    private boolean canAlwaysOpenCommandBlocks = false; // Papyrus
 
     public CraftPlayer(CraftServer server, ServerPlayer entity) {
         super(server, entity);
@@ -2543,6 +2544,21 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
     }
     // Paper end
+
+    // Papyrus start
+    @Override
+    public boolean canAlwaysOpenCommandBlocks() {
+        return canAlwaysOpenCommandBlocks;
+    }
+
+    @Override
+    public void setCanAlwaysOpenCommandBlocks(boolean canAlwaysOpenCommandBlocks) {
+        this.canAlwaysOpenCommandBlocks = canAlwaysOpenCommandBlocks;
+
+        net.minecraft.server.MinecraftServer.getServer().getPlayerList().sendPlayerPermissionLevel(getHandle());
+    }
+    // Papyrus end
+
     // Spigot start
     private final Player.Spigot spigot = new Player.Spigot()
     {
