From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Thu, 21 Nov 2019 11:59:46 +0000
Subject: [PATCH] Implement Block#getCenter


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index 5401ab9f8f6ce12e1c5368dbc3acc78a250b3822..3ab83a7c294eb4dff6f145e80ac70c8473161834 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -104,6 +104,11 @@ public class CraftBlock implements Block {
         return CraftLocation.toBukkit(position, this.getWorld());
     }
 
+    @Override
+    public Location getCenter() {
+        return new Location(getWorld(), position.getX() + 0.5, position.getY() + 0.5, position.getZ() + 0.5);
+    }
+
     @Override
     public Location getLocation(Location loc) {
         if (loc != null) {
