From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Wed, 30 Oct 2019 01:32:35 +0000
Subject: [PATCH] Implement Player canAlwaysOpenCommandBlocks


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
index 4b756c0a4b607faa03b00ab81761335be63c39eb..8776fbaabdb54577d30d3670d5681926e1bea93f 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -398,8 +398,9 @@ public class ServerPlayerGameMode {
             BlockEntity tileentity = this.level.getBlockEntity(pos);
             Block block = iblockdata.getBlock();
 
-            if (block instanceof GameMasterBlock && !this.player.canUseGameMasterBlocks() && !(block instanceof net.minecraft.world.level.block.CommandBlock && (this.player.isCreative() && this.player.getBukkitEntity().hasPermission("minecraft.commandblock")))) { // Paper - command block permission
+            if (block instanceof GameMasterBlock && !this.player.canUseGameMasterBlocks() && !(block instanceof net.minecraft.world.level.block.CommandBlock && (this.player.getBukkitEntity().canAlwaysOpenCommandBlocks() || (this.player.isCreative() && this.player.getBukkitEntity().hasPermission("minecraft.commandblock"))))) { // Paper - command block permission
                 this.level.sendBlockUpdated(pos, iblockdata, iblockdata, 3);
+
                 return false;
             } else if (this.player.blockActionRestricted((Level) this.level, pos, this.gameModeForPlayer)) {
                 return false;
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index d76d8d8db7921d16f87dd162ccd115e351cde106..a5ad4faf00e11e67c18781715762248524e8fc95 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -992,6 +992,9 @@ public abstract class PlayerList {
         GameProfile gameprofile = player.getGameProfile();
         int i = this.server.getProfilePermissions(gameprofile);
 
+        if(player.getBukkitEntity().canAlwaysOpenCommandBlocks() && i < 2)
+            i = 2;
+
         this.sendPlayerPermissionLevel(player, i);
     }
 
diff --git a/src/main/java/net/minecraft/world/item/ItemRestricted.java b/src/main/java/net/minecraft/world/item/ItemRestricted.java
new file mode 100644
index 0000000000000000000000000000000000000000..782eb93c496a48571fdb9dba86bf87be1f1627d1
--- /dev/null
+++ b/src/main/java/net/minecraft/world/item/ItemRestricted.java
@@ -0,0 +1,28 @@
+package net.minecraft.world.item;
+
+import net.minecraft.world.entity.player.EntityHuman;
+import net.minecraft.world.item.ItemBlock;
+import net.minecraft.world.item.context.BlockActionContext;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.state.IBlockData;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+
+import javax.annotation.Nullable;
+
+public class ItemRestricted extends ItemBlock {
+
+    public ItemRestricted(Block block, Info item_info) {
+        super(block, item_info);
+    }
+
+    @Nullable
+    @Override
+    protected IBlockData c(BlockActionContext blockactioncontext) {
+        EntityHuman entityhuman = blockactioncontext.getEntity();
+
+        return entityhuman != null && !(
+                entityhuman.isCreativeAndOp() ||
+                ((CraftPlayer) entityhuman.getBukkitEntity()).canAlwaysOpenCommandBlocks()
+        ) ? null : super.c(blockactioncontext);
+    }
+}
diff --git a/src/main/java/net/minecraft/world/level/block/CommandBlock.java b/src/main/java/net/minecraft/world/level/block/CommandBlock.java
index fd73e9100feac8c7bf9a5fee21a0ab2d502dc3e0..59fb3a3bfb764efc97f4ee13e2d9d3ad64beeff3 100644
--- a/src/main/java/net/minecraft/world/level/block/CommandBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CommandBlock.java
@@ -26,6 +26,7 @@ import net.minecraft.world.phys.BlockHitResult;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import org.bukkit.craftbukkit.entity.CraftPlayer;
 import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
 
 public class CommandBlock extends BaseEntityBlock implements GameMasterBlock {
@@ -129,7 +130,11 @@ public class CommandBlock extends BaseEntityBlock implements GameMasterBlock {
     public InteractionResult use(BlockState state, Level world, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
         BlockEntity tileentity = world.getBlockEntity(pos);
 
-        if (tileentity instanceof CommandBlockEntity && (player.canUseGameMasterBlocks() || (player.isCreative() && player.getBukkitEntity().hasPermission("minecraft.commandblock")))) { // Paper - command block permission
+        if (tileentity instanceof CommandBlockEntity && (
+                player.canUseGameMasterBlocks()
+                || ((CraftPlayer) player.getBukkitEntity()).canAlwaysOpenCommandBlocks()
+                || (player.isCreative() && player.getBukkitEntity().hasPermission("minecraft.commandblock"))
+        )) {
             player.openCommandBlock((CommandBlockEntity) tileentity);
             return InteractionResult.sidedSuccess(world.isClientSide);
         } else {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 76e2ae09855e0efaaa0856d2f49e4968adbccbdc..7cf40184b42b1ad53428b26ba45b05764b3770b0 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -152,6 +152,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private static final boolean DISABLE_CHANNEL_LIMIT = System.getProperty("paper.disableChannelLimit") != null; // Paper - add a flag to disable the channel limit
     private long lastSaveTime;
     // Paper end
+    private boolean canAlwaysOpenCommandBlocks = false; // Papyrus
 
     public CraftPlayer(CraftServer server, ServerPlayer entity) {
         super(server, entity);
@@ -2359,6 +2360,20 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
     // Paper end
 
+    // Papyrus start
+    @Override
+    public boolean canAlwaysOpenCommandBlocks() {
+        return canAlwaysOpenCommandBlocks;
+    }
+
+    @Override
+    public void setCanAlwaysOpenCommandBlocks(boolean canAlwaysOpenCommandBlocks) {
+        this.canAlwaysOpenCommandBlocks = canAlwaysOpenCommandBlocks;
+
+        net.minecraft.server.MinecraftServer.getServer().getPlayerList().sendPlayerPermissionLevel(getHandle());
+    }
+    // Papyrus end
+
     // Spigot start
     private final Player.Spigot spigot = new Player.Spigot()
     {
