From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sun, 3 Nov 2019 16:08:17 +0000
Subject: [PATCH] Hybrid mode implementation


diff --git a/src/main/java/co/aikar/timings/MinecraftTimings.java b/src/main/java/co/aikar/timings/MinecraftTimings.java
index bb5f079a10b15d7b9f2ab5a8af67c559ffa5b371..32b51674b32b694713c339fc92ce3d7061a00d45 100644
--- a/src/main/java/co/aikar/timings/MinecraftTimings.java
+++ b/src/main/java/co/aikar/timings/MinecraftTimings.java
@@ -49,6 +49,8 @@ public final class MinecraftTimings {
 
     public static final Timing midTickChunkTasks = Timings.ofSafe("Mid Tick Chunk Tasks");
 
+    public static final Timing hybridEngineTimer = Timings.ofSafe("Hybrid Engine");
+
     private static final Map<Class<?>, String> taskNameCache = new MapMaker().weakKeys().makeMap();
 
     private MinecraftTimings() {}
diff --git a/src/main/java/fr/thekinrar/papyrus/Metrics.java b/src/main/java/fr/thekinrar/papyrus/Metrics.java
index 7c284c9b2ffd4145b6cffb73b1cfcc3b5c3fdaae..ff7346e595cbe2b63aa31886a4d521dc3283ad4c 100644
--- a/src/main/java/fr/thekinrar/papyrus/Metrics.java
+++ b/src/main/java/fr/thekinrar/papyrus/Metrics.java
@@ -593,7 +593,7 @@ public class Metrics {
                 }));
 
                 metrics.addCustomChart(new Metrics.SingleLineChart("players", () -> Bukkit.getOnlinePlayers().size()));
-                metrics.addCustomChart(new Metrics.SimplePie("online_mode", () -> Bukkit.getOnlineMode() ? "online" : "offline"));
+                metrics.addCustomChart(new Metrics.SimplePie("online_mode", () -> Bukkit.getOnlineMode() ? "online" : (Bukkit.getHybridMode() ? "hybrid" : "offline")));
                 metrics.addCustomChart(new Metrics.SimplePie("papyrus_version", () -> (Metrics.class.getPackage().getImplementationVersion() != null) ? Metrics.class.getPackage().getImplementationVersion() : "unknown"));
 
                 metrics.addCustomChart(new Metrics.DrilldownPie("java_version", () -> {
diff --git a/src/main/java/fr/thekinrar/papyrus/PapyrusConfig.java b/src/main/java/fr/thekinrar/papyrus/PapyrusConfig.java
index 4168eafc0d8893e9b677a5c4bde415b900cb8262..90d8a45b39c332565311db744ff72627aa61c7da 100644
--- a/src/main/java/fr/thekinrar/papyrus/PapyrusConfig.java
+++ b/src/main/java/fr/thekinrar/papyrus/PapyrusConfig.java
@@ -1,34 +1,24 @@
 package fr.thekinrar.papyrus;
 
-import co.aikar.timings.Timings;
-import co.aikar.timings.TimingsManager;
-import com.destroystokyo.paper.PaperCommand;
-import com.destroystokyo.paper.io.chunk.ChunkTaskManager;
-import com.google.common.base.Strings;
 import com.google.common.base.Throwables;
-import com.google.common.collect.Lists;
 import net.minecraft.server.MinecraftServer;
 import org.bukkit.Bukkit;
 import org.bukkit.ChatColor;
 import org.bukkit.command.Command;
-import org.bukkit.configuration.ConfigurationSection;
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.bukkit.configuration.file.YamlConfiguration;
-import org.spigotmc.SpigotConfig;
-import org.spigotmc.WatchdogThread;
 
 import java.io.File;
 import java.io.IOException;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
-import java.nio.charset.StandardCharsets;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 import java.util.logging.Level;
-import java.util.logging.Logger;
 import java.util.regex.Pattern;
 
 public class PapyrusConfig {
@@ -190,4 +180,9 @@ public class PapyrusConfig {
         config.addDefault(path, def);
         return config.getString(path, config.getString(path));
     }
+
+    public static boolean hybridMode = false;
+    private static void hybridMode() {
+        hybridMode = getBoolean("settings.hybrid-mode", false);
+    }
 }
diff --git a/src/main/java/fr/thekinrar/papyrus/PapyrusPermissions.java b/src/main/java/fr/thekinrar/papyrus/PapyrusPermissions.java
new file mode 100644
index 0000000000000000000000000000000000000000..8d1cd1c5c197e73f955563ec1312a7294e13649f
--- /dev/null
+++ b/src/main/java/fr/thekinrar/papyrus/PapyrusPermissions.java
@@ -0,0 +1,13 @@
+package fr.thekinrar.papyrus;
+
+import org.bukkit.permissions.Permission;
+import org.bukkit.util.permissions.DefaultPermissions;
+
+public final class PapyrusPermissions {
+    private static final String ROOT= "papyrus.command";
+
+    public static void registerPermissions() {
+        Permission parent = DefaultPermissions.registerPermission(ROOT, "Gives the user the ability to use all Papyrus commands");
+        parent.recalculatePermissibles();
+    }
+}
diff --git a/src/main/java/fr/thekinrar/papyrus/hybrid/HybridPlayer.java b/src/main/java/fr/thekinrar/papyrus/hybrid/HybridPlayer.java
new file mode 100644
index 0000000000000000000000000000000000000000..9e689ecad1618108d6f9d3a9a3d15b738cb0c12e
--- /dev/null
+++ b/src/main/java/fr/thekinrar/papyrus/hybrid/HybridPlayer.java
@@ -0,0 +1,46 @@
+package fr.thekinrar.papyrus.hybrid;
+
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
+import org.bukkit.potion.PotionEffect;
+import org.bukkit.potion.PotionEffectType;
+
+public class HybridPlayer {
+
+    private final CraftPlayer player;
+    private final CraftServer server;
+
+    private int ticks = -1;
+
+    public HybridPlayer(CraftPlayer player) {
+        this.player = player;
+        this.server = player.getHandle().server.server;
+    }
+
+    public void tick() {
+        ++ticks;
+
+        if(ticks == 0) {
+            server.getHandle().sendAllPlayerInfo(player.getHandle());
+        }
+
+        if(ticks % 20 == 0)
+            addEffect(PotionEffectType.BLINDNESS, 0);
+    }
+
+    private void addEffect(PotionEffectType type, int amplifier) {
+        player.addPotionEffect(
+                new PotionEffect(type, 100, amplifier, false, false, false),
+                true,
+                EntityPotionEffectEvent.Cause.HYBRID_AUTH
+        );
+    }
+
+    public void clearEffects() {
+        player.removePotionEffect(PotionEffectType.BLINDNESS, EntityPotionEffectEvent.Cause.HYBRID_AUTH);
+
+        player.getHandle().lastSentExp = -1;
+        server.getHandle().sendAllPlayerInfo(player.getHandle());
+    }
+}
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 4d1f2fa55b69f140630354f4f8ca264c78f2ad97..c495feda2b06a8ffebd379dc8a03198949c7920f 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1565,6 +1565,18 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.profiler.popPush("levels");
         Iterator iterator = this.getAllLevels().iterator();
 
+        // Papyrus start
+        if(server.getHybridMode()) {
+            MinecraftTimings.hybridEngineTimer.startTiming();
+            for(ServerPlayer player : playerList.players) {
+                if(player.unauthenticated) {
+                    player.getBukkitEntity().getHybrid().tick();
+                }
+            }
+            MinecraftTimings.hybridEngineTimer.stopTiming();
+        }
+        // Papyrus end
+
         // CraftBukkit start
         // Run tasks that are waiting on processing
         MinecraftTimings.processQueueTimer.startTiming(); // Spigot
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 6d2d2ba4a83d887f4d52188305a4a28c2ee2f014..fcaffd89a004f3244f81da9f005d35e04807036e 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -729,7 +729,12 @@ public class ServerPlayer extends Player {
 
             if (this.totalExperience != this.lastSentExp) {
                 this.lastSentExp = this.totalExperience;
+
+                if(this.unauthenticated) {
+                this.connection.send(new ClientboundSetExperiencePacket(0, 0, 0));
+                } else {
                 this.connection.send(new ClientboundSetExperiencePacket(this.experienceProgress, this.totalExperience, this.experienceLevel));
+                }
             }
 
             if (this.tickCount % 20 == 0) {
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 9a1fe222b328062edc5c3ba28e0ec4b45ee335db..1125ef1f182e3cdcac63f8c34be4009749bdfca8 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1401,6 +1401,13 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
                             }
                             // Paper end
 
+                            // Papyrus start - Hybrid mode
+                            if(player.unauthenticated) {
+                                this.teleport(this.player.getX(), this.player.getY(), this.player.getZ(), this.player.getYRot(), this.player.getXRot());
+                                return;
+                            }
+                            // Papyrus end
+
                             if (!this.player.isChangingDimension() && (!this.player.getLevel().getGameRules().getBoolean(GameRules.RULE_DISABLE_ELYTRA_MOVEMENT_CHECK) || !this.player.isFallFlying())) {
                                 float f2 = this.player.isFallFlying() ? 300.0F : 100.0F;
 
@@ -1720,6 +1727,8 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
 
         switch (packetplayinblockdig_enumplayerdigtype) {
             case SWAP_ITEM_WITH_OFFHAND:
+                if(player.unauthenticated) return; // Papyrus
+
                 if (!this.player.isSpectator()) {
                     ItemStack itemstack = this.player.getItemInHand(InteractionHand.OFF_HAND);
 
@@ -1747,6 +1756,8 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
 
                 return;
             case DROP_ITEM:
+                if(player.unauthenticated) return; // Papyrus
+
                 if (!this.player.isSpectator()) {
                     // limit how quickly items can be dropped
                     // If the ticks aren't the same then the count starts from 0 and we update the lastDropTick.
@@ -1768,12 +1779,16 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
 
                 return;
             case DROP_ALL_ITEMS:
+                if(player.unauthenticated) return; // Papyrus
+
                 if (!this.player.isSpectator()) {
                     this.player.drop(true);
                 }
 
                 return;
             case RELEASE_USE_ITEM:
+                if(player.unauthenticated) return; // Papyrus
+
                 this.player.releaseUsingItem();
                 return;
             case START_DESTROY_BLOCK:
@@ -2105,6 +2120,10 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
 
             // CraftBukkit start
             boolean isSync = s.startsWith("/");
+            if(player.unauthenticated && !isSync) {
+                ServerGamePacketListenerImpl.LOGGER.warn(this.player.getScoreboardName() + " tried chatting while unauthenticated: " + s);
+                return;
+            }
             if (isSync) {
                 try {
                     this.server.server.playerCommandState = true;
@@ -3154,6 +3173,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
     // CraftBukkit start
     private static final ResourceLocation CUSTOM_REGISTER = new ResourceLocation("register");
     private static final ResourceLocation CUSTOM_UNREGISTER = new ResourceLocation("unregister");
+    private static final ResourceLocation CUSTOM_HYBRID = new ResourceLocation("bungeecord:hybrid");
 
     private static final ResourceLocation MINECRAFT_BRAND = new ResourceLocation("brand"); // Paper - Brand support
 
@@ -3180,6 +3200,18 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
                 ServerGamePacketListenerImpl.LOGGER.error("Couldn\'t unregister custom payload", ex);
                 this.disconnect("Invalid payload UNREGISTER!", org.bukkit.event.player.PlayerKickEvent.Cause.INVALID_PAYLOAD); // Paper - kick event cause
             }
+        } else if (packet.identifier.equals(CUSTOM_HYBRID)) {
+            try {
+                String[] data = packet.data.toString(java.nio.charset.StandardCharsets.UTF_8).split("\0");
+                if("authenticate".equals(data[0])) {
+                    player.unauthenticated = false;
+                    player.getBukkitEntity().getHybrid().clearEffects();
+                } else {
+                    throw new IllegalArgumentException("Unknown subchannel: " + data[0]);
+                }
+            } catch (Exception ex) {
+                ServerGamePacketListenerImpl.LOGGER.error("Couldn't process hybrid custom payload", ex);
+            }
         } else {
             try {
                 byte[] data = new byte[packet.data.readableBytes()];
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index c361775f1f928cfbc35ca3a0acd03e25fb42b82e..c671c1ba69b68d2339d3bceae5734e9cd7811a06 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -233,6 +233,12 @@ public abstract class PlayerList {
 
         if (nbttagcompound == null) player.fudgeSpawnLocation(worldserver1); // Paper - only move to spawn on first login, otherwise, stay where you are....
 
+        // Papyrus start
+        if(cserver.getHybridMode()) {
+            player.getBukkitEntity().initHybrid();
+        }
+        // Papyrus end
+
         player.setLevel(worldserver1);
         String s1 = "local";
 
@@ -275,7 +281,7 @@ public abstract class PlayerList {
         playerconnection.send(new ClientboundSetCarriedItemPacket(player.getInventory().selected));
         playerconnection.send(new ClientboundUpdateRecipesPacket(this.server.getRecipeManager().getRecipes()));
         playerconnection.send(new ClientboundUpdateTagsPacket(this.server.getTags().serializeToNetwork((RegistryAccess) this.registryHolder)));
-        playerconnection.send(new ClientboundEntityEventPacket(player, (byte) (worldserver1.getGameRules().getBoolean(GameRules.RULE_REDUCEDDEBUGINFO) ? 22 : 23))); // Paper - fix this rule not being initialized on the client
+        playerconnection.send(new ClientboundEntityEventPacket(player, (byte) (hasReducedDebugInfo(player) ? 22 : 23))); // Paper - fix this rule not being initialized on the client
         this.sendPlayerPermissionLevel(player);
         player.getStats().markAllDirty();
         player.getRecipeBook().sendInitialRecipeBook(player);
@@ -1307,18 +1313,32 @@ public abstract class PlayerList {
     }
 
     public void sendAllPlayerInfo(ServerPlayer player) {
-        player.inventoryMenu.sendAllDataToRemote();
+        // Papyrus start
+        if(player.unauthenticated)
+            new net.minecraft.world.inventory.InventoryMenu(new net.minecraft.world.entity.player.Inventory(player), true, player).sendAllDataToRemote();
+        else // Papyrus end
+            player.inventoryMenu.sendAllDataToRemote();
         // entityplayer.triggerHealthUpdate();
         player.getBukkitEntity().updateScaledHealth(); // CraftBukkit - Update scaled health on respawn and worldchange
         player.connection.send(new ClientboundSetCarriedItemPacket(player.getInventory().selected));
         // CraftBukkit start - from GameRules
-        int i = player.level.getGameRules().getBoolean(GameRules.RULE_REDUCEDDEBUGINFO) ? 22 : 23;
-        player.connection.send(new ClientboundEntityEventPacket(player, (byte) i));
-        float immediateRespawn = player.level.getGameRules().getBoolean(GameRules.RULE_DO_IMMEDIATE_RESPAWN) ? 1.0F: 0.0F;
+        updateReducedDebugInfo(player); // Papyrus
+        float immediateRespawn = player.level.getGameRules().getBoolean(GameRules.RULE_DO_IMMEDIATE_RESPAWN) ? 1.0F : 0.0F;
         player.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.IMMEDIATE_RESPAWN, immediateRespawn));
         // CraftBukkit end
     }
 
+    // Papyrus start
+    public void updateReducedDebugInfo(ServerPlayer player) {
+        int i = hasReducedDebugInfo(player) ? 22 : 23;
+        player.connection.send(new ClientboundEntityEventPacket(player, (byte) i));
+    }
+    // Papyrus end
+
+    private boolean hasReducedDebugInfo(ServerPlayer player) {
+        return player.level.getGameRules().getBoolean(GameRules.RULE_REDUCEDDEBUGINFO) || player.unauthenticated;
+    }
+
     public int getPlayerCount() {
         return this.players.size();
     }
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 1149a15486016ac101c5976b45b7d1c1109244ce..7c3f2818943a2af1e415e1ccd53660061d5ce542 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1107,9 +1107,12 @@ public abstract class LivingEntity extends Entity {
                 override = new MobEffectInstance(mobeffect1).update(mobeffect);
             }
 
-            EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, mobeffect1, mobeffect, cause, override);
-            if (event.isCancelled()) {
-                return false;
+            if(!cause.equals(EntityPotionEffectEvent.Cause.HYBRID_AUTH)) { // Papyrus
+                EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, mobeffect1, mobeffect, cause, override);
+                if (event.isCancelled()) {
+                    return false;
+                }
+                override = event.isOverride(); // Papyrus - see below
             }
             // CraftBukkit end
 
@@ -1118,7 +1121,7 @@ public abstract class LivingEntity extends Entity {
                 this.onEffectAdded(mobeffect, entity);
                 return true;
                 // CraftBukkit start
-            } else if (event.isOverride()) {
+            } else if (override) { // Papyrus - read override directly
                 mobeffect1.update(mobeffect);
                 this.onEffectUpdated(mobeffect1, true, entity);
                 // CraftBukkit end
@@ -1176,9 +1179,11 @@ public abstract class LivingEntity extends Entity {
             return null;
         }
 
-        EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, effect, null, cause);
-        if (event.isCancelled()) {
-            return null;
+        if(!cause.equals(EntityPotionEffectEvent.Cause.HYBRID_AUTH)) {
+            EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, effect, null, cause);
+            if (event.isCancelled()) {
+                return null;
+            }
         }
 
         return (MobEffectInstance) this.activeEffects.remove(mobeffectlist);
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index cea92f1dc663bf0648b2bd877d86ca380a517bc9..09c23f5fb0c40ce4d631c0e2f7279abbd1404811 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -193,6 +193,10 @@ public abstract class Player extends LivingEntity {
     }
     // CraftBukkit end
 
+    // Papyrus start
+    public boolean unauthenticated = fr.thekinrar.papyrus.PapyrusConfig.hybridMode;
+    // Papyrus end
+
     public Player(Level world, BlockPos pos, float yaw, GameProfile profile) {
         super(EntityType.PLAYER, world);
         this.lastItemInMainHand = ItemStack.EMPTY;
@@ -745,10 +749,7 @@ public abstract class Player extends LivingEntity {
             org.bukkit.entity.Player player = (org.bukkit.entity.Player) this.getBukkitEntity();
             Item drop = (Item) entityitem.getBukkitEntity();
 
-            PlayerDropItemEvent event = new PlayerDropItemEvent(player, drop);
-            this.level.getCraftServer().getPluginManager().callEvent(event);
-
-            if (event.isCancelled() && !alwaysSucceed) { // Paper - Fix duplicating /give items on item drop cancel
+            if (hybridCall(new PlayerDropItemEvent(player, drop)) && !alwaysSucceed) { // Paper - Fix duplicating /give items on item drop cancel
                 org.bukkit.inventory.ItemStack cur = player.getInventory().getItemInHand();
                 if (retainOwnership && (cur == null || cur.getAmount() == 0)) {
                     // The complete stack was dropped
@@ -777,6 +778,14 @@ public abstract class Player extends LivingEntity {
         }
     }
 
+    public boolean hybridCall(org.bukkit.event.Event event) {
+        if(unauthenticated)
+            return true;
+
+        this.level.getCraftServer().getPluginManager().callEvent(event);
+        return event instanceof org.bukkit.event.Cancellable && ((org.bukkit.event.Cancellable) event).isCancelled();
+    }
+
     public float getDestroySpeed(BlockState block) {
         float f = this.inventory.getDestroySpeed(block);
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 5994d8732f7e28ca391910277dd62004c1927111..d33898311e46322e4cf31db04c69b737aedddcb5 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -462,6 +462,7 @@ public final class CraftServer implements Server {
             if (!com.destroystokyo.paper.PaperConfig.loadPermsBeforePlugins) this.loadCustomPermissions(); // Paper
             this.helpMap.initializeCommands();
             this.syncCommands();
+            fr.thekinrar.papyrus.PapyrusPermissions.registerPermissions(); // Papyrus
         }
     }
 
@@ -944,6 +945,7 @@ public final class CraftServer implements Server {
         com.destroystokyo.paper.PaperConfig.registerCommands(); // Paper
         this.overrideAllCommandBlockCommands = this.commandsConfiguration.getStringList("command-block-overrides").contains("*");
         this.ignoreVanillaPermissions = this.commandsConfiguration.getBoolean("ignore-vanilla-permissions");
+        fr.thekinrar.papyrus.PapyrusConfig.registerCommands(); // Papyrus
 
         int pollCount = 0;
 
@@ -2668,4 +2670,24 @@ public final class CraftServer implements Server {
     }
 
     // Paper end
+
+    // Papyrus start
+    @Override
+    public boolean getHybridMode() {
+        if (fr.thekinrar.papyrus.PapyrusConfig.hybridMode) {
+            if (getOnlineMode()) {
+                getLogger().warning("Hybrid mode enabled but server is in online mode. Disabling hybrid mode.");
+                fr.thekinrar.papyrus.PapyrusConfig.hybridMode = false;
+            } else if(!org.spigotmc.SpigotConfig.bungee) {
+                getLogger().warning("Hybrid mode enabled but bungee support is disabled. Disabling hybrid mode.");
+                fr.thekinrar.papyrus.PapyrusConfig.hybridMode = false;
+            } else if(com.destroystokyo.paper.PaperConfig.bungeeOnlineMode) {
+                getLogger().warning("Hybrid mode enabled but bungee support is online mode. Disabling hybrid mode.");
+                fr.thekinrar.papyrus.PapyrusConfig.hybridMode = false;
+            }
+        }
+
+        return fr.thekinrar.papyrus.PapyrusConfig.hybridMode;
+    }
+    // Papyrus end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index fe71123dc07cb13ffad8f13e57aa9bda1cb0abf5..1ab1aa1515c3364baa46a2ffda86a4c326b1500c 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -436,7 +436,11 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
 
     @Override
     public boolean addPotionEffect(PotionEffect effect, boolean force) {
-        this.getHandle().addEffect(new MobEffectInstance(MobEffect.byId(effect.getType().getId()), effect.getDuration(), effect.getAmplifier(), effect.isAmbient(), effect.hasParticles(), effect.hasIcon()), EntityPotionEffectEvent.Cause.PLUGIN); // Paper - Don't ignore icon
+        return addPotionEffect(effect, force, EntityPotionEffectEvent.Cause.PLUGIN);
+    }
+
+    public boolean addPotionEffect(PotionEffect effect, boolean force, EntityPotionEffectEvent.Cause cause) {
+        this.getHandle().addEffect(new MobEffectInstance(MobEffect.byId(effect.getType().getId()), effect.getDuration(), effect.getAmplifier(), effect.isAmbient(), effect.hasParticles(), effect.hasIcon()), cause);
         return true;
     }
 
@@ -465,6 +469,10 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         this.getHandle().removeEffect(MobEffect.byId(type.getId()), EntityPotionEffectEvent.Cause.PLUGIN);
     }
 
+    public void removePotionEffect(PotionEffectType type, EntityPotionEffectEvent.Cause cause) {
+        this.getHandle().removeEffect(MobEffect.byId(type.getId()), cause);
+    }
+
     @Override
     public Collection<PotionEffect> getActivePotionEffects() {
         List<PotionEffect> effects = new ArrayList<PotionEffect>();
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index a8031f894154a6da5904d1c5c4b8f89b1b4880d3..775e23bbc29eb6b7f9a44d17249da6dc780da031 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -154,6 +154,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private long lastSaveTime;
     // Paper end
     private boolean canAlwaysOpenCommandBlocks = false; // Papyrus
+    private fr.thekinrar.papyrus.hybrid.HybridPlayer hybrid;
 
     public CraftPlayer(CraftServer server, ServerPlayer entity) {
         super(server, entity);
@@ -161,6 +162,10 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.firstPlayed = System.currentTimeMillis();
     }
 
+    public fr.thekinrar.papyrus.hybrid.HybridPlayer getHybrid() {
+        return hybrid;
+    }
+
     public GameProfile getProfile() {
         return this.getHandle().getGameProfile();
     }
@@ -2448,6 +2453,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
         net.minecraft.server.MinecraftServer.getServer().getPlayerList().sendPlayerPermissionLevel(getHandle());
     }
+
+    @Override
+    public boolean isAuthenticated() {
+        return !getHandle().unauthenticated;
+    }
+
+    public void initHybrid() {
+        if(hybrid == null) {
+            hybrid = new fr.thekinrar.papyrus.hybrid.HybridPlayer(this);
+        }
+    }
     // Papyrus end
 
     // Spigot start
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 31bbf018e3b7cafb1e186c7d00b421fb133331a3..8833e4c6eca0473a9bd674e345ce84cffa8eee21 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -539,7 +539,12 @@ public class CraftEventFactory {
         if (cancelledBlock) {
             event.setUseInteractedBlock(Event.Result.DENY);
         }
-        craftServer.getPluginManager().callEvent(event);
+
+        if(who != null && who.unauthenticated) {
+            event.setCancelled(true);
+        } else {
+            craftServer.getPluginManager().callEvent(event);
+        }
 
         return event;
     }
