From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Fri, 25 Jun 2021 21:37:27 +0200
Subject: [PATCH] Implement disableVanillaEntityCollisions API


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 6fa47becd0f83ac4273ef3a10c314aa27b08184b..577e5c70b92158a2bb8daefba34b29ba25155d30 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -397,6 +397,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     @javax.annotation.Nullable
     private UUID originWorld;
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
+    public boolean disableVanillaEntityCollisions; // Papyrus
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -1349,6 +1350,8 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             return movement;
         }
 
+        if(disableVanillaEntityCollisions) return moveVector; // Papyrus
+
         final Level world = this.level;
         final AABB currBoundingBox = this.getBoundingBox();
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 9368ec01e498f913bc5b7b3e77fe87659090d9b5..cb1bdbd61400f34eb46f9e1e7d6af08af7c77f85 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1369,6 +1369,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return getHandle().isInPowderSnow || getHandle().wasInPowderSnow; // depending on the location in the entity "tick" either could be needed.
     }
     // Paper end
+
     // Paper Start - Collision API
     @Override
     public boolean collidesAt(@org.jetbrains.annotations.NotNull Location location) {
@@ -1384,4 +1385,16 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return !this.getHandle().level.noCollision(this.getHandle(), aabb);
     }
     // Paper End - Collision API
+
+    // Papyrus start
+    @Override
+    public void setIgnoresVanillaEntityCollisions(boolean ignores) {
+        getHandle().disableVanillaEntityCollisions = ignores;
+    }
+
+    @Override
+    public boolean ignoresVanillaEntityCollisions() {
+        return getHandle().disableVanillaEntityCollisions;
+    }
+    // Papyrus end
 }
