From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Fri, 25 Jun 2021 21:37:27 +0200
Subject: [PATCH] Implement disableVanillaEntityCollisions API


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index f925a8d550ecbf2044a37bfe58b30d6578c5f6af..f270dd9542d15a3784c296ebc91cf7b579eb70a6 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -398,6 +398,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     @javax.annotation.Nullable
     private UUID originWorld;
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
+    public boolean disableVanillaEntityCollisions; // Papyrus
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -1349,6 +1350,8 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             return movement;
         }
 
+        if(disableVanillaEntityCollisions) return moveVector; // Papyrus
+
         final Level world = this.level;
         final AABB currBoundingBox = this.getBoundingBox();
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 6549ade8e19807c523e5a1dc68b66585aad438b1..0bf762105d1ab45964efaa291b399c33289d3f12 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1340,6 +1340,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return getHandle().isInPowderSnow || getHandle().wasInPowderSnow; // depending on the location in the entity "tick" either could be needed.
     }
     // Paper end
+
     // Paper Start - Collision API
     @Override
     public boolean collidesAt(@org.jetbrains.annotations.NotNull Location location) {
@@ -1355,4 +1356,16 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return !this.getHandle().level.noCollision(this.getHandle(), aabb);
     }
     // Paper End - Collision API
+
+    // Papyrus start
+    @Override
+    public void setIgnoresVanillaEntityCollisions(boolean ignores) {
+        getHandle().disableVanillaEntityCollisions = ignores;
+    }
+
+    @Override
+    public boolean ignoresVanillaEntityCollisions() {
+        return getHandle().disableVanillaEntityCollisions;
+    }
+    // Papyrus end
 }
