From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Thu, 21 Nov 2019 11:59:46 +0000
Subject: [PATCH] Implement Block#getCenter


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index ca320ef8a28fe15b32197669419981c4c2fcafd9..1e8db9aa758e1ee34d1d10509c95f0c05217fd22 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -95,6 +95,11 @@ public class CraftBlock implements Block {
         return new Location(this.getWorld(), this.position.getX(), this.position.getY(), this.position.getZ());
     }
 
+    @Override
+    public Location getCenter() {
+        return new Location(getWorld(), position.getX() + 0.5, position.getY() + 0.5, position.getZ() + 0.5);
+    }
+
     @Override
     public Location getLocation(Location loc) {
         if (loc != null) {
