From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Fri, 25 Jun 2021 21:37:27 +0200
Subject: [PATCH] Implement disableVanillaEntityCollisions API


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 513c34aa02d63f7e3c178eade818e156af4541db..f38a30fe3f120d2c419146e6c04b0f4136237ae7 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -417,6 +417,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     private UUID originWorld;
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
     public boolean collidingWithWorldBorder; // Paper
+    public boolean disableVanillaEntityCollisions; // Papyrus
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -1450,6 +1451,8 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             return movement;
         }
 
+        if(disableVanillaEntityCollisions) return moveVector; // Papyrus
+
         final Level world = this.level;
         final AABB currBoundingBox = this.getBoundingBox();
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index fc0dc8e607cc24020106ea1af92b4421a5f9393d..bf8b73017a928fd5060db4e26be5cbc74606c565 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1435,6 +1435,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return getHandle().isInPowderSnow || getHandle().wasInPowderSnow; // depending on the location in the entity "tick" either could be needed.
     }
     // Paper end
+
     // Paper Start - Collision API
     @Override
     public boolean collidesAt(@org.jetbrains.annotations.NotNull Location location) {
@@ -1450,4 +1451,16 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return !this.getHandle().level().noCollision(this.getHandle(), aabb);
     }
     // Paper End - Collision API
+
+    // Papyrus start
+    @Override
+    public void setIgnoresVanillaEntityCollisions(boolean ignores) {
+        getHandle().disableVanillaEntityCollisions = ignores;
+    }
+
+    @Override
+    public boolean ignoresVanillaEntityCollisions() {
+        return getHandle().disableVanillaEntityCollisions;
+    }
+    // Papyrus end
 }
