From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sat, 23 Nov 2019 02:29:24 +0000
Subject: [PATCH] Case-insensitive command matching for command blocks

(Which matches the behaviour for players and console)

diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index 6fec6a47538da4c0c5a4505e9bedf492bb3376dd..de914f586ef1a3708cf3542e966382e5ed137a62 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -238,6 +238,7 @@ public class Commands {
 
         String[] args = command.split(" ");
         if (args.length == 0) return 0; // Paper - empty commands shall not be dispatched
+        args[0] = args[0].toLowerCase(); // Papyrus - lower case match
 
         String cmd = args[0];
         if (cmd.startsWith("minecraft:")) cmd = cmd.substring("minecraft:".length());
