From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sun, 3 Nov 2019 16:08:17 +0000
Subject: [PATCH] Hybrid mode implementation


diff --git a/src/main/java/co/aikar/timings/MinecraftTimings.java b/src/main/java/co/aikar/timings/MinecraftTimings.java
index 112029cb275d45dced60807820f1bfe9f394496d..38f624e7b8ef2b76433aa4d19a05000ff20c4352 100644
--- a/src/main/java/co/aikar/timings/MinecraftTimings.java
+++ b/src/main/java/co/aikar/timings/MinecraftTimings.java
@@ -51,6 +51,8 @@ public final class MinecraftTimings {
 
     public static final Timing midTickChunkTasks = Timings.ofSafe("Mid Tick Chunk Tasks");
 
+    public static final Timing hybridEngineTimer = Timings.ofSafe("Hybrid Engine");
+
     private static final Map<Class<?>, String> taskNameCache = new MapMaker().weakKeys().makeMap();
 
     private MinecraftTimings() {}
diff --git a/src/main/java/fr/thekinrar/papyrus/Metrics.java b/src/main/java/fr/thekinrar/papyrus/Metrics.java
index 7c284c9b2ffd4145b6cffb73b1cfcc3b5c3fdaae..ff7346e595cbe2b63aa31886a4d521dc3283ad4c 100644
--- a/src/main/java/fr/thekinrar/papyrus/Metrics.java
+++ b/src/main/java/fr/thekinrar/papyrus/Metrics.java
@@ -593,7 +593,7 @@ public class Metrics {
                 }));
 
                 metrics.addCustomChart(new Metrics.SingleLineChart("players", () -> Bukkit.getOnlinePlayers().size()));
-                metrics.addCustomChart(new Metrics.SimplePie("online_mode", () -> Bukkit.getOnlineMode() ? "online" : "offline"));
+                metrics.addCustomChart(new Metrics.SimplePie("online_mode", () -> Bukkit.getOnlineMode() ? "online" : (Bukkit.getHybridMode() ? "hybrid" : "offline")));
                 metrics.addCustomChart(new Metrics.SimplePie("papyrus_version", () -> (Metrics.class.getPackage().getImplementationVersion() != null) ? Metrics.class.getPackage().getImplementationVersion() : "unknown"));
 
                 metrics.addCustomChart(new Metrics.DrilldownPie("java_version", () -> {
diff --git a/src/main/java/fr/thekinrar/papyrus/PapyrusConfig.java b/src/main/java/fr/thekinrar/papyrus/PapyrusConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..78e39d50f83070aa94677ddc1a6a1400fce713df
--- /dev/null
+++ b/src/main/java/fr/thekinrar/papyrus/PapyrusConfig.java
@@ -0,0 +1,143 @@
+package fr.thekinrar.papyrus;
+
+import com.google.common.base.Throwables;
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.Bukkit;
+import org.bukkit.command.Command;
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.configuration.file.YamlConfiguration;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+
+public class PapyrusConfig
+{
+
+    private static File CONFIG_FILE;
+    /*========================================================================*/
+    public static YamlConfiguration config;
+    static int version;
+    static Map<String, Command> commands;
+    private static boolean verbose;
+    /*========================================================================*/
+    private static boolean metricsStarted;
+
+    public static void init(File configFile)
+    {
+        CONFIG_FILE = configFile;
+        config = new YamlConfiguration();
+        try
+        {
+            config.load( CONFIG_FILE );
+        } catch ( IOException ex )
+        {
+        } catch ( InvalidConfigurationException ex )
+        {
+            Bukkit.getLogger().log( Level.SEVERE, "Could not load papyrus.yml, please correct your syntax errors", ex );
+            throw Throwables.propagate( ex );
+        }
+
+        config.options().copyDefaults( true );
+        verbose = getBoolean("verbose", false);
+
+        commands = new HashMap<String, Command>();
+
+        version = getInt( "config-version", 1 );
+        set( "config-version", 1 );
+        readConfig( PapyrusConfig.class, null );
+    }
+
+    public static void registerCommands()
+    {
+        for ( Map.Entry<String, Command> entry : commands.entrySet() )
+        {
+            MinecraftServer.getServer().server.getCommandMap().register( entry.getKey(), "Papyrus", entry.getValue() );
+        }
+
+        if(!metricsStarted) {
+            Metrics.PapyrusMetrics.startMetrics();
+            metricsStarted = true;
+        }
+    }
+
+    public static void readConfig(Class<?> clazz, Object instance) // Paper - package-private -> public
+    {
+        for ( Method method : clazz.getDeclaredMethods() )
+        {
+            if ( Modifier.isPrivate( method.getModifiers() ) )
+            {
+                if ( method.getParameterTypes().length == 0 && method.getReturnType() == Void.TYPE )
+                {
+                    try
+                    {
+                        method.setAccessible( true );
+                        method.invoke( instance );
+                    } catch ( InvocationTargetException ex )
+                    {
+                        throw Throwables.propagate( ex.getCause() );
+                    } catch ( Exception ex )
+                    {
+                        Bukkit.getLogger().log( Level.SEVERE, "Error invoking " + method, ex );
+                    }
+                }
+            }
+        }
+        save();
+    }
+    public static void save() {
+        try
+        {
+            config.save( CONFIG_FILE );
+        } catch ( IOException ex )
+        {
+            Bukkit.getLogger().log( Level.SEVERE, "Could not save " + CONFIG_FILE, ex );
+        }
+    }
+
+    private static void set(String path, Object val)
+    {
+        config.set( path, val );
+    }
+
+    private static boolean getBoolean(String path, boolean def)
+    {
+        config.addDefault( path, def );
+        return config.getBoolean( path, config.getBoolean( path ) );
+    }
+
+    private static int getInt(String path, int def)
+    {
+        config.addDefault( path, def );
+        return config.getInt( path, config.getInt( path ) );
+    }
+
+    private static <T> List getList(String path, T def)
+    {
+        config.addDefault( path, def );
+        return (List<T>) config.getList( path, config.getList( path ) );
+    }
+
+    private static String getString(String path, String def)
+    {
+        config.addDefault( path, def );
+        return config.getString( path, config.getString( path ) );
+    }
+
+    private static double getDouble(String path, double def)
+    {
+        config.addDefault( path, def );
+        return config.getDouble( path, config.getDouble( path ) );
+    }
+
+    public static boolean hybridMode = false;
+    private static void hybridMode() {
+        hybridMode = getBoolean("settings.hybrid-mode", false);
+    }
+}
diff --git a/src/main/java/fr/thekinrar/papyrus/PapyrusPermissions.java b/src/main/java/fr/thekinrar/papyrus/PapyrusPermissions.java
new file mode 100644
index 0000000000000000000000000000000000000000..8d1cd1c5c197e73f955563ec1312a7294e13649f
--- /dev/null
+++ b/src/main/java/fr/thekinrar/papyrus/PapyrusPermissions.java
@@ -0,0 +1,13 @@
+package fr.thekinrar.papyrus;
+
+import org.bukkit.permissions.Permission;
+import org.bukkit.util.permissions.DefaultPermissions;
+
+public final class PapyrusPermissions {
+    private static final String ROOT= "papyrus.command";
+
+    public static void registerPermissions() {
+        Permission parent = DefaultPermissions.registerPermission(ROOT, "Gives the user the ability to use all Papyrus commands");
+        parent.recalculatePermissibles();
+    }
+}
diff --git a/src/main/java/fr/thekinrar/papyrus/hybrid/HybridPlayer.java b/src/main/java/fr/thekinrar/papyrus/hybrid/HybridPlayer.java
new file mode 100644
index 0000000000000000000000000000000000000000..9e689ecad1618108d6f9d3a9a3d15b738cb0c12e
--- /dev/null
+++ b/src/main/java/fr/thekinrar/papyrus/hybrid/HybridPlayer.java
@@ -0,0 +1,46 @@
+package fr.thekinrar.papyrus.hybrid;
+
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
+import org.bukkit.potion.PotionEffect;
+import org.bukkit.potion.PotionEffectType;
+
+public class HybridPlayer {
+
+    private final CraftPlayer player;
+    private final CraftServer server;
+
+    private int ticks = -1;
+
+    public HybridPlayer(CraftPlayer player) {
+        this.player = player;
+        this.server = player.getHandle().server.server;
+    }
+
+    public void tick() {
+        ++ticks;
+
+        if(ticks == 0) {
+            server.getHandle().sendAllPlayerInfo(player.getHandle());
+        }
+
+        if(ticks % 20 == 0)
+            addEffect(PotionEffectType.BLINDNESS, 0);
+    }
+
+    private void addEffect(PotionEffectType type, int amplifier) {
+        player.addPotionEffect(
+                new PotionEffect(type, 100, amplifier, false, false, false),
+                true,
+                EntityPotionEffectEvent.Cause.HYBRID_AUTH
+        );
+    }
+
+    public void clearEffects() {
+        player.removePotionEffect(PotionEffectType.BLINDNESS, EntityPotionEffectEvent.Cause.HYBRID_AUTH);
+
+        player.getHandle().lastSentExp = -1;
+        server.getHandle().sendAllPlayerInfo(player.getHandle());
+    }
+}
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index f741888b20268f6d4588763653839f086759f0db..f7f9efe5e2511f75ff455474ee8564da6510ed1f 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1501,6 +1501,18 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.profiler.popPush("levels");
         //Iterator iterator = this.getAllLevels().iterator(); // Paper - moved down
 
+        // Papyrus start
+        if(server.getHybridMode()) {
+            MinecraftTimings.hybridEngineTimer.startTiming();
+            for(ServerPlayer player : playerList.players) {
+                if(player.unauthenticated) {
+                    player.getBukkitEntity().getHybrid().tick();
+                }
+            }
+            MinecraftTimings.hybridEngineTimer.stopTiming();
+        }
+        // Papyrus end
+
         // CraftBukkit start
         // Run tasks that are waiting on processing
         MinecraftTimings.processQueueTimer.startTiming(); // Spigot
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index dc5fda83a66afbfeb7897fc20b4742899d8aca08..5d771e4e8889a843e9c87d3c4a33e8e43f6c5413 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -815,7 +815,12 @@ public class ServerPlayer extends Player {
 
             if (this.totalExperience != this.lastSentExp) {
                 this.lastSentExp = this.totalExperience;
+
+                if(this.unauthenticated) {
+                this.connection.send(new ClientboundSetExperiencePacket(0, 0, 0));
+                } else {
                 this.connection.send(new ClientboundSetExperiencePacket(this.experienceProgress, this.totalExperience, this.experienceLevel));
+                }
             }
 
             if (this.tickCount % 20 == 0) {
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 80e050603e36134754136a8ee2bd14624f976e91..b67371e1a4a01e0d7bb332815b114c081f94a502 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1484,6 +1484,13 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
                             }
                             // Paper end
 
+                            // Papyrus start - Hybrid mode
+                            if(player.unauthenticated) {
+                                this.teleport(this.player.getX(), this.player.getY(), this.player.getZ(), this.player.getYRot(), this.player.getXRot());
+                                return;
+                            }
+                            // Papyrus end
+
                             if (!this.player.isChangingDimension() && (!this.player.level().getGameRules().getBoolean(GameRules.RULE_DISABLE_ELYTRA_MOVEMENT_CHECK) || !this.player.isFallFlying())) {
                                 float f2 = this.player.isFallFlying() ? 300.0F : 100.0F;
 
@@ -1843,6 +1850,8 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
 
         switch (packetplayinblockdig_enumplayerdigtype) {
             case SWAP_ITEM_WITH_OFFHAND:
+                if(player.unauthenticated) return; // Papyrus
+
                 if (!this.player.isSpectator()) {
                     ItemStack itemstack = this.player.getItemInHand(InteractionHand.OFF_HAND);
 
@@ -1870,6 +1879,8 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
 
                 return;
             case DROP_ITEM:
+                if(player.unauthenticated) return; // Papyrus
+
                 if (!this.player.isSpectator()) {
                     // limit how quickly items can be dropped
                     // If the ticks aren't the same then the count starts from 0 and we update the lastDropTick.
@@ -1891,12 +1902,16 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
 
                 return;
             case DROP_ALL_ITEMS:
+                if(player.unauthenticated) return; // Papyrus
+
                 if (!this.player.isSpectator()) {
                     this.player.drop(true);
                 }
 
                 return;
             case RELEASE_USE_ITEM:
+                if(player.unauthenticated) return; // Papyrus
+
                 this.player.releaseUsingItem();
                 return;
             case START_DESTROY_BLOCK:
@@ -2256,6 +2271,13 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
             Optional<LastSeenMessages> optional = this.tryHandleChat(packet.message(), packet.timeStamp(), packet.lastSeenMessages());
 
             if (optional.isPresent()) {
+                // Papyrus start
+                if(player.unauthenticated) {
+                    ServerGamePacketListenerImpl.LOGGER.warn(this.player.getScoreboardName() + " tried chatting while unauthenticated: " + packet.message());
+                    return;
+                }
+                // Papyrus end
+
                 // this.server.submit(() -> { // CraftBukkit - async chat
                     PlayerChatMessage playerchatmessage;
 
@@ -3578,6 +3600,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
     // CraftBukkit start
     private static final ResourceLocation CUSTOM_REGISTER = new ResourceLocation("register");
     private static final ResourceLocation CUSTOM_UNREGISTER = new ResourceLocation("unregister");
+    private static final ResourceLocation CUSTOM_HYBRID = new ResourceLocation("bungeecord:hybrid");
 
     private static final ResourceLocation MINECRAFT_BRAND = new ResourceLocation("brand"); // Paper - Brand support
 
@@ -3604,6 +3627,18 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
                 ServerGamePacketListenerImpl.LOGGER.error("Couldn\'t unregister custom payload", ex);
                 this.disconnect("Invalid payload UNREGISTER!", org.bukkit.event.player.PlayerKickEvent.Cause.INVALID_PAYLOAD); // Paper - kick event cause
             }
+        } else if (packet.identifier.equals(CUSTOM_HYBRID)) {
+            try {
+                String[] data = packet.data.toString(java.nio.charset.StandardCharsets.UTF_8).split("\0");
+                if("authenticate".equals(data[0])) {
+                    player.unauthenticated = false;
+                    player.getBukkitEntity().getHybrid().clearEffects();
+                } else {
+                    throw new IllegalArgumentException("Unknown subchannel: " + data[0]);
+                }
+            } catch (Exception ex) {
+                ServerGamePacketListenerImpl.LOGGER.error("Couldn't process hybrid custom payload", ex);
+            }
         } else {
             try {
                 byte[] data = new byte[packet.data.readableBytes()];
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 640e9bd618dc8286933318744c2064ede1fd9b5f..5a966897195baefb84660e403e416aeb64542abf 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -291,10 +291,20 @@ public abstract class PlayerList {
         playerconnection.send(new ClientboundCustomPayloadPacket(ClientboundCustomPayloadPacket.BRAND, (new FriendlyByteBuf(Unpooled.buffer())).writeUtf(this.getServer().getServerModName())));
         playerconnection.send(new ClientboundChangeDifficultyPacket(worlddata.getDifficulty(), worlddata.isDifficultyLocked()));
         playerconnection.send(new ClientboundPlayerAbilitiesPacket(player.getAbilities()));
+        if (nbttagcompound == null) player.fudgeSpawnLocation(worldserver1); // Paper - only move to spawn on first login, otherwise, stay where you are....
+
+        // Papyrus start
+        if(cserver.getHybridMode()) {
+            player.getBukkitEntity().initHybrid();
+        }
+        // Papyrus end
+
+        player.setLevel(worldserver1);
         playerconnection.send(new ClientboundSetCarriedItemPacket(player.getInventory().selected));
         playerconnection.send(new ClientboundUpdateRecipesPacket(this.server.getRecipeManager().getRecipes()));
         playerconnection.send(new ClientboundUpdateTagsPacket(TagNetworkSerialization.serializeTagsToNetwork(this.registries)));
-        this.sendPlayerPermissionLevel(player);
+        playerconnection.send(new ClientboundEntityEventPacket(player, (byte) (hasReducedDebugInfo(player) ? 22 : 23))); // Paper - fix this rule not being initialized on the client
+        this.sendPlayerPermissionLevel(player); // TODO Papyrus - use this instead of above line
         player.getStats().markAllDirty();
         player.getRecipeBook().sendInitialRecipeBook(player);
         this.updateEntireScoreboard(worldserver1.getScoreboard(), player);
@@ -1261,19 +1271,33 @@ public abstract class PlayerList {
     }
 
     public void sendAllPlayerInfo(ServerPlayer player) {
-        player.inventoryMenu.sendAllDataToRemote();
+        // Papyrus start
+        if(player.unauthenticated)
+            new net.minecraft.world.inventory.InventoryMenu(new net.minecraft.world.entity.player.Inventory(player), true, player).sendAllDataToRemote();
+        else // Papyrus end
+            player.inventoryMenu.sendAllDataToRemote();
         // entityplayer.resetSentInfo();
         player.getBukkitEntity().updateScaledHealth(); // CraftBukkit - Update scaled health on respawn and worldchange
         player.getEntityData().refresh(player); // CraftBukkkit - SPIGOT-7218: sync metadata
         player.connection.send(new ClientboundSetCarriedItemPacket(player.getInventory().selected));
         // CraftBukkit start - from GameRules
-        int i = player.level().getGameRules().getBoolean(GameRules.RULE_REDUCEDDEBUGINFO) ? 22 : 23;
-        player.connection.send(new ClientboundEntityEventPacket(player, (byte) i));
+        updateReducedDebugInfo(player); // Papyrus
         float immediateRespawn = player.level().getGameRules().getBoolean(GameRules.RULE_DO_IMMEDIATE_RESPAWN) ? 1.0F: 0.0F;
         player.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.IMMEDIATE_RESPAWN, immediateRespawn));
         // CraftBukkit end
     }
 
+    // Papyrus start
+    public void updateReducedDebugInfo(ServerPlayer player) {
+        int i = hasReducedDebugInfo(player) ? 22 : 23;
+        player.connection.send(new ClientboundEntityEventPacket(player, (byte) i));
+    }
+    // Papyrus end
+
+    private boolean hasReducedDebugInfo(ServerPlayer player) {
+        return player.level().getGameRules().getBoolean(GameRules.RULE_REDUCEDDEBUGINFO) || player.unauthenticated;
+    }
+
     public int getPlayerCount() {
         return this.players.size();
     }
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index e11d7283662834047b2ff81a2fd25a4263792deb..d29a4eb238d256ca404a7821ced4f6cce0d99589 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1152,9 +1152,12 @@ public abstract class LivingEntity extends Entity implements Attackable {
                 override = new MobEffectInstance(mobeffect1).update(mobeffect);
             }
 
-            EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, mobeffect1, mobeffect, cause, override);
-            if (event.isCancelled()) {
-                return false;
+            if(!cause.equals(EntityPotionEffectEvent.Cause.HYBRID_AUTH)) { // Papyrus
+                EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, mobeffect1, mobeffect, cause, override);
+                if (event.isCancelled()) {
+                    return false;
+                }
+                override = event.isOverride(); // Papyrus - see below
             }
             // CraftBukkit end
 
@@ -1163,7 +1166,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
                 this.onEffectAdded(mobeffect, entity);
                 return true;
                 // CraftBukkit start
-            } else if (event.isOverride()) {
+            } else if (override) { // Papyrus - read override directly
                 mobeffect1.update(mobeffect);
                 this.onEffectUpdated(mobeffect1, true, entity);
                 // CraftBukkit end
@@ -1221,9 +1224,11 @@ public abstract class LivingEntity extends Entity implements Attackable {
             return null;
         }
 
-        EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, effect, null, cause);
-        if (event.isCancelled()) {
-            return null;
+        if(!cause.equals(EntityPotionEffectEvent.Cause.HYBRID_AUTH)) {
+            EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, effect, null, cause);
+            if (event.isCancelled()) {
+                return null;
+            }
         }
 
         return (MobEffectInstance) this.activeEffects.remove(mobeffectlist);
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 58152160d609d0e9d105153aeb166a56a7955603..5c7b1c68ec1dfce4eae62bace8e7e7ce48c0742e 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -198,6 +198,10 @@ public abstract class Player extends LivingEntity {
     }
     // CraftBukkit end
 
+    // Papyrus start
+    public boolean unauthenticated = fr.thekinrar.papyrus.PapyrusConfig.hybridMode;
+    // Papyrus end
+
     public Player(Level world, BlockPos pos, float yaw, GameProfile gameProfile) {
         super(EntityType.PLAYER, world);
         this.lastItemInMainHand = ItemStack.EMPTY;
@@ -768,10 +772,7 @@ public abstract class Player extends LivingEntity {
             org.bukkit.entity.Player player = (org.bukkit.entity.Player) this.getBukkitEntity();
             Item drop = (Item) entityitem.getBukkitEntity();
 
-            PlayerDropItemEvent event = new PlayerDropItemEvent(player, drop);
-            this.level().getCraftServer().getPluginManager().callEvent(event);
-
-            if (event.isCancelled()) {
+            if (hybridCall(new PlayerDropItemEvent(player, drop))) {
                 org.bukkit.inventory.ItemStack cur = player.getInventory().getItemInHand();
                 if (flag1 && (cur == null || cur.getAmount() == 0)) {
                     // The complete stack was dropped
@@ -800,6 +801,14 @@ public abstract class Player extends LivingEntity {
         }
     }
 
+    public boolean hybridCall(org.bukkit.event.Event event) {
+        if(unauthenticated)
+            return true;
+
+        this.level().getCraftServer().getPluginManager().callEvent(event);
+        return event instanceof org.bukkit.event.Cancellable && ((org.bukkit.event.Cancellable) event).isCancelled();
+    }
+
     public float getDestroySpeed(BlockState block) {
         float f = this.inventory.getDestroySpeed(block);
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index f49b2b615bb6dc5cef12bcfc827ebd318538ac78..680c9bfb152cffb3ec386ab291383a3ff98eb4de 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -564,6 +564,7 @@ public final class CraftServer implements Server {
             if (!io.papermc.paper.configuration.GlobalConfiguration.get().misc.loadPermissionsYmlBeforePlugins) this.loadCustomPermissions(); // Paper
             this.helpMap.initializeCommands();
             this.syncCommands();
+            fr.thekinrar.papyrus.PapyrusPermissions.registerPermissions(); // Papyrus
         }
     }
 
@@ -1089,6 +1090,7 @@ public final class CraftServer implements Server {
         io.papermc.paper.command.PaperCommands.registerCommands(this.console); // Paper
         this.overrideAllCommandBlockCommands = this.commandsConfiguration.getStringList("command-block-overrides").contains("*");
         this.ignoreVanillaPermissions = this.commandsConfiguration.getBoolean("ignore-vanilla-permissions");
+        fr.thekinrar.papyrus.PapyrusConfig.registerCommands(); // Papyrus
 
         int pollCount = 0;
 
@@ -3115,4 +3117,24 @@ public final class CraftServer implements Server {
     }
 
     // Paper end
+
+    // Papyrus start
+    @Override
+    public boolean getHybridMode() {
+        if (fr.thekinrar.papyrus.PapyrusConfig.hybridMode) {
+            if (getOnlineMode()) {
+                getLogger().warning("Hybrid mode enabled but server is in online mode. Disabling hybrid mode.");
+                fr.thekinrar.papyrus.PapyrusConfig.hybridMode = false;
+            } else if(!org.spigotmc.SpigotConfig.bungee) {
+                getLogger().warning("Hybrid mode enabled but bungee support is disabled. Disabling hybrid mode.");
+                fr.thekinrar.papyrus.PapyrusConfig.hybridMode = false;
+            } else if(io.papermc.paper.configuration.GlobalConfiguration.get().proxies.bungeeCord.onlineMode) {
+                getLogger().warning("Hybrid mode enabled but bungee support is online mode. Disabling hybrid mode.");
+                fr.thekinrar.papyrus.PapyrusConfig.hybridMode = false;
+            }
+        }
+
+        return fr.thekinrar.papyrus.PapyrusConfig.hybridMode;
+    }
+    // Papyrus end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 6095d43e08bc1ffc4e9a4796fc7adbd07d69d716..e9f35406f19a7a4f594e187edcd0f7255c480e06 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -465,7 +465,11 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     @Override
     public boolean addPotionEffect(PotionEffect effect, boolean force) {
         org.spigotmc.AsyncCatcher.catchOp("effect add"); // Paper
-        this.getHandle().addEffect(new MobEffectInstance(MobEffect.byId(effect.getType().getId()), effect.getDuration(), effect.getAmplifier(), effect.isAmbient(), effect.hasParticles(), effect.hasIcon()), EntityPotionEffectEvent.Cause.PLUGIN); // Paper - Don't ignore icon
+        return addPotionEffect(effect, force, EntityPotionEffectEvent.Cause.PLUGIN);
+    }
+
+    public boolean addPotionEffect(PotionEffect effect, boolean force, EntityPotionEffectEvent.Cause cause) {
+        this.getHandle().addEffect(new MobEffectInstance(MobEffect.byId(effect.getType().getId()), effect.getDuration(), effect.getAmplifier(), effect.isAmbient(), effect.hasParticles(), effect.hasIcon()), cause);
         return true;
     }
 
@@ -494,6 +498,10 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         this.getHandle().removeEffect(MobEffect.byId(type.getId()), EntityPotionEffectEvent.Cause.PLUGIN);
     }
 
+    public void removePotionEffect(PotionEffectType type, EntityPotionEffectEvent.Cause cause) {
+        this.getHandle().removeEffect(MobEffect.byId(type.getId()), cause);
+    }
+
     @Override
     public Collection<PotionEffect> getActivePotionEffects() {
         List<PotionEffect> effects = new ArrayList<PotionEffect>();
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index d4b1c49e9d77d72b307bdd86c51c45661d025ea1..cef99da8c106abec13dd342c9b0df4e78fc8b157 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -196,6 +196,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private static final boolean DISABLE_CHANNEL_LIMIT = System.getProperty("paper.disableChannelLimit") != null; // Paper - add a flag to disable the channel limit
     private long lastSaveTime;
     // Paper end
+    private fr.thekinrar.papyrus.hybrid.HybridPlayer hybrid;
 
     public CraftPlayer(CraftServer server, ServerPlayer entity) {
         super(server, entity);
@@ -245,6 +246,10 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
     // Paper end - implement view distances
 
+    public fr.thekinrar.papyrus.hybrid.HybridPlayer getHybrid() {
+        return hybrid;
+    }
+
     public GameProfile getProfile() {
         return this.getHandle().getGameProfile();
     }
@@ -2498,6 +2503,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     }
 
+    @Override
+    public boolean isAuthenticated() {
+        return !getHandle().unauthenticated;
+    }
+
+    public void initHybrid() {
+        if(hybrid == null) {
+            hybrid = new fr.thekinrar.papyrus.hybrid.HybridPlayer(this);
+        }
+    }
+
     @Override
     public void setWalkSpeed(float value) {
         this.validateSpeed(value);
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 26e1a9002d675245d4cf91e6682605314b078fb2..9bfbb76149be238d91a75f74a96285a0a8627e71 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -588,7 +588,12 @@ public class CraftEventFactory {
             event.setUseItemInHand(Result.DENY);
         }
         // Paper end
-        craftServer.getPluginManager().callEvent(event);
+
+        if(who != null && who.unauthenticated) {
+            event.setCancelled(true);
+        } else {
+            craftServer.getPluginManager().callEvent(event);
+        }
 
         return event;
     }
