From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Fri, 25 Jun 2021 21:37:27 +0200
Subject: [PATCH] Implement disableVanillaEntityCollisions API


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 9bb44918af119d9afae4a0a050c6a5381f028364..b067353b3d5c3c02995a24c940e7a61cd2a40017 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -326,6 +326,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
     private org.bukkit.util.Vector origin;
     @javax.annotation.Nullable
     private UUID originWorld;
+    public boolean disableVanillaEntityCollisions; // Papyrus
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -1253,6 +1254,8 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
             return movement;
         }
 
+        if(disableVanillaEntityCollisions) return moveVector; // Papyrus
+
         final Level world = this.level;
         final AABB currBoundingBox = this.getBoundingBox();
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 5aae88e20bc04560d6ad52cfcaa872d28bfcee8f..c8e0b4c30e6e93b8dd9bdd14478505d0b1a2f727 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1287,4 +1287,16 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return !entity.valid && entity.level.addFreshEntity(entity, reason);
     }
     // Paper end
+
+    // Papyrus start
+    @Override
+    public void setIgnoresVanillaEntityCollisions(boolean ignores) {
+        getHandle().disableVanillaEntityCollisions = ignores;
+    }
+
+    @Override
+    public boolean ignoresVanillaEntityCollisions() {
+        return getHandle().disableVanillaEntityCollisions;
+    }
+    // Papyrus end
 }
