From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Tue, 1 Oct 2019 10:12:37 +0100
Subject: [PATCH] Support Material not being an enum anymore


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 921253a06daa414aed7dc6824effc65db09ea7a5..05b26c448c1053ce9ab23f2503868041033b759c 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -560,7 +560,7 @@ public class PaperWorldConfig {
 
         altItemDespawnRateEnabled = getBoolean(path + ".enabled", false);
 
-        Map<Material, Integer> altItemDespawnRateMapDefault = new EnumMap<>(Material.class);
+        Map<Material, Integer> altItemDespawnRateMapDefault = new HashMap<>(Material.values().length);
         altItemDespawnRateMapDefault.put(Material.COBBLESTONE, 300);
         for (Material key : altItemDespawnRateMapDefault.keySet()) {
             config.addDefault("world-settings.default." + path + ".items." + key, altItemDespawnRateMapDefault.get(key));
@@ -582,7 +582,7 @@ public class PaperWorldConfig {
             altItemDespawnRateEnabled = false;
         }
 
-        altItemDespawnRateMap = new EnumMap<>(Material.class);
+        altItemDespawnRateMap = new HashMap<>(Material.values().length);
         if (!altItemDespawnRateEnabled) {
             return;
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBed.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBed.java
index 418e525e6193680f203840309e8c5b90c2f49d7a..2a23c8903717896d54c822d767f233d50a20d3e9 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBed.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBed.java
@@ -18,38 +18,38 @@ public class CraftBed extends CraftBlockEntityState<TileEntityBed> implements Be
 
     @Override
     public DyeColor getColor() {
-        switch (getType()) {
-            case BLACK_BED:
+        switch (getType().name()) {
+            case "BLACK_BED":
                 return DyeColor.BLACK;
-            case BLUE_BED:
+            case "BLUE_BED":
                 return DyeColor.BLUE;
-            case BROWN_BED:
+            case "BROWN_BED":
                 return DyeColor.BROWN;
-            case CYAN_BED:
+            case "CYAN_BED":
                 return DyeColor.CYAN;
-            case GRAY_BED:
+            case "GRAY_BED":
                 return DyeColor.GRAY;
-            case GREEN_BED:
+            case "GREEN_BED":
                 return DyeColor.GREEN;
-            case LIGHT_BLUE_BED:
+            case "LIGHT_BLUE_BED":
                 return DyeColor.LIGHT_BLUE;
-            case LIGHT_GRAY_BED:
+            case "LIGHT_GRAY_BED":
                 return DyeColor.LIGHT_GRAY;
-            case LIME_BED:
+            case "LIME_BED":
                 return DyeColor.LIME;
-            case MAGENTA_BED:
+            case "MAGENTA_BED":
                 return DyeColor.MAGENTA;
-            case ORANGE_BED:
+            case "ORANGE_BED":
                 return DyeColor.ORANGE;
-            case PINK_BED:
+            case "PINK_BED":
                 return DyeColor.PINK;
-            case PURPLE_BED:
+            case "PURPLE_BED":
                 return DyeColor.PURPLE;
-            case RED_BED:
+            case "RED_BED":
                 return DyeColor.RED;
-            case WHITE_BED:
+            case "WHITE_BED":
                 return DyeColor.WHITE;
-            case YELLOW_BED:
+            case "YELLOW_BED":
                 return DyeColor.YELLOW;
             default:
                 throw new IllegalArgumentException("Unknown DyeColor for " + getType());
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index 4b6cfe09c256b43ca59f9c8a52330004c198438b..f30397ac979145ff452df2593901fe0afc33de35 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -332,159 +332,159 @@ public class CraftBlock implements Block {
         // Paper end
         Material material = getType();
 
-        switch (material) {
-            case ACACIA_SIGN:
-            case ACACIA_WALL_SIGN:
-            case BIRCH_SIGN:
-            case BIRCH_WALL_SIGN:
-            case CRIMSON_SIGN:
-            case CRIMSON_WALL_SIGN:
-            case DARK_OAK_SIGN:
-            case DARK_OAK_WALL_SIGN:
-            case JUNGLE_SIGN:
-            case JUNGLE_WALL_SIGN:
-            case OAK_SIGN:
-            case OAK_WALL_SIGN:
-            case SPRUCE_SIGN:
-            case SPRUCE_WALL_SIGN:
-            case WARPED_SIGN:
-            case WARPED_WALL_SIGN:
+        switch (material.name()) {
+        case "ACACIA_SIGN":
+        case "ACACIA_WALL_SIGN":
+        case "BIRCH_SIGN":
+        case "BIRCH_WALL_SIGN":
+        case "CRIMSON_SIGN":
+        case "CRIMSON_WALL_SIGN":
+        case "DARK_OAK_SIGN":
+        case "DARK_OAK_WALL_SIGN":
+        case "JUNGLE_SIGN":
+        case "JUNGLE_WALL_SIGN":
+        case "OAK_SIGN":
+        case "OAK_WALL_SIGN":
+        case "SPRUCE_SIGN":
+        case "SPRUCE_WALL_SIGN":
+        case "WARPED_SIGN":
+        case "WARPED_WALL_SIGN":
                 return new CraftSign(this);
-            case CHEST:
-            case TRAPPED_CHEST:
+        case "CHEST":
+        case "TRAPPED_CHEST":
                 return new CraftChest(this);
-            case FURNACE:
+        case "FURNACE":
                 return new CraftFurnaceFurnace(this);
-            case DISPENSER:
+        case "DISPENSER":
                 return new CraftDispenser(this);
-            case DROPPER:
+        case "DROPPER":
                 return new CraftDropper(this);
-            case END_GATEWAY:
+        case "END_GATEWAY":
                 return new CraftEndGateway(this);
-            case HOPPER:
+        case "HOPPER":
                 return new CraftHopper(this);
-            case SPAWNER:
+        case "SPAWNER":
                 return new CraftCreatureSpawner(this);
-            case JUKEBOX:
+        case "JUKEBOX":
                 return new CraftJukebox(this);
-            case BREWING_STAND:
+        case "BREWING_STAND":
                 return new CraftBrewingStand(this);
-            case CREEPER_HEAD:
-            case CREEPER_WALL_HEAD:
-            case DRAGON_HEAD:
-            case DRAGON_WALL_HEAD:
-            case PLAYER_HEAD:
-            case PLAYER_WALL_HEAD:
-            case SKELETON_SKULL:
-            case SKELETON_WALL_SKULL:
-            case WITHER_SKELETON_SKULL:
-            case WITHER_SKELETON_WALL_SKULL:
-            case ZOMBIE_HEAD:
-            case ZOMBIE_WALL_HEAD:
+        case "CREEPER_HEAD":
+        case "CREEPER_WALL_HEAD":
+        case "DRAGON_HEAD":
+        case "DRAGON_WALL_HEAD":
+        case "PLAYER_HEAD":
+        case "PLAYER_WALL_HEAD":
+        case "SKELETON_SKULL":
+        case "SKELETON_WALL_SKULL":
+        case "WITHER_SKELETON_SKULL":
+        case "WITHER_SKELETON_WALL_SKULL":
+        case "ZOMBIE_HEAD":
+        case "ZOMBIE_WALL_HEAD":
                 return new CraftSkull(this);
-            case COMMAND_BLOCK:
-            case CHAIN_COMMAND_BLOCK:
-            case REPEATING_COMMAND_BLOCK:
+        case "COMMAND_BLOCK":
+        case "CHAIN_COMMAND_BLOCK":
+        case "REPEATING_COMMAND_BLOCK":
                 return new CraftCommandBlock(this);
-            case BEACON:
+        case "BEACON":
                 return new CraftBeacon(this);
-            case BLACK_BANNER:
-            case BLACK_WALL_BANNER:
-            case BLUE_BANNER:
-            case BLUE_WALL_BANNER:
-            case BROWN_BANNER:
-            case BROWN_WALL_BANNER:
-            case CYAN_BANNER:
-            case CYAN_WALL_BANNER:
-            case GRAY_BANNER:
-            case GRAY_WALL_BANNER:
-            case GREEN_BANNER:
-            case GREEN_WALL_BANNER:
-            case LIGHT_BLUE_BANNER:
-            case LIGHT_BLUE_WALL_BANNER:
-            case LIGHT_GRAY_BANNER:
-            case LIGHT_GRAY_WALL_BANNER:
-            case LIME_BANNER:
-            case LIME_WALL_BANNER:
-            case MAGENTA_BANNER:
-            case MAGENTA_WALL_BANNER:
-            case ORANGE_BANNER:
-            case ORANGE_WALL_BANNER:
-            case PINK_BANNER:
-            case PINK_WALL_BANNER:
-            case PURPLE_BANNER:
-            case PURPLE_WALL_BANNER:
-            case RED_BANNER:
-            case RED_WALL_BANNER:
-            case WHITE_BANNER:
-            case WHITE_WALL_BANNER:
-            case YELLOW_BANNER:
-            case YELLOW_WALL_BANNER:
+        case "BLACK_BANNER":
+        case "BLACK_WALL_BANNER":
+        case "BLUE_BANNER":
+        case "BLUE_WALL_BANNER":
+        case "BROWN_BANNER":
+        case "BROWN_WALL_BANNER":
+        case "CYAN_BANNER":
+        case "CYAN_WALL_BANNER":
+        case "GRAY_BANNER":
+        case "GRAY_WALL_BANNER":
+        case "GREEN_BANNER":
+        case "GREEN_WALL_BANNER":
+        case "LIGHT_BLUE_BANNER":
+        case "LIGHT_BLUE_WALL_BANNER":
+        case "LIGHT_GRAY_BANNER":
+        case "LIGHT_GRAY_WALL_BANNER":
+        case "LIME_BANNER":
+        case "LIME_WALL_BANNER":
+        case "MAGENTA_BANNER":
+        case "MAGENTA_WALL_BANNER":
+        case "ORANGE_BANNER":
+        case "ORANGE_WALL_BANNER":
+        case "PINK_BANNER":
+        case "PINK_WALL_BANNER":
+        case "PURPLE_BANNER":
+        case "PURPLE_WALL_BANNER":
+        case "RED_BANNER":
+        case "RED_WALL_BANNER":
+        case "WHITE_BANNER":
+        case "WHITE_WALL_BANNER":
+        case "YELLOW_BANNER":
+        case "YELLOW_WALL_BANNER":
                 return new CraftBanner(this);
-            case STRUCTURE_BLOCK:
+        case "STRUCTURE_BLOCK":
                 return new CraftStructureBlock(this);
-            case SHULKER_BOX:
-            case WHITE_SHULKER_BOX:
-            case ORANGE_SHULKER_BOX:
-            case MAGENTA_SHULKER_BOX:
-            case LIGHT_BLUE_SHULKER_BOX:
-            case YELLOW_SHULKER_BOX:
-            case LIME_SHULKER_BOX:
-            case PINK_SHULKER_BOX:
-            case GRAY_SHULKER_BOX:
-            case LIGHT_GRAY_SHULKER_BOX:
-            case CYAN_SHULKER_BOX:
-            case PURPLE_SHULKER_BOX:
-            case BLUE_SHULKER_BOX:
-            case BROWN_SHULKER_BOX:
-            case GREEN_SHULKER_BOX:
-            case RED_SHULKER_BOX:
-            case BLACK_SHULKER_BOX:
+        case "SHULKER_BOX":
+        case "WHITE_SHULKER_BOX":
+        case "ORANGE_SHULKER_BOX":
+        case "MAGENTA_SHULKER_BOX":
+        case "LIGHT_BLUE_SHULKER_BOX":
+        case "YELLOW_SHULKER_BOX":
+        case "LIME_SHULKER_BOX":
+        case "PINK_SHULKER_BOX":
+        case "GRAY_SHULKER_BOX":
+        case "LIGHT_GRAY_SHULKER_BOX":
+        case "CYAN_SHULKER_BOX":
+        case "PURPLE_SHULKER_BOX":
+        case "BLUE_SHULKER_BOX":
+        case "BROWN_SHULKER_BOX":
+        case "GREEN_SHULKER_BOX":
+        case "RED_SHULKER_BOX":
+        case "BLACK_SHULKER_BOX":
                 return new CraftShulkerBox(this);
-            case ENCHANTING_TABLE:
+        case "ENCHANTING_TABLE":
                 return new CraftEnchantingTable(this);
-            case ENDER_CHEST:
+        case "ENDER_CHEST":
                 return new CraftEnderChest(this);
-            case DAYLIGHT_DETECTOR:
+        case "DAYLIGHT_DETECTOR":
                 return new CraftDaylightDetector(this);
-            case COMPARATOR:
+        case "COMPARATOR":
                 return new CraftComparator(this);
-            case BLACK_BED:
-            case BLUE_BED:
-            case BROWN_BED:
-            case CYAN_BED:
-            case GRAY_BED:
-            case GREEN_BED:
-            case LIGHT_BLUE_BED:
-            case LIGHT_GRAY_BED:
-            case LIME_BED:
-            case MAGENTA_BED:
-            case ORANGE_BED:
-            case PINK_BED:
-            case PURPLE_BED:
-            case RED_BED:
-            case WHITE_BED:
-            case YELLOW_BED:
+        case "BLACK_BED":
+        case "BLUE_BED":
+        case "BROWN_BED":
+        case "CYAN_BED":
+        case "GRAY_BED":
+        case "GREEN_BED":
+        case "LIGHT_BLUE_BED":
+        case "LIGHT_GRAY_BED":
+        case "LIME_BED":
+        case "MAGENTA_BED":
+        case "ORANGE_BED":
+        case "PINK_BED":
+        case "PURPLE_BED":
+        case "RED_BED":
+        case "WHITE_BED":
+        case "YELLOW_BED":
                 return new CraftBed(this);
-            case CONDUIT:
+        case "CONDUIT":
                 return new CraftConduit(this);
-            case BARREL:
+        case "BARREL":
                 return new CraftBarrel(this);
-            case BELL:
+        case "BELL":
                 return new CraftBell(this);
-            case BLAST_FURNACE:
+        case "BLAST_FURNACE":
                 return new CraftBlastFurnace(this);
-            case CAMPFIRE:
-            case SOUL_CAMPFIRE:
+        case "CAMPFIRE":
+        case "SOUL_CAMPFIRE":
                 return new CraftCampfire(this);
-            case JIGSAW:
+        case "JIGSAW":
                 return new CraftJigsaw(this);
-            case LECTERN:
+        case "LECTERN":
                 return new CraftLectern(this);
-            case SMOKER:
+        case "SMOKER":
                 return new CraftSmoker(this);
-            case BEEHIVE:
-            case BEE_NEST:
+        case "BEEHIVE":
+        case "BEE_NEST":
                 return new CraftBeehive(this);
             default:
                 TileEntity tileEntity = world.getTileEntity(position);
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftSkull.java b/src/main/java/org/bukkit/craftbukkit/block/CraftSkull.java
index 00fb749f4d181d8d830496cf741d589e10af5098..0ee6f013f7326d604effd04d4b0acff6334579e5 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftSkull.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftSkull.java
@@ -141,24 +141,24 @@ public class CraftSkull extends CraftBlockEntityState<TileEntitySkull> implement
 
     @Override
     public SkullType getSkullType() {
-        switch (getType()) {
-            case SKELETON_SKULL:
-            case SKELETON_WALL_SKULL:
+        switch (getType().name()) {
+            case "SKELETON_SKULL":
+            case "SKELETON_WALL_SKULL":
                 return SkullType.SKELETON;
-            case WITHER_SKELETON_SKULL:
-            case WITHER_SKELETON_WALL_SKULL:
+            case "WITHER_SKELETON_SKULL":
+            case "WITHER_SKELETON_WALL_SKULL":
                 return SkullType.WITHER;
-            case ZOMBIE_HEAD:
-            case ZOMBIE_WALL_HEAD:
+            case "ZOMBIE_HEAD":
+            case "ZOMBIE_WALL_HEAD":
                 return SkullType.ZOMBIE;
-            case PLAYER_HEAD:
-            case PLAYER_WALL_HEAD:
+            case "PLAYER_HEAD":
+            case "PLAYER_WALL_HEAD":
                 return SkullType.PLAYER;
-            case CREEPER_HEAD:
-            case CREEPER_WALL_HEAD:
+            case "CREEPER_HEAD":
+            case "CREEPER_WALL_HEAD":
                 return SkullType.CREEPER;
-            case DRAGON_HEAD:
-            case DRAGON_WALL_HEAD:
+            case "DRAGON_HEAD":
+            case "DRAGON_WALL_HEAD":
                 return SkullType.DRAGON;
             default:
                 throw new IllegalArgumentException("Unknown SkullType for " + getType());
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 926440e846eff2c1aaa262aa2b3975b7dd225332..f68d48f4c5b6d35f7bbf0def19b6e5b804a0c65d 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1367,14 +1367,14 @@ public class CraftEventFactory {
         org.bukkit.World bukkitWorld = world.getWorld();
         Block igniter = bukkitWorld.getBlockAt(source.getX(), source.getY(), source.getZ());
         IgniteCause cause;
-        switch (igniter.getType()) {
-            case LAVA:
+        switch (igniter.getType().name()) {
+            case "LAVA":
                 cause = IgniteCause.LAVA;
                 break;
-            case DISPENSER:
+            case "DISPENSER":
                 cause = IgniteCause.FLINT_AND_STEEL;
                 break;
-            case FIRE: // Fire or any other unknown block counts as SPREAD.
+            case "FIRE": // Fire or any other unknown block counts as SPREAD.
             default:
                 cause = IgniteCause.SPREAD;
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
index 6966b9d1ce674232d3f867798fa58bd0933ff69e..6258163d1a252890c115fc2dd1ff49933c807189 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
@@ -50,222 +50,222 @@ public final class CraftItemFactory implements ItemFactory {
 
     private ItemMeta getItemMeta(Material material, CraftMetaItem meta) {
         material = CraftLegacy.fromLegacy(material); // This may be called from legacy item stacks, try to get the right material
-        switch (material) {
-        case AIR:
+        switch (material.name()) {
+        case "AIR":
             return null;
-        case WRITTEN_BOOK:
+        case "WRITTEN_BOOK":
             return meta instanceof CraftMetaBookSigned ? meta : new CraftMetaBookSigned(meta);
-        case WRITABLE_BOOK:
+        case "WRITABLE_BOOK":
             return meta != null && meta.getClass().equals(CraftMetaBook.class) ? meta : new CraftMetaBook(meta);
-        case CREEPER_HEAD:
-        case CREEPER_WALL_HEAD:
-        case DRAGON_HEAD:
-        case DRAGON_WALL_HEAD:
-        case PLAYER_HEAD:
-        case PLAYER_WALL_HEAD:
-        case SKELETON_SKULL:
-        case SKELETON_WALL_SKULL:
-        case WITHER_SKELETON_SKULL:
-        case WITHER_SKELETON_WALL_SKULL:
-        case ZOMBIE_HEAD:
-        case ZOMBIE_WALL_HEAD:
+        case "CREEPER_HEAD":
+        case "CREEPER_WALL_HEAD":
+        case "DRAGON_HEAD":
+        case "DRAGON_WALL_HEAD":
+        case "PLAYER_HEAD":
+        case "PLAYER_WALL_HEAD":
+        case "SKELETON_SKULL":
+        case "SKELETON_WALL_SKULL":
+        case "WITHER_SKELETON_SKULL":
+        case "WITHER_SKELETON_WALL_SKULL":
+        case "ZOMBIE_HEAD":
+        case "ZOMBIE_WALL_HEAD":
             return meta instanceof CraftMetaSkull ? meta : new CraftMetaSkull(meta);
-        case LEATHER_HELMET:
-        case LEATHER_HORSE_ARMOR:
-        case LEATHER_CHESTPLATE:
-        case LEATHER_LEGGINGS:
-        case LEATHER_BOOTS:
+        case "LEATHER_HELMET":
+        case "LEATHER_HORSE_ARMOR":
+        case "LEATHER_CHESTPLATE":
+        case "LEATHER_LEGGINGS":
+        case "LEATHER_BOOTS":
             return meta instanceof CraftMetaLeatherArmor ? meta : new CraftMetaLeatherArmor(meta);
-        case POTION:
-        case SPLASH_POTION:
-        case LINGERING_POTION:
-        case TIPPED_ARROW:
+        case "POTION":
+        case "SPLASH_POTION":
+        case "LINGERING_POTION":
+        case "TIPPED_ARROW":
             return meta instanceof CraftMetaPotion ? meta : new CraftMetaPotion(meta);
-        case FILLED_MAP:
+        case "FILLED_MAP":
             return meta instanceof CraftMetaMap ? meta : new CraftMetaMap(meta);
-        case FIREWORK_ROCKET:
+        case "FIREWORK_ROCKET":
             return meta instanceof CraftMetaFirework ? meta : new CraftMetaFirework(meta);
-        case FIREWORK_STAR:
+        case "FIREWORK_STAR":
             return meta instanceof CraftMetaCharge ? meta : new CraftMetaCharge(meta);
-        case ENCHANTED_BOOK:
+        case "ENCHANTED_BOOK":
             return meta instanceof CraftMetaEnchantedBook ? meta : new CraftMetaEnchantedBook(meta);
-        case BLACK_BANNER:
-        case BLACK_WALL_BANNER:
-        case BLUE_BANNER:
-        case BLUE_WALL_BANNER:
-        case BROWN_BANNER:
-        case BROWN_WALL_BANNER:
-        case CYAN_BANNER:
-        case CYAN_WALL_BANNER:
-        case GRAY_BANNER:
-        case GRAY_WALL_BANNER:
-        case GREEN_BANNER:
-        case GREEN_WALL_BANNER:
-        case LIGHT_BLUE_BANNER:
-        case LIGHT_BLUE_WALL_BANNER:
-        case LIGHT_GRAY_BANNER:
-        case LIGHT_GRAY_WALL_BANNER:
-        case LIME_BANNER:
-        case LIME_WALL_BANNER:
-        case MAGENTA_BANNER:
-        case MAGENTA_WALL_BANNER:
-        case ORANGE_BANNER:
-        case ORANGE_WALL_BANNER:
-        case PINK_BANNER:
-        case PINK_WALL_BANNER:
-        case PURPLE_BANNER:
-        case PURPLE_WALL_BANNER:
-        case RED_BANNER:
-        case RED_WALL_BANNER:
-        case WHITE_BANNER:
-        case WHITE_WALL_BANNER:
-        case YELLOW_BANNER:
-        case YELLOW_WALL_BANNER:
+        case "BLACK_BANNER":
+        case "BLACK_WALL_BANNER":
+        case "BLUE_BANNER":
+        case "BLUE_WALL_BANNER":
+        case "BROWN_BANNER":
+        case "BROWN_WALL_BANNER":
+        case "CYAN_BANNER":
+        case "CYAN_WALL_BANNER":
+        case "GRAY_BANNER":
+        case "GRAY_WALL_BANNER":
+        case "GREEN_BANNER":
+        case "GREEN_WALL_BANNER":
+        case "LIGHT_BLUE_BANNER":
+        case "LIGHT_BLUE_WALL_BANNER":
+        case "LIGHT_GRAY_BANNER":
+        case "LIGHT_GRAY_WALL_BANNER":
+        case "LIME_BANNER":
+        case "LIME_WALL_BANNER":
+        case "MAGENTA_BANNER":
+        case "MAGENTA_WALL_BANNER":
+        case "ORANGE_BANNER":
+        case "ORANGE_WALL_BANNER":
+        case "PINK_BANNER":
+        case "PINK_WALL_BANNER":
+        case "PURPLE_BANNER":
+        case "PURPLE_WALL_BANNER":
+        case "RED_BANNER":
+        case "RED_WALL_BANNER":
+        case "WHITE_BANNER":
+        case "WHITE_WALL_BANNER":
+        case "YELLOW_BANNER":
+        case "YELLOW_WALL_BANNER":
             return meta instanceof CraftMetaBanner ? meta : new CraftMetaBanner(meta);
-        case BAT_SPAWN_EGG:
-        case BEE_SPAWN_EGG:
-        case BLAZE_SPAWN_EGG:
-        case CAT_SPAWN_EGG:
-        case CAVE_SPIDER_SPAWN_EGG:
-        case CHICKEN_SPAWN_EGG:
-        case COD_SPAWN_EGG:
-        case COW_SPAWN_EGG:
-        case CREEPER_SPAWN_EGG:
-        case DOLPHIN_SPAWN_EGG:
-        case DONKEY_SPAWN_EGG:
-        case DROWNED_SPAWN_EGG:
-        case ELDER_GUARDIAN_SPAWN_EGG:
-        case ENDERMAN_SPAWN_EGG:
-        case ENDERMITE_SPAWN_EGG:
-        case EVOKER_SPAWN_EGG:
-        case FOX_SPAWN_EGG:
-        case GHAST_SPAWN_EGG:
-        case GUARDIAN_SPAWN_EGG:
-        case HOGLIN_SPAWN_EGG:
-        case HORSE_SPAWN_EGG:
-        case HUSK_SPAWN_EGG:
-        case LLAMA_SPAWN_EGG:
-        case MAGMA_CUBE_SPAWN_EGG:
-        case MOOSHROOM_SPAWN_EGG:
-        case MULE_SPAWN_EGG:
-        case OCELOT_SPAWN_EGG:
-        case PANDA_SPAWN_EGG:
-        case PARROT_SPAWN_EGG:
-        case PHANTOM_SPAWN_EGG:
-        case PIGLIN_SPAWN_EGG:
-        case PIG_SPAWN_EGG:
-        case PILLAGER_SPAWN_EGG:
-        case POLAR_BEAR_SPAWN_EGG:
-        case PUFFERFISH_SPAWN_EGG:
-        case RABBIT_SPAWN_EGG:
-        case RAVAGER_SPAWN_EGG:
-        case SALMON_SPAWN_EGG:
-        case SHEEP_SPAWN_EGG:
-        case SHULKER_SPAWN_EGG:
-        case SILVERFISH_SPAWN_EGG:
-        case SKELETON_HORSE_SPAWN_EGG:
-        case SKELETON_SPAWN_EGG:
-        case SLIME_SPAWN_EGG:
-        case SPIDER_SPAWN_EGG:
-        case SQUID_SPAWN_EGG:
-        case STRAY_SPAWN_EGG:
-        case STRIDER_SPAWN_EGG:
-        case TRADER_LLAMA_SPAWN_EGG:
-        case TROPICAL_FISH_SPAWN_EGG:
-        case TURTLE_SPAWN_EGG:
-        case VEX_SPAWN_EGG:
-        case VILLAGER_SPAWN_EGG:
-        case VINDICATOR_SPAWN_EGG:
-        case WANDERING_TRADER_SPAWN_EGG:
-        case WITCH_SPAWN_EGG:
-        case WITHER_SKELETON_SPAWN_EGG:
-        case WOLF_SPAWN_EGG:
-        case ZOGLIN_SPAWN_EGG:
-        case ZOMBIE_HORSE_SPAWN_EGG:
-        case ZOMBIE_SPAWN_EGG:
-        case ZOMBIE_VILLAGER_SPAWN_EGG:
-        case ZOMBIFIED_PIGLIN_SPAWN_EGG:
+        case "BAT_SPAWN_EGG":
+        case "BEE_SPAWN_EGG":
+        case "BLAZE_SPAWN_EGG":
+        case "CAT_SPAWN_EGG":
+        case "CAVE_SPIDER_SPAWN_EGG":
+        case "CHICKEN_SPAWN_EGG":
+        case "COD_SPAWN_EGG":
+        case "COW_SPAWN_EGG":
+        case "CREEPER_SPAWN_EGG":
+        case "DOLPHIN_SPAWN_EGG":
+        case "DONKEY_SPAWN_EGG":
+        case "DROWNED_SPAWN_EGG":
+        case "ELDER_GUARDIAN_SPAWN_EGG":
+        case "ENDERMAN_SPAWN_EGG":
+        case "ENDERMITE_SPAWN_EGG":
+        case "EVOKER_SPAWN_EGG":
+        case "FOX_SPAWN_EGG":
+        case "GHAST_SPAWN_EGG":
+        case "GUARDIAN_SPAWN_EGG":
+        case "HOGLIN_SPAWN_EGG":
+        case "HORSE_SPAWN_EGG":
+        case "HUSK_SPAWN_EGG":
+        case "LLAMA_SPAWN_EGG":
+        case "MAGMA_CUBE_SPAWN_EGG":
+        case "MOOSHROOM_SPAWN_EGG":
+        case "MULE_SPAWN_EGG":
+        case "OCELOT_SPAWN_EGG":
+        case "PANDA_SPAWN_EGG":
+        case "PARROT_SPAWN_EGG":
+        case "PHANTOM_SPAWN_EGG":
+        case "PIGLIN_SPAWN_EGG":
+        case "PIG_SPAWN_EGG":
+        case "PILLAGER_SPAWN_EGG":
+        case "POLAR_BEAR_SPAWN_EGG":
+        case "PUFFERFISH_SPAWN_EGG":
+        case "RABBIT_SPAWN_EGG":
+        case "RAVAGER_SPAWN_EGG":
+        case "SALMON_SPAWN_EGG":
+        case "SHEEP_SPAWN_EGG":
+        case "SHULKER_SPAWN_EGG":
+        case "SILVERFISH_SPAWN_EGG":
+        case "SKELETON_HORSE_SPAWN_EGG":
+        case "SKELETON_SPAWN_EGG":
+        case "SLIME_SPAWN_EGG":
+        case "SPIDER_SPAWN_EGG":
+        case "SQUID_SPAWN_EGG":
+        case "STRAY_SPAWN_EGG":
+        case "STRIDER_SPAWN_EGG":
+        case "TRADER_LLAMA_SPAWN_EGG":
+        case "TROPICAL_FISH_SPAWN_EGG":
+        case "TURTLE_SPAWN_EGG":
+        case "VEX_SPAWN_EGG":
+        case "VILLAGER_SPAWN_EGG":
+        case "VINDICATOR_SPAWN_EGG":
+        case "WANDERING_TRADER_SPAWN_EGG":
+        case "WITCH_SPAWN_EGG":
+        case "WITHER_SKELETON_SPAWN_EGG":
+        case "WOLF_SPAWN_EGG":
+        case "ZOGLIN_SPAWN_EGG":
+        case "ZOMBIE_HORSE_SPAWN_EGG":
+        case "ZOMBIE_SPAWN_EGG":
+        case "ZOMBIE_VILLAGER_SPAWN_EGG":
+        case "ZOMBIFIED_PIGLIN_SPAWN_EGG":
             return meta instanceof CraftMetaSpawnEgg ? meta : new CraftMetaSpawnEgg(meta);
-        case ARMOR_STAND:
+        case "ARMOR_STAND":
             return meta instanceof CraftMetaArmorStand ? meta : new CraftMetaArmorStand(meta);
-        case KNOWLEDGE_BOOK:
+        case "KNOWLEDGE_BOOK":
             return meta instanceof CraftMetaKnowledgeBook ? meta : new CraftMetaKnowledgeBook(meta);
-        case FURNACE:
-        case CHEST:
-        case TRAPPED_CHEST:
-        case JUKEBOX:
-        case DISPENSER:
-        case DROPPER:
-        case ACACIA_SIGN:
-        case ACACIA_WALL_SIGN:
-        case BIRCH_SIGN:
-        case BIRCH_WALL_SIGN:
-        case CRIMSON_SIGN:
-        case CRIMSON_WALL_SIGN:
-        case DARK_OAK_SIGN:
-        case DARK_OAK_WALL_SIGN:
-        case JUNGLE_SIGN:
-        case JUNGLE_WALL_SIGN:
-        case OAK_SIGN:
-        case OAK_WALL_SIGN:
-        case SPRUCE_SIGN:
-        case SPRUCE_WALL_SIGN:
-        case WARPED_SIGN:
-        case WARPED_WALL_SIGN:
-        case SPAWNER:
-        case BREWING_STAND:
-        case ENCHANTING_TABLE:
-        case COMMAND_BLOCK:
-        case REPEATING_COMMAND_BLOCK:
-        case CHAIN_COMMAND_BLOCK:
-        case BEACON:
-        case DAYLIGHT_DETECTOR:
-        case HOPPER:
-        case COMPARATOR:
-        case SHIELD:
-        case STRUCTURE_BLOCK:
-        case SHULKER_BOX:
-        case WHITE_SHULKER_BOX:
-        case ORANGE_SHULKER_BOX:
-        case MAGENTA_SHULKER_BOX:
-        case LIGHT_BLUE_SHULKER_BOX:
-        case YELLOW_SHULKER_BOX:
-        case LIME_SHULKER_BOX:
-        case PINK_SHULKER_BOX:
-        case GRAY_SHULKER_BOX:
-        case LIGHT_GRAY_SHULKER_BOX:
-        case CYAN_SHULKER_BOX:
-        case PURPLE_SHULKER_BOX:
-        case BLUE_SHULKER_BOX:
-        case BROWN_SHULKER_BOX:
-        case GREEN_SHULKER_BOX:
-        case RED_SHULKER_BOX:
-        case BLACK_SHULKER_BOX:
-        case ENDER_CHEST:
-        case BARREL:
-        case BELL:
-        case BLAST_FURNACE:
-        case CAMPFIRE:
-        case SOUL_CAMPFIRE:
-        case JIGSAW:
-        case LECTERN:
-        case SMOKER:
-        case BEEHIVE:
-        case BEE_NEST:
+        case "FURNACE":
+        case "CHEST":
+        case "TRAPPED_CHEST":
+        case "JUKEBOX":
+        case "DISPENSER":
+        case "DROPPER":
+        case "ACACIA_SIGN":
+        case "ACACIA_WALL_SIGN":
+        case "BIRCH_SIGN":
+        case "BIRCH_WALL_SIGN":
+        case "CRIMSON_SIGN":
+        case "CRIMSON_WALL_SIGN":
+        case "DARK_OAK_SIGN":
+        case "DARK_OAK_WALL_SIGN":
+        case "JUNGLE_SIGN":
+        case "JUNGLE_WALL_SIGN":
+        case "OAK_SIGN":
+        case "OAK_WALL_SIGN":
+        case "SPRUCE_SIGN":
+        case "SPRUCE_WALL_SIGN":
+        case "WARPED_SIGN":
+        case "WARPED_WALL_SIGN":
+        case "SPAWNER":
+        case "BREWING_STAND":
+        case "ENCHANTING_TABLE":
+        case "COMMAND_BLOCK":
+        case "REPEATING_COMMAND_BLOCK":
+        case "CHAIN_COMMAND_BLOCK":
+        case "BEACON":
+        case "DAYLIGHT_DETECTOR":
+        case "HOPPER":
+        case "COMPARATOR":
+        case "SHIELD":
+        case "STRUCTURE_BLOCK":
+        case "SHULKER_BOX":
+        case "WHITE_SHULKER_BOX":
+        case "ORANGE_SHULKER_BOX":
+        case "MAGENTA_SHULKER_BOX":
+        case "LIGHT_BLUE_SHULKER_BOX":
+        case "YELLOW_SHULKER_BOX":
+        case "LIME_SHULKER_BOX":
+        case "PINK_SHULKER_BOX":
+        case "GRAY_SHULKER_BOX":
+        case "LIGHT_GRAY_SHULKER_BOX":
+        case "CYAN_SHULKER_BOX":
+        case "PURPLE_SHULKER_BOX":
+        case "BLUE_SHULKER_BOX":
+        case "BROWN_SHULKER_BOX":
+        case "GREEN_SHULKER_BOX":
+        case "RED_SHULKER_BOX":
+        case "BLACK_SHULKER_BOX":
+        case "ENDER_CHEST":
+        case "BARREL":
+        case "BELL":
+        case "BLAST_FURNACE":
+        case "CAMPFIRE":
+        case "SOUL_CAMPFIRE":
+        case "JIGSAW":
+        case "LECTERN":
+        case "SMOKER":
+        case "BEEHIVE":
+        case "BEE_NEST":
             return new CraftMetaBlockState(meta, material);
-        case TROPICAL_FISH_BUCKET:
+        case "TROPICAL_FISH_BUCKET":
             return meta instanceof CraftMetaTropicalFishBucket ? meta : new CraftMetaTropicalFishBucket(meta);
-        case CROSSBOW:
+        case "CROSSBOW":
             return meta instanceof CraftMetaCrossbow ? meta : new CraftMetaCrossbow(meta);
-        case SUSPICIOUS_STEW:
+        case "SUSPICIOUS_STEW":
             return meta instanceof CraftMetaSuspiciousStew ? meta : new CraftMetaSuspiciousStew(meta);
-        case COD_BUCKET:
-        case PUFFERFISH_BUCKET:
-        case SALMON_BUCKET:
-        case ITEM_FRAME:
-        case PAINTING:
+        case "COD_BUCKET":
+        case "PUFFERFISH_BUCKET":
+        case "SALMON_BUCKET":
+        case "ITEM_FRAME":
+        case "PAINTING":
             return meta instanceof CraftMetaEntityTag ? meta : new CraftMetaEntityTag(meta);
-        case COMPASS:
+        case "COMPASS":
             return meta instanceof CraftMetaCompass ? meta : new CraftMetaCompass(meta);
         default:
             return new CraftMetaItem(meta);
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 09d7a86b5f6cffdf6664ad657dd4f8dd8eabdd70..f3eaa65483216ee6a57f411e052aa657f5ce32bf 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -279,220 +279,220 @@ public final class CraftItemStack extends ItemStack {
         if (!hasItemMeta(item)) {
             return CraftItemFactory.instance().getItemMeta(getType(item));
         }
-        switch (getType(item)) {
-            case WRITTEN_BOOK:
+        switch (getType(item).name()) {
+            case "WRITTEN_BOOK":
                 return new CraftMetaBookSigned(item.getTag());
-            case WRITABLE_BOOK:
+            case "WRITABLE_BOOK":
                 return new CraftMetaBook(item.getTag());
-            case CREEPER_HEAD:
-            case CREEPER_WALL_HEAD:
-            case DRAGON_HEAD:
-            case DRAGON_WALL_HEAD:
-            case PLAYER_HEAD:
-            case PLAYER_WALL_HEAD:
-            case SKELETON_SKULL:
-            case SKELETON_WALL_SKULL:
-            case WITHER_SKELETON_SKULL:
-            case WITHER_SKELETON_WALL_SKULL:
-            case ZOMBIE_HEAD:
-            case ZOMBIE_WALL_HEAD:
+            case "CREEPER_HEAD":
+            case "CREEPER_WALL_HEAD":
+            case "DRAGON_HEAD":
+            case "DRAGON_WALL_HEAD":
+            case "PLAYER_HEAD":
+            case "PLAYER_WALL_HEAD":
+            case "SKELETON_SKULL":
+            case "SKELETON_WALL_SKULL":
+            case "WITHER_SKELETON_SKULL":
+            case "WITHER_SKELETON_WALL_SKULL":
+            case "ZOMBIE_HEAD":
+            case "ZOMBIE_WALL_HEAD":
                 return new CraftMetaSkull(item.getTag());
-            case LEATHER_HELMET:
-            case LEATHER_HORSE_ARMOR:
-            case LEATHER_CHESTPLATE:
-            case LEATHER_LEGGINGS:
-            case LEATHER_BOOTS:
+            case "LEATHER_HELMET":
+            case "LEATHER_HORSE_ARMOR":
+            case "LEATHER_CHESTPLATE":
+            case "LEATHER_LEGGINGS":
+            case "LEATHER_BOOTS":
                 return new CraftMetaLeatherArmor(item.getTag());
-            case POTION:
-            case SPLASH_POTION:
-            case LINGERING_POTION:
-            case TIPPED_ARROW:
+            case "POTION":
+            case "SPLASH_POTION":
+            case "LINGERING_POTION":
+            case "TIPPED_ARROW":
                 return new CraftMetaPotion(item.getTag());
-            case FILLED_MAP:
+            case "FILLED_MAP":
                 return new CraftMetaMap(item.getTag());
-            case FIREWORK_ROCKET:
+            case "FIREWORK_ROCKET":
                 return new CraftMetaFirework(item.getTag());
-            case FIREWORK_STAR:
+            case "FIREWORK_STAR":
                 return new CraftMetaCharge(item.getTag());
-            case ENCHANTED_BOOK:
+            case "ENCHANTED_BOOK":
                 return new CraftMetaEnchantedBook(item.getTag());
-            case BLACK_BANNER:
-            case BLACK_WALL_BANNER:
-            case BLUE_BANNER:
-            case BLUE_WALL_BANNER:
-            case BROWN_BANNER:
-            case BROWN_WALL_BANNER:
-            case CYAN_BANNER:
-            case CYAN_WALL_BANNER:
-            case GRAY_BANNER:
-            case GRAY_WALL_BANNER:
-            case GREEN_BANNER:
-            case GREEN_WALL_BANNER:
-            case LIGHT_BLUE_BANNER:
-            case LIGHT_BLUE_WALL_BANNER:
-            case LIGHT_GRAY_BANNER:
-            case LIGHT_GRAY_WALL_BANNER:
-            case LIME_BANNER:
-            case LIME_WALL_BANNER:
-            case MAGENTA_BANNER:
-            case MAGENTA_WALL_BANNER:
-            case ORANGE_BANNER:
-            case ORANGE_WALL_BANNER:
-            case PINK_BANNER:
-            case PINK_WALL_BANNER:
-            case PURPLE_BANNER:
-            case PURPLE_WALL_BANNER:
-            case RED_BANNER:
-            case RED_WALL_BANNER:
-            case WHITE_BANNER:
-            case WHITE_WALL_BANNER:
-            case YELLOW_BANNER:
-            case YELLOW_WALL_BANNER:
+            case "BLACK_BANNER":
+            case "BLACK_WALL_BANNER":
+            case "BLUE_BANNER":
+            case "BLUE_WALL_BANNER":
+            case "BROWN_BANNER":
+            case "BROWN_WALL_BANNER":
+            case "CYAN_BANNER":
+            case "CYAN_WALL_BANNER":
+            case "GRAY_BANNER":
+            case "GRAY_WALL_BANNER":
+            case "GREEN_BANNER":
+            case "GREEN_WALL_BANNER":
+            case "LIGHT_BLUE_BANNER":
+            case "LIGHT_BLUE_WALL_BANNER":
+            case "LIGHT_GRAY_BANNER":
+            case "LIGHT_GRAY_WALL_BANNER":
+            case "LIME_BANNER":
+            case "LIME_WALL_BANNER":
+            case "MAGENTA_BANNER":
+            case "MAGENTA_WALL_BANNER":
+            case "ORANGE_BANNER":
+            case "ORANGE_WALL_BANNER":
+            case "PINK_BANNER":
+            case "PINK_WALL_BANNER":
+            case "PURPLE_BANNER":
+            case "PURPLE_WALL_BANNER":
+            case "RED_BANNER":
+            case "RED_WALL_BANNER":
+            case "WHITE_BANNER":
+            case "WHITE_WALL_BANNER":
+            case "YELLOW_BANNER":
+            case "YELLOW_WALL_BANNER":
                 return new CraftMetaBanner(item.getTag());
-            case BAT_SPAWN_EGG:
-            case BEE_SPAWN_EGG:
-            case BLAZE_SPAWN_EGG:
-            case CAT_SPAWN_EGG:
-            case CAVE_SPIDER_SPAWN_EGG:
-            case CHICKEN_SPAWN_EGG:
-            case COD_SPAWN_EGG:
-            case COW_SPAWN_EGG:
-            case CREEPER_SPAWN_EGG:
-            case DOLPHIN_SPAWN_EGG:
-            case DONKEY_SPAWN_EGG:
-            case DROWNED_SPAWN_EGG:
-            case ELDER_GUARDIAN_SPAWN_EGG:
-            case ENDERMAN_SPAWN_EGG:
-            case ENDERMITE_SPAWN_EGG:
-            case EVOKER_SPAWN_EGG:
-            case FOX_SPAWN_EGG:
-            case GHAST_SPAWN_EGG:
-            case GUARDIAN_SPAWN_EGG:
-            case HOGLIN_SPAWN_EGG:
-            case HORSE_SPAWN_EGG:
-            case HUSK_SPAWN_EGG:
-            case LLAMA_SPAWN_EGG:
-            case MAGMA_CUBE_SPAWN_EGG:
-            case MOOSHROOM_SPAWN_EGG:
-            case MULE_SPAWN_EGG:
-            case OCELOT_SPAWN_EGG:
-            case PANDA_SPAWN_EGG:
-            case PARROT_SPAWN_EGG:
-            case PHANTOM_SPAWN_EGG:
-            case PIGLIN_SPAWN_EGG:
-            case PIG_SPAWN_EGG:
-            case PILLAGER_SPAWN_EGG:
-            case POLAR_BEAR_SPAWN_EGG:
-            case PUFFERFISH_SPAWN_EGG:
-            case RABBIT_SPAWN_EGG:
-            case RAVAGER_SPAWN_EGG:
-            case SALMON_SPAWN_EGG:
-            case SHEEP_SPAWN_EGG:
-            case SHULKER_SPAWN_EGG:
-            case SILVERFISH_SPAWN_EGG:
-            case SKELETON_HORSE_SPAWN_EGG:
-            case SKELETON_SPAWN_EGG:
-            case SLIME_SPAWN_EGG:
-            case SPIDER_SPAWN_EGG:
-            case SQUID_SPAWN_EGG:
-            case STRAY_SPAWN_EGG:
-            case STRIDER_SPAWN_EGG:
-            case TRADER_LLAMA_SPAWN_EGG:
-            case TROPICAL_FISH_SPAWN_EGG:
-            case TURTLE_SPAWN_EGG:
-            case VEX_SPAWN_EGG:
-            case VILLAGER_SPAWN_EGG:
-            case VINDICATOR_SPAWN_EGG:
-            case WANDERING_TRADER_SPAWN_EGG:
-            case WITCH_SPAWN_EGG:
-            case WITHER_SKELETON_SPAWN_EGG:
-            case WOLF_SPAWN_EGG:
-            case ZOGLIN_SPAWN_EGG:
-            case ZOMBIE_HORSE_SPAWN_EGG:
-            case ZOMBIE_SPAWN_EGG:
-            case ZOMBIE_VILLAGER_SPAWN_EGG:
-            case ZOMBIFIED_PIGLIN_SPAWN_EGG:
+            case "BAT_SPAWN_EGG":
+            case "BEE_SPAWN_EGG":
+            case "BLAZE_SPAWN_EGG":
+            case "CAT_SPAWN_EGG":
+            case "CAVE_SPIDER_SPAWN_EGG":
+            case "CHICKEN_SPAWN_EGG":
+            case "COD_SPAWN_EGG":
+            case "COW_SPAWN_EGG":
+            case "CREEPER_SPAWN_EGG":
+            case "DOLPHIN_SPAWN_EGG":
+            case "DONKEY_SPAWN_EGG":
+            case "DROWNED_SPAWN_EGG":
+            case "ELDER_GUARDIAN_SPAWN_EGG":
+            case "ENDERMAN_SPAWN_EGG":
+            case "ENDERMITE_SPAWN_EGG":
+            case "EVOKER_SPAWN_EGG":
+            case "FOX_SPAWN_EGG":
+            case "GHAST_SPAWN_EGG":
+            case "GUARDIAN_SPAWN_EGG":
+            case "HOGLIN_SPAWN_EGG":
+            case "HORSE_SPAWN_EGG":
+            case "HUSK_SPAWN_EGG":
+            case "LLAMA_SPAWN_EGG":
+            case "MAGMA_CUBE_SPAWN_EGG":
+            case "MOOSHROOM_SPAWN_EGG":
+            case "MULE_SPAWN_EGG":
+            case "OCELOT_SPAWN_EGG":
+            case "PANDA_SPAWN_EGG":
+            case "PARROT_SPAWN_EGG":
+            case "PHANTOM_SPAWN_EGG":
+            case "PIGLIN_SPAWN_EGG":
+            case "PIG_SPAWN_EGG":
+            case "PILLAGER_SPAWN_EGG":
+            case "POLAR_BEAR_SPAWN_EGG":
+            case "PUFFERFISH_SPAWN_EGG":
+            case "RABBIT_SPAWN_EGG":
+            case "RAVAGER_SPAWN_EGG":
+            case "SALMON_SPAWN_EGG":
+            case "SHEEP_SPAWN_EGG":
+            case "SHULKER_SPAWN_EGG":
+            case "SILVERFISH_SPAWN_EGG":
+            case "SKELETON_HORSE_SPAWN_EGG":
+            case "SKELETON_SPAWN_EGG":
+            case "SLIME_SPAWN_EGG":
+            case "SPIDER_SPAWN_EGG":
+            case "SQUID_SPAWN_EGG":
+            case "STRAY_SPAWN_EGG":
+            case "STRIDER_SPAWN_EGG":
+            case "TRADER_LLAMA_SPAWN_EGG":
+            case "TROPICAL_FISH_SPAWN_EGG":
+            case "TURTLE_SPAWN_EGG":
+            case "VEX_SPAWN_EGG":
+            case "VILLAGER_SPAWN_EGG":
+            case "VINDICATOR_SPAWN_EGG":
+            case "WANDERING_TRADER_SPAWN_EGG":
+            case "WITCH_SPAWN_EGG":
+            case "WITHER_SKELETON_SPAWN_EGG":
+            case "WOLF_SPAWN_EGG":
+            case "ZOGLIN_SPAWN_EGG":
+            case "ZOMBIE_HORSE_SPAWN_EGG":
+            case "ZOMBIE_SPAWN_EGG":
+            case "ZOMBIE_VILLAGER_SPAWN_EGG":
+            case "ZOMBIFIED_PIGLIN_SPAWN_EGG":
                 return new CraftMetaSpawnEgg(item.getTag());
-            case ARMOR_STAND:
+            case "ARMOR_STAND":
                 return new CraftMetaArmorStand(item.getTag());
-            case KNOWLEDGE_BOOK:
+            case "KNOWLEDGE_BOOK":
                 return new CraftMetaKnowledgeBook(item.getTag());
-            case FURNACE:
-            case CHEST:
-            case TRAPPED_CHEST:
-            case JUKEBOX:
-            case DISPENSER:
-            case DROPPER:
-            case ACACIA_SIGN:
-            case ACACIA_WALL_SIGN:
-            case BIRCH_SIGN:
-            case BIRCH_WALL_SIGN:
-            case CRIMSON_SIGN:
-            case CRIMSON_WALL_SIGN:
-            case DARK_OAK_SIGN:
-            case DARK_OAK_WALL_SIGN:
-            case JUNGLE_SIGN:
-            case JUNGLE_WALL_SIGN:
-            case OAK_SIGN:
-            case OAK_WALL_SIGN:
-            case SPRUCE_SIGN:
-            case SPRUCE_WALL_SIGN:
-            case WARPED_SIGN:
-            case WARPED_WALL_SIGN:
-            case SPAWNER:
-            case BREWING_STAND:
-            case ENCHANTING_TABLE:
-            case COMMAND_BLOCK:
-            case REPEATING_COMMAND_BLOCK:
-            case CHAIN_COMMAND_BLOCK:
-            case BEACON:
-            case DAYLIGHT_DETECTOR:
-            case HOPPER:
-            case COMPARATOR:
-            case SHIELD:
-            case STRUCTURE_BLOCK:
-            case SHULKER_BOX:
-            case WHITE_SHULKER_BOX:
-            case ORANGE_SHULKER_BOX:
-            case MAGENTA_SHULKER_BOX:
-            case LIGHT_BLUE_SHULKER_BOX:
-            case YELLOW_SHULKER_BOX:
-            case LIME_SHULKER_BOX:
-            case PINK_SHULKER_BOX:
-            case GRAY_SHULKER_BOX:
-            case LIGHT_GRAY_SHULKER_BOX:
-            case CYAN_SHULKER_BOX:
-            case PURPLE_SHULKER_BOX:
-            case BLUE_SHULKER_BOX:
-            case BROWN_SHULKER_BOX:
-            case GREEN_SHULKER_BOX:
-            case RED_SHULKER_BOX:
-            case BLACK_SHULKER_BOX:
-            case ENDER_CHEST:
-            case BARREL:
-            case BELL:
-            case BLAST_FURNACE:
-            case CAMPFIRE:
-            case SOUL_CAMPFIRE:
-            case JIGSAW:
-            case LECTERN:
-            case SMOKER:
-            case BEEHIVE:
-            case BEE_NEST:
+            case "FURNACE":
+            case "CHEST":
+            case "TRAPPED_CHEST":
+            case "JUKEBOX":
+            case "DISPENSER":
+            case "DROPPER":
+            case "ACACIA_SIGN":
+            case "ACACIA_WALL_SIGN":
+            case "BIRCH_SIGN":
+            case "BIRCH_WALL_SIGN":
+            case "CRIMSON_SIGN":
+            case "CRIMSON_WALL_SIGN":
+            case "DARK_OAK_SIGN":
+            case "DARK_OAK_WALL_SIGN":
+            case "JUNGLE_SIGN":
+            case "JUNGLE_WALL_SIGN":
+            case "OAK_SIGN":
+            case "OAK_WALL_SIGN":
+            case "SPRUCE_SIGN":
+            case "SPRUCE_WALL_SIGN":
+            case "WARPED_SIGN":
+            case "WARPED_WALL_SIGN":
+            case "SPAWNER":
+            case "BREWING_STAND":
+            case "ENCHANTING_TABLE":
+            case "COMMAND_BLOCK":
+            case "REPEATING_COMMAND_BLOCK":
+            case "CHAIN_COMMAND_BLOCK":
+            case "BEACON":
+            case "DAYLIGHT_DETECTOR":
+            case "HOPPER":
+            case "COMPARATOR":
+            case "SHIELD":
+            case "STRUCTURE_BLOCK":
+            case "SHULKER_BOX":
+            case "WHITE_SHULKER_BOX":
+            case "ORANGE_SHULKER_BOX":
+            case "MAGENTA_SHULKER_BOX":
+            case "LIGHT_BLUE_SHULKER_BOX":
+            case "YELLOW_SHULKER_BOX":
+            case "LIME_SHULKER_BOX":
+            case "PINK_SHULKER_BOX":
+            case "GRAY_SHULKER_BOX":
+            case "LIGHT_GRAY_SHULKER_BOX":
+            case "CYAN_SHULKER_BOX":
+            case "PURPLE_SHULKER_BOX":
+            case "BLUE_SHULKER_BOX":
+            case "BROWN_SHULKER_BOX":
+            case "GREEN_SHULKER_BOX":
+            case "RED_SHULKER_BOX":
+            case "BLACK_SHULKER_BOX":
+            case "ENDER_CHEST":
+            case "BARREL":
+            case "BELL":
+            case "BLAST_FURNACE":
+            case "CAMPFIRE":
+            case "SOUL_CAMPFIRE":
+            case "JIGSAW":
+            case "LECTERN":
+            case "SMOKER":
+            case "BEEHIVE":
+            case "BEE_NEST":
                 return new CraftMetaBlockState(item.getTag(), CraftMagicNumbers.getMaterial(item.getItem()));
-            case TROPICAL_FISH_BUCKET:
+            case "TROPICAL_FISH_BUCKET":
                 return new CraftMetaTropicalFishBucket(item.getTag());
-            case CROSSBOW:
+            case "CROSSBOW":
                 return new CraftMetaCrossbow(item.getTag());
-            case SUSPICIOUS_STEW:
+            case "SUSPICIOUS_STEW":
                 return new CraftMetaSuspiciousStew(item.getTag());
-            case COD_BUCKET:
-            case PUFFERFISH_BUCKET:
-            case SALMON_BUCKET:
-            case ITEM_FRAME:
-            case PAINTING:
+            case "COD_BUCKET":
+            case "PUFFERFISH_BUCKET":
+            case "SALMON_BUCKET":
+            case "ITEM_FRAME":
+            case "PAINTING":
                 return new CraftMetaEntityTag(item.getTag());
-            case COMPASS:
+            case "COMPASS":
                 return new CraftMetaCompass(item.getTag());
             default:
                 return new CraftMetaItem(item.getTag());
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaArmorStand.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaArmorStand.java
index 6b460841a8428727dd55a841fe5af06de9b1bdd1..a2ee6a222b40fda54eeeb4891f012a779387ad88 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaArmorStand.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaArmorStand.java
@@ -146,8 +146,8 @@ public class CraftMetaArmorStand extends CraftMetaItem implements com.destroysto
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case ARMOR_STAND:
+        switch (type.name()) {
+            case "ARMOR_STAND":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
index 27e7d7919992a3d1567db8619e67d57b62a3a5d6..70ad2e9d733b687a7d9fbcf345db9f38ac0df93d 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
@@ -205,39 +205,39 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case BLACK_BANNER:
-            case BLACK_WALL_BANNER:
-            case BLUE_BANNER:
-            case BLUE_WALL_BANNER:
-            case BROWN_BANNER:
-            case BROWN_WALL_BANNER:
-            case CYAN_BANNER:
-            case CYAN_WALL_BANNER:
-            case GRAY_BANNER:
-            case GRAY_WALL_BANNER:
-            case GREEN_BANNER:
-            case GREEN_WALL_BANNER:
-            case LIGHT_BLUE_BANNER:
-            case LIGHT_BLUE_WALL_BANNER:
-            case LIGHT_GRAY_BANNER:
-            case LIGHT_GRAY_WALL_BANNER:
-            case LIME_BANNER:
-            case LIME_WALL_BANNER:
-            case MAGENTA_BANNER:
-            case MAGENTA_WALL_BANNER:
-            case ORANGE_BANNER:
-            case ORANGE_WALL_BANNER:
-            case PINK_BANNER:
-            case PINK_WALL_BANNER:
-            case PURPLE_BANNER:
-            case PURPLE_WALL_BANNER:
-            case RED_BANNER:
-            case RED_WALL_BANNER:
-            case WHITE_BANNER:
-            case WHITE_WALL_BANNER:
-            case YELLOW_BANNER:
-            case YELLOW_WALL_BANNER:
+        switch (type.name()) {
+            case "BLACK_BANNER":
+            case "BLACK_WALL_BANNER":
+            case "BLUE_BANNER":
+            case "BLUE_WALL_BANNER":
+            case "BROWN_BANNER":
+            case "BROWN_WALL_BANNER":
+            case "CYAN_BANNER":
+            case "CYAN_WALL_BANNER":
+            case "GRAY_BANNER":
+            case "GRAY_WALL_BANNER":
+            case "GREEN_BANNER":
+            case "GREEN_WALL_BANNER":
+            case "LIGHT_BLUE_BANNER":
+            case "LIGHT_BLUE_WALL_BANNER":
+            case "LIGHT_GRAY_BANNER":
+            case "LIGHT_GRAY_WALL_BANNER":
+            case "LIME_BANNER":
+            case "LIME_WALL_BANNER":
+            case "MAGENTA_BANNER":
+            case "MAGENTA_WALL_BANNER":
+            case "ORANGE_BANNER":
+            case "ORANGE_WALL_BANNER":
+            case "PINK_BANNER":
+            case "PINK_WALL_BANNER":
+            case "PURPLE_BANNER":
+            case "PURPLE_WALL_BANNER":
+            case "RED_BANNER":
+            case "RED_WALL_BANNER":
+            case "WHITE_BANNER":
+            case "WHITE_WALL_BANNER":
+            case "YELLOW_BANNER":
+            case "YELLOW_WALL_BANNER":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBlockState.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBlockState.java
index d003a2e5d910a9198e9eefdd3bfb6d173c57e2ec..8adeb82720032e007d53791baca86f11dcf36bf1 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBlockState.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBlockState.java
@@ -185,69 +185,69 @@ public class CraftMetaBlockState extends CraftMetaItem implements BlockStateMeta
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case FURNACE:
-            case CHEST:
-            case TRAPPED_CHEST:
-            case JUKEBOX:
-            case DISPENSER:
-            case DROPPER:
-            case ACACIA_SIGN:
-            case ACACIA_WALL_SIGN:
-            case BIRCH_SIGN:
-            case BIRCH_WALL_SIGN:
-            case CRIMSON_SIGN:
-            case CRIMSON_WALL_SIGN:
-            case DARK_OAK_SIGN:
-            case DARK_OAK_WALL_SIGN:
-            case JUNGLE_SIGN:
-            case JUNGLE_WALL_SIGN:
-            case OAK_SIGN:
-            case OAK_WALL_SIGN:
-            case SPRUCE_SIGN:
-            case SPRUCE_WALL_SIGN:
-            case WARPED_SIGN:
-            case WARPED_WALL_SIGN:
-            case SPAWNER:
-            case BREWING_STAND:
-            case ENCHANTING_TABLE:
-            case COMMAND_BLOCK:
-            case REPEATING_COMMAND_BLOCK:
-            case CHAIN_COMMAND_BLOCK:
-            case BEACON:
-            case DAYLIGHT_DETECTOR:
-            case HOPPER:
-            case COMPARATOR:
-            case SHIELD:
-            case STRUCTURE_BLOCK:
-            case SHULKER_BOX:
-            case WHITE_SHULKER_BOX:
-            case ORANGE_SHULKER_BOX:
-            case MAGENTA_SHULKER_BOX:
-            case LIGHT_BLUE_SHULKER_BOX:
-            case YELLOW_SHULKER_BOX:
-            case LIME_SHULKER_BOX:
-            case PINK_SHULKER_BOX:
-            case GRAY_SHULKER_BOX:
-            case LIGHT_GRAY_SHULKER_BOX:
-            case CYAN_SHULKER_BOX:
-            case PURPLE_SHULKER_BOX:
-            case BLUE_SHULKER_BOX:
-            case BROWN_SHULKER_BOX:
-            case GREEN_SHULKER_BOX:
-            case RED_SHULKER_BOX:
-            case BLACK_SHULKER_BOX:
-            case ENDER_CHEST:
-            case BARREL:
-            case BELL:
-            case BLAST_FURNACE:
-            case CAMPFIRE:
-            case SOUL_CAMPFIRE:
-            case JIGSAW:
-            case LECTERN:
-            case SMOKER:
-            case BEEHIVE:
-            case BEE_NEST:
+        switch (type.name()) {
+            case "FURNACE":
+            case "CHEST":
+            case "TRAPPED_CHEST":
+            case "JUKEBOX":
+            case "DISPENSER":
+            case "DROPPER":
+            case "ACACIA_SIGN":
+            case "ACACIA_WALL_SIGN":
+            case "BIRCH_SIGN":
+            case "BIRCH_WALL_SIGN":
+            case "CRIMSON_SIGN":
+            case "CRIMSON_WALL_SIGN":
+            case "DARK_OAK_SIGN":
+            case "DARK_OAK_WALL_SIGN":
+            case "JUNGLE_SIGN":
+            case "JUNGLE_WALL_SIGN":
+            case "OAK_SIGN":
+            case "OAK_WALL_SIGN":
+            case "SPRUCE_SIGN":
+            case "SPRUCE_WALL_SIGN":
+            case "WARPED_SIGN":
+            case "WARPED_WALL_SIGN":
+            case "SPAWNER":
+            case "BREWING_STAND":
+            case "ENCHANTING_TABLE":
+            case "COMMAND_BLOCK":
+            case "REPEATING_COMMAND_BLOCK":
+            case "CHAIN_COMMAND_BLOCK":
+            case "BEACON":
+            case "DAYLIGHT_DETECTOR":
+            case "HOPPER":
+            case "COMPARATOR":
+            case "SHIELD":
+            case "STRUCTURE_BLOCK":
+            case "SHULKER_BOX":
+            case "WHITE_SHULKER_BOX":
+            case "ORANGE_SHULKER_BOX":
+            case "MAGENTA_SHULKER_BOX":
+            case "LIGHT_BLUE_SHULKER_BOX":
+            case "YELLOW_SHULKER_BOX":
+            case "LIME_SHULKER_BOX":
+            case "PINK_SHULKER_BOX":
+            case "GRAY_SHULKER_BOX":
+            case "LIGHT_GRAY_SHULKER_BOX":
+            case "CYAN_SHULKER_BOX":
+            case "PURPLE_SHULKER_BOX":
+            case "BLUE_SHULKER_BOX":
+            case "BROWN_SHULKER_BOX":
+            case "GREEN_SHULKER_BOX":
+            case "RED_SHULKER_BOX":
+            case "BLACK_SHULKER_BOX":
+            case "ENDER_CHEST":
+            case "BARREL":
+            case "BELL":
+            case "BLAST_FURNACE":
+            case "CAMPFIRE":
+            case "SOUL_CAMPFIRE":
+            case "JIGSAW":
+            case "LECTERN":
+            case "SMOKER":
+            case "BEEHIVE":
+            case "BEE_NEST":
                 return true;
         }
         return false;
@@ -271,259 +271,259 @@ public class CraftMetaBlockState extends CraftMetaItem implements BlockStateMeta
     public BlockState getBlockState() {
         Material stateMaterial = material; // Only actually used for jigsaws
         if (blockEntityTag != null) {
-            switch (material) {
-                case SHIELD:
+            switch (material.name()) {
+                case "SHIELD":
                     blockEntityTag.setString("id", "banner");
                     stateMaterial = shieldToBannerHack(blockEntityTag);
                     break;
-                case SHULKER_BOX:
-                case WHITE_SHULKER_BOX:
-                case ORANGE_SHULKER_BOX:
-                case MAGENTA_SHULKER_BOX:
-                case LIGHT_BLUE_SHULKER_BOX:
-                case YELLOW_SHULKER_BOX:
-                case LIME_SHULKER_BOX:
-                case PINK_SHULKER_BOX:
-                case GRAY_SHULKER_BOX:
-                case LIGHT_GRAY_SHULKER_BOX:
-                case CYAN_SHULKER_BOX:
-                case PURPLE_SHULKER_BOX:
-                case BLUE_SHULKER_BOX:
-                case BROWN_SHULKER_BOX:
-                case GREEN_SHULKER_BOX:
-                case RED_SHULKER_BOX:
-                case BLACK_SHULKER_BOX:
+                case "SHULKER_BOX":
+                case "WHITE_SHULKER_BOX":
+                case "ORANGE_SHULKER_BOX":
+                case "MAGENTA_SHULKER_BOX":
+                case "LIGHT_BLUE_SHULKER_BOX":
+                case "YELLOW_SHULKER_BOX":
+                case "LIME_SHULKER_BOX":
+                case "PINK_SHULKER_BOX":
+                case "GRAY_SHULKER_BOX":
+                case "LIGHT_GRAY_SHULKER_BOX":
+                case "CYAN_SHULKER_BOX":
+                case "PURPLE_SHULKER_BOX":
+                case "BLUE_SHULKER_BOX":
+                case "BROWN_SHULKER_BOX":
+                case "GREEN_SHULKER_BOX":
+                case "RED_SHULKER_BOX":
+                case "BLACK_SHULKER_BOX":
                     blockEntityTag.setString("id", "shulker_box");
                     break;
-                case BEE_NEST:
-                case BEEHIVE:
+                case "BEE_NEST":
+                case "BEEHIVE":
                     blockEntityTag.setString("id", "beehive");
                     break;
             }
         }
         TileEntity te = (blockEntityTag == null) ? null : TileEntity.create(CraftMagicNumbers.getBlock(stateMaterial).getBlockData(), blockEntityTag);
 
-        switch (material) {
-        case ACACIA_SIGN:
-        case ACACIA_WALL_SIGN:
-        case BIRCH_SIGN:
-        case BIRCH_WALL_SIGN:
-        case CRIMSON_SIGN:
-        case CRIMSON_WALL_SIGN:
-        case DARK_OAK_SIGN:
-        case DARK_OAK_WALL_SIGN:
-        case JUNGLE_SIGN:
-        case JUNGLE_WALL_SIGN:
-        case OAK_SIGN:
-        case OAK_WALL_SIGN:
-        case SPRUCE_SIGN:
-        case SPRUCE_WALL_SIGN:
-        case WARPED_SIGN:
-        case WARPED_WALL_SIGN:
+        switch (material.name()) {
+        case "ACACIA_SIGN":
+        case "ACACIA_WALL_SIGN":
+        case "BIRCH_SIGN":
+        case "BIRCH_WALL_SIGN":
+        case "CRIMSON_SIGN":
+        case "CRIMSON_WALL_SIGN":
+        case "DARK_OAK_SIGN":
+        case "DARK_OAK_WALL_SIGN":
+        case "JUNGLE_SIGN":
+        case "JUNGLE_WALL_SIGN":
+        case "OAK_SIGN":
+        case "OAK_WALL_SIGN":
+        case "SPRUCE_SIGN":
+        case "SPRUCE_WALL_SIGN":
+        case "WARPED_SIGN":
+        case "WARPED_WALL_SIGN":
             if (te == null) {
                 te = new TileEntitySign();
             }
             return new CraftSign(material, (TileEntitySign) te);
-        case CHEST:
-        case TRAPPED_CHEST:
+        case "CHEST":
+        case "TRAPPED_CHEST":
             if (te == null) {
                 te = new TileEntityChest();
             }
             return new CraftChest(material, (TileEntityChest) te);
-        case FURNACE:
+        case "FURNACE":
             if (te == null) {
                 te = new TileEntityFurnaceFurnace();
             }
             return new CraftFurnaceFurnace(material, (TileEntityFurnaceFurnace) te);
-        case DISPENSER:
+        case "DISPENSER":
             if (te == null) {
                 te = new TileEntityDispenser();
             }
             return new CraftDispenser(material, (TileEntityDispenser) te);
-        case DROPPER:
+        case "DROPPER":
             if (te == null) {
                 te = new TileEntityDropper();
             }
             return new CraftDropper(material, (TileEntityDropper) te);
-        case END_GATEWAY:
+        case "END_GATEWAY":
             if (te == null) {
                 te = new TileEntityEndGateway();
             }
             return new CraftEndGateway(material, (TileEntityEndGateway) te);
-        case HOPPER:
+        case "HOPPER":
             if (te == null) {
                 te = new TileEntityHopper();
             }
             return new CraftHopper(material, (TileEntityHopper) te);
-        case SPAWNER:
+        case "SPAWNER":
             if (te == null) {
                 te = new TileEntityMobSpawner();
             }
             return new CraftCreatureSpawner(material, (TileEntityMobSpawner) te);
-        case JUKEBOX:
+        case "JUKEBOX":
             if (te == null) {
                 te = new TileEntityJukeBox();
             }
             return new CraftJukebox(material, (TileEntityJukeBox) te);
-        case BREWING_STAND:
+        case "BREWING_STAND":
             if (te == null) {
                 te = new TileEntityBrewingStand();
             }
             return new CraftBrewingStand(material, (TileEntityBrewingStand) te);
-        case CREEPER_HEAD:
-        case CREEPER_WALL_HEAD:
-        case DRAGON_HEAD:
-        case DRAGON_WALL_HEAD:
-        case PLAYER_HEAD:
-        case PLAYER_WALL_HEAD:
-        case SKELETON_SKULL:
-        case SKELETON_WALL_SKULL:
-        case WITHER_SKELETON_SKULL:
-        case WITHER_SKELETON_WALL_SKULL:
-        case ZOMBIE_HEAD:
-        case ZOMBIE_WALL_HEAD:
+        case "CREEPER_HEAD":
+        case "CREEPER_WALL_HEAD":
+        case "DRAGON_HEAD":
+        case "DRAGON_WALL_HEAD":
+        case "PLAYER_HEAD":
+        case "PLAYER_WALL_HEAD":
+        case "SKELETON_SKULL":
+        case "SKELETON_WALL_SKULL":
+        case "WITHER_SKELETON_SKULL":
+        case "WITHER_SKELETON_WALL_SKULL":
+        case "ZOMBIE_HEAD":
+        case "ZOMBIE_WALL_HEAD":
             if (te == null) {
                 te = new TileEntitySkull();
             }
             return new CraftSkull(material, (TileEntitySkull) te);
-        case COMMAND_BLOCK:
-        case REPEATING_COMMAND_BLOCK:
-        case CHAIN_COMMAND_BLOCK:
+        case "COMMAND_BLOCK":
+        case "REPEATING_COMMAND_BLOCK":
+        case "CHAIN_COMMAND_BLOCK":
             if (te == null) {
                 te = new TileEntityCommand();
             }
             return new CraftCommandBlock(material, (TileEntityCommand) te);
-        case BEACON:
+        case "BEACON":
             if (te == null) {
                 te = new TileEntityBeacon();
             }
             return new CraftBeacon(material, (TileEntityBeacon) te);
-        case SHIELD:
+        case "SHIELD":
             if (te == null) {
                 te = new TileEntityBanner();
             }
             ((TileEntityBanner) te).color = (blockEntityTag == null) ? EnumColor.WHITE : EnumColor.fromColorIndex(blockEntityTag.getInt(CraftMetaBanner.BASE.NBT));
-        case BLACK_BANNER:
-        case BLACK_WALL_BANNER:
-        case BLUE_BANNER:
-        case BLUE_WALL_BANNER:
-        case BROWN_BANNER:
-        case BROWN_WALL_BANNER:
-        case CYAN_BANNER:
-        case CYAN_WALL_BANNER:
-        case GRAY_BANNER:
-        case GRAY_WALL_BANNER:
-        case GREEN_BANNER:
-        case GREEN_WALL_BANNER:
-        case LIGHT_BLUE_BANNER:
-        case LIGHT_BLUE_WALL_BANNER:
-        case LIGHT_GRAY_BANNER:
-        case LIGHT_GRAY_WALL_BANNER:
-        case LIME_BANNER:
-        case LIME_WALL_BANNER:
-        case MAGENTA_BANNER:
-        case MAGENTA_WALL_BANNER:
-        case ORANGE_BANNER:
-        case ORANGE_WALL_BANNER:
-        case PINK_BANNER:
-        case PINK_WALL_BANNER:
-        case PURPLE_BANNER:
-        case PURPLE_WALL_BANNER:
-        case RED_BANNER:
-        case RED_WALL_BANNER:
-        case WHITE_BANNER:
-        case WHITE_WALL_BANNER:
-        case YELLOW_BANNER:
-        case YELLOW_WALL_BANNER:
+        case "BLACK_BANNER":
+        case "BLACK_WALL_BANNER":
+        case "BLUE_BANNER":
+        case "BLUE_WALL_BANNER":
+        case "BROWN_BANNER":
+        case "BROWN_WALL_BANNER":
+        case "CYAN_BANNER":
+        case "CYAN_WALL_BANNER":
+        case "GRAY_BANNER":
+        case "GRAY_WALL_BANNER":
+        case "GREEN_BANNER":
+        case "GREEN_WALL_BANNER":
+        case "LIGHT_BLUE_BANNER":
+        case "LIGHT_BLUE_WALL_BANNER":
+        case "LIGHT_GRAY_BANNER":
+        case "LIGHT_GRAY_WALL_BANNER":
+        case "LIME_BANNER":
+        case "LIME_WALL_BANNER":
+        case "MAGENTA_BANNER":
+        case "MAGENTA_WALL_BANNER":
+        case "ORANGE_BANNER":
+        case "ORANGE_WALL_BANNER":
+        case "PINK_BANNER":
+        case "PINK_WALL_BANNER":
+        case "PURPLE_BANNER":
+        case "PURPLE_WALL_BANNER":
+        case "RED_BANNER":
+        case "RED_WALL_BANNER":
+        case "WHITE_BANNER":
+        case "WHITE_WALL_BANNER":
+        case "YELLOW_BANNER":
+        case "YELLOW_WALL_BANNER":
             if (te == null) {
                 te = new TileEntityBanner();
             }
             return new CraftBanner(material == Material.SHIELD ? shieldToBannerHack(blockEntityTag) : material, (TileEntityBanner) te);
-        case STRUCTURE_BLOCK:
+        case "STRUCTURE_BLOCK":
             if (te == null) {
                 te = new TileEntityStructure();
             }
             return new CraftStructureBlock(material, (TileEntityStructure) te);
-        case SHULKER_BOX:
-        case WHITE_SHULKER_BOX:
-        case ORANGE_SHULKER_BOX:
-        case MAGENTA_SHULKER_BOX:
-        case LIGHT_BLUE_SHULKER_BOX:
-        case YELLOW_SHULKER_BOX:
-        case LIME_SHULKER_BOX:
-        case PINK_SHULKER_BOX:
-        case GRAY_SHULKER_BOX:
-        case LIGHT_GRAY_SHULKER_BOX:
-        case CYAN_SHULKER_BOX:
-        case PURPLE_SHULKER_BOX:
-        case BLUE_SHULKER_BOX:
-        case BROWN_SHULKER_BOX:
-        case GREEN_SHULKER_BOX:
-        case RED_SHULKER_BOX:
-        case BLACK_SHULKER_BOX:
+        case "SHULKER_BOX":
+        case "WHITE_SHULKER_BOX":
+        case "ORANGE_SHULKER_BOX":
+        case "MAGENTA_SHULKER_BOX":
+        case "LIGHT_BLUE_SHULKER_BOX":
+        case "YELLOW_SHULKER_BOX":
+        case "LIME_SHULKER_BOX":
+        case "PINK_SHULKER_BOX":
+        case "GRAY_SHULKER_BOX":
+        case "LIGHT_GRAY_SHULKER_BOX":
+        case "CYAN_SHULKER_BOX":
+        case "PURPLE_SHULKER_BOX":
+        case "BLUE_SHULKER_BOX":
+        case "BROWN_SHULKER_BOX":
+        case "GREEN_SHULKER_BOX":
+        case "RED_SHULKER_BOX":
+        case "BLACK_SHULKER_BOX":
             if (te == null) {
                 te = new TileEntityShulkerBox();
             }
             return new CraftShulkerBox(material, (TileEntityShulkerBox) te);
-        case ENCHANTING_TABLE:
+        case "ENCHANTING_TABLE":
             if (te == null) {
                 te = new TileEntityEnchantTable();
             }
             return new CraftEnchantingTable(material, (TileEntityEnchantTable) te);
-        case ENDER_CHEST:
-            if (te == null) {
+        case "ENDER_CHEST":
+            if (te == null){
                 te = new TileEntityEnderChest();
             }
             return new CraftEnderChest(material, (TileEntityEnderChest) te);
-        case DAYLIGHT_DETECTOR:
-            if (te == null) {
+        case "DAYLIGHT_DETECTOR":
+            if (te == null){
                 te = new TileEntityLightDetector();
             }
             return new CraftDaylightDetector(material, (TileEntityLightDetector) te);
-        case COMPARATOR:
-            if (te == null) {
+        case "COMPARATOR":
+            if (te == null){
                 te = new TileEntityComparator();
             }
             return new CraftComparator(material, (TileEntityComparator) te);
-        case BARREL:
-            if (te == null) {
+        case "BARREL":
+            if (te == null){
                 te = new TileEntityBarrel();
             }
             return new CraftBarrel(material, (TileEntityBarrel) te);
-        case BELL:
-            if (te == null) {
+        case "BELL":
+            if (te == null){
                 te = new TileEntityBell();
             }
             return new CraftBell(material, (TileEntityBell) te);
-        case BLAST_FURNACE:
-            if (te == null) {
+        case "BLAST_FURNACE":
+            if (te == null){
                 te = new TileEntityBlastFurnace();
             }
             return new CraftBlastFurnace(material, (TileEntityBlastFurnace) te);
-        case CAMPFIRE:
-        case SOUL_CAMPFIRE:
+        case "CAMPFIRE":
+        case "SOUL_CAMPFIRE":
             if (te == null) {
                 te = new TileEntityCampfire();
             }
             return new CraftCampfire(material, (TileEntityCampfire) te);
-        case JIGSAW:
-            if (te == null) {
+        case "JIGSAW":
+            if (te == null){
                 te = new TileEntityJigsaw();
             }
             return new CraftJigsaw(material, (TileEntityJigsaw) te);
-        case LECTERN:
-            if (te == null) {
+        case "LECTERN":
+            if (te == null){
                 te = new TileEntityLectern();
             }
             return new CraftLectern(material, (TileEntityLectern) te);
-        case SMOKER:
-            if (te == null) {
+        case "SMOKER":
+            if (te == null){
                 te = new TileEntitySmoker();
             }
             return new CraftSmoker(material, (TileEntitySmoker) te);
-        case BEE_NEST:
-        case BEEHIVE:
-            if (te == null) {
+        case "BEE_NEST":
+        case "BEEHIVE":
+            if (te == null){
                 te = new TileEntityBeehive();
             }
             return new CraftBeehive(material, (TileEntityBeehive) te);
@@ -537,168 +537,168 @@ public class CraftMetaBlockState extends CraftMetaItem implements BlockStateMeta
         Validate.notNull(blockState, "blockState must not be null");
 
         boolean valid;
-        switch (material) {
-        case ACACIA_SIGN:
-        case ACACIA_WALL_SIGN:
-        case BIRCH_SIGN:
-        case BIRCH_WALL_SIGN:
-        case CRIMSON_SIGN:
-        case CRIMSON_WALL_SIGN:
-        case DARK_OAK_SIGN:
-        case DARK_OAK_WALL_SIGN:
-        case JUNGLE_SIGN:
-        case JUNGLE_WALL_SIGN:
-        case OAK_SIGN:
-        case OAK_WALL_SIGN:
-        case SPRUCE_SIGN:
-        case SPRUCE_WALL_SIGN:
-        case WARPED_SIGN:
-        case WARPED_WALL_SIGN:
+        switch (material.name()) {
+        case "ACACIA_SIGN":
+        case "ACACIA_WALL_SIGN":
+        case "BIRCH_SIGN":
+        case "BIRCH_WALL_SIGN":
+        case "CRIMSON_SIGN":
+        case "CRIMSON_WALL_SIGN":
+        case "DARK_OAK_SIGN":
+        case "DARK_OAK_WALL_SIGN":
+        case "JUNGLE_SIGN":
+        case "JUNGLE_WALL_SIGN":
+        case "OAK_SIGN":
+        case "OAK_WALL_SIGN":
+        case "SPRUCE_SIGN":
+        case "SPRUCE_WALL_SIGN":
+        case "WARPED_SIGN":
+        case "WARPED_WALL_SIGN":
             valid = blockState instanceof CraftSign;
             break;
-        case CHEST:
-        case TRAPPED_CHEST:
+        case "CHEST":
+        case "TRAPPED_CHEST":
             valid = blockState instanceof CraftChest;
             break;
-        case FURNACE:
+        case "FURNACE":
             valid = blockState instanceof CraftFurnaceFurnace;
             break;
-        case DISPENSER:
+        case "DISPENSER":
             valid = blockState instanceof CraftDispenser;
             break;
-        case DROPPER:
+        case "DROPPER":
             valid = blockState instanceof CraftDropper;
             break;
-        case END_GATEWAY:
+        case "END_GATEWAY":
             valid = blockState instanceof CraftEndGateway;
             break;
-        case HOPPER:
+        case "HOPPER":
             valid = blockState instanceof CraftHopper;
             break;
-        case SPAWNER:
+        case "SPAWNER":
             valid = blockState instanceof CraftCreatureSpawner;
             break;
-        case JUKEBOX:
+        case "JUKEBOX":
             valid = blockState instanceof CraftJukebox;
             break;
-        case BREWING_STAND:
+        case "BREWING_STAND":
             valid = blockState instanceof CraftBrewingStand;
             break;
-        case CREEPER_HEAD:
-        case CREEPER_WALL_HEAD:
-        case DRAGON_HEAD:
-        case DRAGON_WALL_HEAD:
-        case PLAYER_HEAD:
-        case PLAYER_WALL_HEAD:
-        case SKELETON_SKULL:
-        case SKELETON_WALL_SKULL:
-        case WITHER_SKELETON_SKULL:
-        case WITHER_SKELETON_WALL_SKULL:
-        case ZOMBIE_HEAD:
-        case ZOMBIE_WALL_HEAD:
+        case "CREEPER_HEAD":
+        case "CREEPER_WALL_HEAD":
+        case "DRAGON_HEAD":
+        case "DRAGON_WALL_HEAD":
+        case "PLAYER_HEAD":
+        case "PLAYER_WALL_HEAD":
+        case "SKELETON_SKULL":
+        case "SKELETON_WALL_SKULL":
+        case "WITHER_SKELETON_SKULL":
+        case "WITHER_SKELETON_WALL_SKULL":
+        case "ZOMBIE_HEAD":
+        case "ZOMBIE_WALL_HEAD":
             valid = blockState instanceof CraftSkull;
             break;
-        case COMMAND_BLOCK:
-        case REPEATING_COMMAND_BLOCK:
-        case CHAIN_COMMAND_BLOCK:
+        case "COMMAND_BLOCK":
+        case "REPEATING_COMMAND_BLOCK":
+        case "CHAIN_COMMAND_BLOCK":
             valid = blockState instanceof CraftCommandBlock;
             break;
-        case BEACON:
+        case "BEACON":
             valid = blockState instanceof CraftBeacon;
             break;
-        case SHIELD:
-        case BLACK_BANNER:
-        case BLACK_WALL_BANNER:
-        case BLUE_BANNER:
-        case BLUE_WALL_BANNER:
-        case BROWN_BANNER:
-        case BROWN_WALL_BANNER:
-        case CYAN_BANNER:
-        case CYAN_WALL_BANNER:
-        case GRAY_BANNER:
-        case GRAY_WALL_BANNER:
-        case GREEN_BANNER:
-        case GREEN_WALL_BANNER:
-        case LIGHT_BLUE_BANNER:
-        case LIGHT_BLUE_WALL_BANNER:
-        case LIGHT_GRAY_BANNER:
-        case LIGHT_GRAY_WALL_BANNER:
-        case LIME_BANNER:
-        case LIME_WALL_BANNER:
-        case MAGENTA_BANNER:
-        case MAGENTA_WALL_BANNER:
-        case ORANGE_BANNER:
-        case ORANGE_WALL_BANNER:
-        case PINK_BANNER:
-        case PINK_WALL_BANNER:
-        case PURPLE_BANNER:
-        case PURPLE_WALL_BANNER:
-        case RED_BANNER:
-        case RED_WALL_BANNER:
-        case WHITE_BANNER:
-        case WHITE_WALL_BANNER:
-        case YELLOW_BANNER:
-        case YELLOW_WALL_BANNER:
+        case "SHIELD":
+        case "BLACK_BANNER":
+        case "BLACK_WALL_BANNER":
+        case "BLUE_BANNER":
+        case "BLUE_WALL_BANNER":
+        case "BROWN_BANNER":
+        case "BROWN_WALL_BANNER":
+        case "CYAN_BANNER":
+        case "CYAN_WALL_BANNER":
+        case "GRAY_BANNER":
+        case "GRAY_WALL_BANNER":
+        case "GREEN_BANNER":
+        case "GREEN_WALL_BANNER":
+        case "LIGHT_BLUE_BANNER":
+        case "LIGHT_BLUE_WALL_BANNER":
+        case "LIGHT_GRAY_BANNER":
+        case "LIGHT_GRAY_WALL_BANNER":
+        case "LIME_BANNER":
+        case "LIME_WALL_BANNER":
+        case "MAGENTA_BANNER":
+        case "MAGENTA_WALL_BANNER":
+        case "ORANGE_BANNER":
+        case "ORANGE_WALL_BANNER":
+        case "PINK_BANNER":
+        case "PINK_WALL_BANNER":
+        case "PURPLE_BANNER":
+        case "PURPLE_WALL_BANNER":
+        case "RED_BANNER":
+        case "RED_WALL_BANNER":
+        case "WHITE_BANNER":
+        case "WHITE_WALL_BANNER":
+        case "YELLOW_BANNER":
+        case "YELLOW_WALL_BANNER":
             valid = blockState instanceof CraftBanner;
             break;
-        case STRUCTURE_BLOCK:
+        case "STRUCTURE_BLOCK":
             valid = blockState instanceof CraftStructureBlock;
             break;
-        case SHULKER_BOX:
-        case WHITE_SHULKER_BOX:
-        case ORANGE_SHULKER_BOX:
-        case MAGENTA_SHULKER_BOX:
-        case LIGHT_BLUE_SHULKER_BOX:
-        case YELLOW_SHULKER_BOX:
-        case LIME_SHULKER_BOX:
-        case PINK_SHULKER_BOX:
-        case GRAY_SHULKER_BOX:
-        case LIGHT_GRAY_SHULKER_BOX:
-        case CYAN_SHULKER_BOX:
-        case PURPLE_SHULKER_BOX:
-        case BLUE_SHULKER_BOX:
-        case BROWN_SHULKER_BOX:
-        case GREEN_SHULKER_BOX:
-        case RED_SHULKER_BOX:
-        case BLACK_SHULKER_BOX:
+        case "SHULKER_BOX":
+        case "WHITE_SHULKER_BOX":
+        case "ORANGE_SHULKER_BOX":
+        case "MAGENTA_SHULKER_BOX":
+        case "LIGHT_BLUE_SHULKER_BOX":
+        case "YELLOW_SHULKER_BOX":
+        case "LIME_SHULKER_BOX":
+        case "PINK_SHULKER_BOX":
+        case "GRAY_SHULKER_BOX":
+        case "LIGHT_GRAY_SHULKER_BOX":
+        case "CYAN_SHULKER_BOX":
+        case "PURPLE_SHULKER_BOX":
+        case "BLUE_SHULKER_BOX":
+        case "BROWN_SHULKER_BOX":
+        case "GREEN_SHULKER_BOX":
+        case "RED_SHULKER_BOX":
+        case "BLACK_SHULKER_BOX":
             valid = blockState instanceof CraftShulkerBox;
             break;
-        case ENCHANTING_TABLE:
+        case "ENCHANTING_TABLE":
             valid = blockState instanceof CraftEnchantingTable;
             break;
-        case ENDER_CHEST:
+        case "ENDER_CHEST":
             valid = blockState instanceof CraftEnderChest;
             break;
-        case DAYLIGHT_DETECTOR:
+        case "DAYLIGHT_DETECTOR":
             valid = blockState instanceof CraftDaylightDetector;
             break;
-        case COMPARATOR:
+        case "COMPARATOR":
             valid = blockState instanceof CraftComparator;
             break;
-        case BARREL:
+        case "BARREL":
             valid = blockState instanceof CraftBarrel;
             break;
-        case BELL:
+        case "BELL":
             valid = blockState instanceof CraftBell;
             break;
-        case BLAST_FURNACE:
+        case "BLAST_FURNACE":
             valid = blockState instanceof CraftBlastFurnace;
             break;
-        case CAMPFIRE:
-        case SOUL_CAMPFIRE:
+        case "CAMPFIRE":
+        case "SOUL_CAMPFIRE":
             valid = blockState instanceof CraftCampfire;
             break;
-        case JIGSAW:
+        case "JIGSAW":
             valid = blockState instanceof CraftJigsaw;
             break;
-        case LECTERN:
+        case "LECTERN":
             valid = blockState instanceof CraftLectern;
             break;
-        case SMOKER:
+        case "SMOKER":
             valid = blockState instanceof CraftSmoker;
             break;
-        case BEEHIVE:
-        case BEE_NEST:
+        case "BEEHIVE":
+        case "BEE_NEST":
             valid = blockState instanceof CraftBeehive;
             break;
         default:
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
index 418b5353cae3fa2ada948b416a3fd7422101ae5c..7308208fce610222d8046c09cca62cf34e096fb5 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
@@ -204,9 +204,9 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-        case WRITTEN_BOOK:
-        case WRITABLE_BOOK:
+        switch (type.name()) {
+        case "WRITTEN_BOOK":
+        case "WRITABLE_BOOK":
             return true;
         default:
             return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBookSigned.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBookSigned.java
index 0d58ec9834797ad7b9acaae6353dcf0385c53fd4..b10f394e2161635931777b1c39b7b988eede1bc0 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBookSigned.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBookSigned.java
@@ -51,9 +51,9 @@ class CraftMetaBookSigned extends CraftMetaBook implements BookMeta {
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-        case WRITTEN_BOOK:
-        case WRITABLE_BOOK:
+        switch (type.name()) {
+        case "WRITTEN_BOOK":
+        case "WRITABLE_BOOK":
             return true;
         default:
             return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaCharge.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaCharge.java
index 4a6437f6e5d3ed4f8743dfc5a57bfcdcf315043d..b96c3119af8ddd7347c29f983d2a63ba2d0b17bc 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaCharge.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaCharge.java
@@ -68,8 +68,8 @@ class CraftMetaCharge extends CraftMetaItem implements FireworkEffectMeta {
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case FIREWORK_STAR:
+        switch (type.name()) {
+            case "FIREWORK_STAR":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaCrossbow.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaCrossbow.java
index e9965318d2bd8982557d0c3e687c59b0b1649eb0..789061d3e1e69ae42560c252893e99107c9a8015 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaCrossbow.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaCrossbow.java
@@ -99,8 +99,8 @@ public class CraftMetaCrossbow extends CraftMetaItem implements CrossbowMeta {
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case CROSSBOW:
+        switch (type.name()) {
+            case "CROSSBOW":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaEnchantedBook.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaEnchantedBook.java
index 943af29e5e17d67b09a42e8e0cdcc348b4676ffb..ff004072e6ec2c15e2c5a165cc1ba2f37b90a0ad 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaEnchantedBook.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaEnchantedBook.java
@@ -57,8 +57,8 @@ class CraftMetaEnchantedBook extends CraftMetaItem implements EnchantmentStorage
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case ENCHANTED_BOOK:
+        switch (type.name()) {
+            case "ENCHANTED_BOOK":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaEntityTag.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaEntityTag.java
index a8a5db50ff0e64723002c4a99db9fe673c747b84..6732b146d74685421214ba3ee501afe33fcf64f8 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaEntityTag.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaEntityTag.java
@@ -64,12 +64,12 @@ public class CraftMetaEntityTag extends CraftMetaItem {
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case COD_BUCKET:
-            case PUFFERFISH_BUCKET:
-            case SALMON_BUCKET:
-            case ITEM_FRAME:
-            case PAINTING:
+        switch (type.name()) {
+            case "COD_BUCKET":
+            case "PUFFERFISH_BUCKET":
+            case "SALMON_BUCKET":
+            case "ITEM_FRAME":
+            case "PAINTING":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaFirework.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaFirework.java
index 4016a9edf1796849ae10f52680e9854e61a381b1..ba5cd3a1e1e3eebda790227ea5a3106c08b8402a 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaFirework.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaFirework.java
@@ -255,8 +255,8 @@ class CraftMetaFirework extends CraftMetaItem implements FireworkMeta {
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case FIREWORK_ROCKET:
+        switch(type.name()) {
+            case "FIREWORK_ROCKET":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaKnowledgeBook.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaKnowledgeBook.java
index 317631c3960f1f1f19dad8a033b06bb166d93ecd..19bebaa5074fd8b9f02b6050c449eeba39080810 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaKnowledgeBook.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaKnowledgeBook.java
@@ -84,8 +84,8 @@ public class CraftMetaKnowledgeBook extends CraftMetaItem implements KnowledgeBo
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case KNOWLEDGE_BOOK:
+        switch (type.name()) {
+            case "KNOWLEDGE_BOOK":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaLeatherArmor.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaLeatherArmor.java
index 9624fc13cfed032ace7fc90a890ce3952275c524..796b65616e319763b8a57896830da82728e88c8e 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaLeatherArmor.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaLeatherArmor.java
@@ -67,12 +67,12 @@ class CraftMetaLeatherArmor extends CraftMetaItem implements LeatherArmorMeta {
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case LEATHER_HELMET:
-            case LEATHER_HORSE_ARMOR:
-            case LEATHER_CHESTPLATE:
-            case LEATHER_LEGGINGS:
-            case LEATHER_BOOTS:
+        switch(type.name()) {
+            case "LEATHER_HELMET":
+            case "LEATHER_HORSE_ARMOR":
+            case "LEATHER_CHESTPLATE":
+            case "LEATHER_LEGGINGS":
+            case "LEATHER_BOOTS":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaMap.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaMap.java
index 2d7ddc62aab2d40806ab126a1bc1ed0b8f4d233c..f6cb93e1aa333b06331514a403566514271872fb 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaMap.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaMap.java
@@ -120,8 +120,8 @@ class CraftMetaMap extends CraftMetaItem implements MapMeta {
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case FILLED_MAP:
+        switch (type.name()) {
+            case "FILLED_MAP":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaPotion.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaPotion.java
index df0dda466535f7af41b8fd7f12d7d3a2b56f21b1..b071c572851b4603718fc9714b98f69c296fc2ca 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaPotion.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaPotion.java
@@ -148,11 +148,11 @@ class CraftMetaPotion extends CraftMetaItem implements PotionMeta {
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case POTION:
-            case SPLASH_POTION:
-            case LINGERING_POTION:
-            case TIPPED_ARROW:
+        switch(type.name()) {
+            case "POTION":
+            case "SPLASH_POTION":
+            case "LINGERING_POTION":
+            case "TIPPED_ARROW":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
index 6db0d35eba647a0e81ca464fa52dc4a404ddd2ab..ca3f4448c2a470c35d5d48c5fda7260ed059350b 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
@@ -118,19 +118,19 @@ class CraftMetaSkull extends CraftMetaItem implements SkullMeta {
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case CREEPER_HEAD:
-            case CREEPER_WALL_HEAD:
-            case DRAGON_HEAD:
-            case DRAGON_WALL_HEAD:
-            case PLAYER_HEAD:
-            case PLAYER_WALL_HEAD:
-            case SKELETON_SKULL:
-            case SKELETON_WALL_SKULL:
-            case WITHER_SKELETON_SKULL:
-            case WITHER_SKELETON_WALL_SKULL:
-            case ZOMBIE_HEAD:
-            case ZOMBIE_WALL_HEAD:
+        switch (type.name()) {
+            case "CREEPER_HEAD":
+            case "CREEPER_WALL_HEAD":
+            case "DRAGON_HEAD":
+            case "DRAGON_WALL_HEAD":
+            case "PLAYER_HEAD":
+            case "PLAYER_WALL_HEAD":
+            case "SKELETON_SKULL":
+            case "SKELETON_WALL_SKULL":
+            case "WITHER_SKELETON_SKULL":
+            case "WITHER_SKELETON_WALL_SKULL":
+            case "ZOMBIE_HEAD":
+            case "ZOMBIE_WALL_HEAD":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
index 417160685671ff723ca6c38955d4675bc98f6fa2..21115f8cab519f6e14b2f61f980504b741744f90 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
@@ -111,70 +111,70 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case BAT_SPAWN_EGG:
-            case BEE_SPAWN_EGG:
-            case BLAZE_SPAWN_EGG:
-            case CAT_SPAWN_EGG:
-            case CAVE_SPIDER_SPAWN_EGG:
-            case CHICKEN_SPAWN_EGG:
-            case COD_SPAWN_EGG:
-            case COW_SPAWN_EGG:
-            case CREEPER_SPAWN_EGG:
-            case DOLPHIN_SPAWN_EGG:
-            case DONKEY_SPAWN_EGG:
-            case DROWNED_SPAWN_EGG:
-            case ELDER_GUARDIAN_SPAWN_EGG:
-            case ENDERMAN_SPAWN_EGG:
-            case ENDERMITE_SPAWN_EGG:
-            case EVOKER_SPAWN_EGG:
-            case FOX_SPAWN_EGG:
-            case GHAST_SPAWN_EGG:
-            case GUARDIAN_SPAWN_EGG:
-            case HOGLIN_SPAWN_EGG:
-            case HORSE_SPAWN_EGG:
-            case HUSK_SPAWN_EGG:
-            case LLAMA_SPAWN_EGG:
-            case MAGMA_CUBE_SPAWN_EGG:
-            case MOOSHROOM_SPAWN_EGG:
-            case MULE_SPAWN_EGG:
-            case OCELOT_SPAWN_EGG:
-            case PANDA_SPAWN_EGG:
-            case PARROT_SPAWN_EGG:
-            case PHANTOM_SPAWN_EGG:
-            case PIGLIN_SPAWN_EGG:
-            case PIG_SPAWN_EGG:
-            case PILLAGER_SPAWN_EGG:
-            case POLAR_BEAR_SPAWN_EGG:
-            case PUFFERFISH_SPAWN_EGG:
-            case RABBIT_SPAWN_EGG:
-            case RAVAGER_SPAWN_EGG:
-            case SALMON_SPAWN_EGG:
-            case SHEEP_SPAWN_EGG:
-            case SHULKER_SPAWN_EGG:
-            case SILVERFISH_SPAWN_EGG:
-            case SKELETON_HORSE_SPAWN_EGG:
-            case SKELETON_SPAWN_EGG:
-            case SLIME_SPAWN_EGG:
-            case SPIDER_SPAWN_EGG:
-            case SQUID_SPAWN_EGG:
-            case STRAY_SPAWN_EGG:
-            case STRIDER_SPAWN_EGG:
-            case TRADER_LLAMA_SPAWN_EGG:
-            case TROPICAL_FISH_SPAWN_EGG:
-            case TURTLE_SPAWN_EGG:
-            case VEX_SPAWN_EGG:
-            case VILLAGER_SPAWN_EGG:
-            case VINDICATOR_SPAWN_EGG:
-            case WANDERING_TRADER_SPAWN_EGG:
-            case WITCH_SPAWN_EGG:
-            case WITHER_SKELETON_SPAWN_EGG:
-            case WOLF_SPAWN_EGG:
-            case ZOGLIN_SPAWN_EGG:
-            case ZOMBIE_HORSE_SPAWN_EGG:
-            case ZOMBIE_SPAWN_EGG:
-            case ZOMBIE_VILLAGER_SPAWN_EGG:
-            case ZOMBIFIED_PIGLIN_SPAWN_EGG:
+        switch (type.name()) {
+            case "BAT_SPAWN_EGG":
+            case "BEE_SPAWN_EGG":
+            case "BLAZE_SPAWN_EGG":
+            case "CAT_SPAWN_EGG":
+            case "CAVE_SPIDER_SPAWN_EGG":
+            case "CHICKEN_SPAWN_EGG":
+            case "COD_SPAWN_EGG":
+            case "COW_SPAWN_EGG":
+            case "CREEPER_SPAWN_EGG":
+            case "DOLPHIN_SPAWN_EGG":
+            case "DONKEY_SPAWN_EGG":
+            case "DROWNED_SPAWN_EGG":
+            case "ELDER_GUARDIAN_SPAWN_EGG":
+            case "ENDERMAN_SPAWN_EGG":
+            case "ENDERMITE_SPAWN_EGG":
+            case "EVOKER_SPAWN_EGG":
+            case "FOX_SPAWN_EGG":
+            case "GHAST_SPAWN_EGG":
+            case "GUARDIAN_SPAWN_EGG":
+            case "HOGLIN_SPAWN_EGG":
+            case "HORSE_SPAWN_EGG":
+            case "HUSK_SPAWN_EGG":
+            case "LLAMA_SPAWN_EGG":
+            case "MAGMA_CUBE_SPAWN_EGG":
+            case "MOOSHROOM_SPAWN_EGG":
+            case "MULE_SPAWN_EGG":
+            case "OCELOT_SPAWN_EGG":
+            case "PANDA_SPAWN_EGG":
+            case "PARROT_SPAWN_EGG":
+            case "PHANTOM_SPAWN_EGG":
+            case "PIGLIN_SPAWN_EGG":
+            case "PIG_SPAWN_EGG":
+            case "PILLAGER_SPAWN_EGG":
+            case "POLAR_BEAR_SPAWN_EGG":
+            case "PUFFERFISH_SPAWN_EGG":
+            case "RABBIT_SPAWN_EGG":
+            case "RAVAGER_SPAWN_EGG":
+            case "SALMON_SPAWN_EGG":
+            case "SHEEP_SPAWN_EGG":
+            case "SHULKER_SPAWN_EGG":
+            case "SILVERFISH_SPAWN_EGG":
+            case "SKELETON_HORSE_SPAWN_EGG":
+            case "SKELETON_SPAWN_EGG":
+            case "SLIME_SPAWN_EGG":
+            case "SPIDER_SPAWN_EGG":
+            case "SQUID_SPAWN_EGG":
+            case "STRAY_SPAWN_EGG":
+            case "STRIDER_SPAWN_EGG":
+            case "TRADER_LLAMA_SPAWN_EGG":
+            case "TROPICAL_FISH_SPAWN_EGG":
+            case "TURTLE_SPAWN_EGG":
+            case "VEX_SPAWN_EGG":
+            case "VILLAGER_SPAWN_EGG":
+            case "VINDICATOR_SPAWN_EGG":
+            case "WANDERING_TRADER_SPAWN_EGG":
+            case "WITCH_SPAWN_EGG":
+            case "WITHER_SKELETON_SPAWN_EGG":
+            case "WOLF_SPAWN_EGG":
+            case "ZOGLIN_SPAWN_EGG":
+            case "ZOMBIE_HORSE_SPAWN_EGG":
+            case "ZOMBIE_SPAWN_EGG":
+            case "ZOMBIE_VILLAGER_SPAWN_EGG":
+            case "ZOMBIFIED_PIGLIN_SPAWN_EGG":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaTropicalFishBucket.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaTropicalFishBucket.java
index 7b7e4907cd197e16ea5851b91d9b975658d2dbf3..4bbc9c8f99159f4465efb7e28f5c104048eb2b85 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaTropicalFishBucket.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaTropicalFishBucket.java
@@ -86,8 +86,8 @@ class CraftMetaTropicalFishBucket extends CraftMetaItem implements TropicalFishB
 
     @Override
     boolean applicableTo(Material type) {
-        switch (type) {
-            case TROPICAL_FISH_BUCKET:
+        switch (type.name()) {
+            case "TROPICAL_FISH_BUCKET":
                 return true;
             default:
                 return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftLegacy.java b/src/main/java/org/bukkit/craftbukkit/util/CraftLegacy.java
index c8dde75de231b5fdde88f5bc111c21883d9c3f49..d46cb302b05f1b0413f14aafad77c91a7e29ff9e 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftLegacy.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftLegacy.java
@@ -27,8 +27,7 @@ public final class CraftLegacy {
     }
 
     public static Material[] modern_values() {
-        Material[] values = Material.values();
-        return Arrays.copyOfRange(values, 0, Material.LEGACY_AIR.ordinal());
+        return Material.modernValues();
     }
 
     public static int modern_ordinal(Material material) {
@@ -37,6 +36,6 @@ public final class CraftLegacy {
             throw new NoSuchFieldError("Legacy field ordinal: " + material);
         }
 
-        return material.ordinal();
+        return material.modernOrdinal();
     }
 }
diff --git a/src/test/java/org/bukkit/craftbukkit/inventory/FactoryItemMaterialTest.java b/src/test/java/org/bukkit/craftbukkit/inventory/FactoryItemMaterialTest.java
index c279579c2bd54363090c6c99fbd92798a6b97c79..9fb07e8b0f4ad178fe92a57497b701b35d239b0c 100644
--- a/src/test/java/org/bukkit/craftbukkit/inventory/FactoryItemMaterialTest.java
+++ b/src/test/java/org/bukkit/craftbukkit/inventory/FactoryItemMaterialTest.java
@@ -34,7 +34,7 @@ public class FactoryItemMaterialTest extends AbstractTestingBase {
         materials = list.toArray(new Material[list.size()]);
     }
 
-    static String name(Enum<?> from, Enum<?> to) {
+    static String name(Material from, Material to) {
         if (from.getClass() == to.getClass()) {
             return buffer.delete(0, Integer.MAX_VALUE).append(from.getClass().getName()).append(' ').append(from.name()).append(" to ").append(to.name()).toString();
         }
