From a63ba7f62f3976a8b513bda61b2b2caaae51d9dc Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Tue, 29 Oct 2019 22:55:41 +0000
Subject: [PATCH] Fix #1 - Wrong command block edit message


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index bc9ffbd7b..3f5f45668 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -645,7 +645,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
             tileentitycommand.b(!event.getNewNeedsRedstone());
             commandblocklistenerabstract.e();
             if (!UtilColor.b(event.getNewCommand())) {
-                this.player.sendMessage(new ChatMessage("advMode.setCommand.success", new Object[]{s}));
+                this.player.sendMessage(new ChatMessage("advMode.setCommand.success", event.getNewCommand()));
             }
         }
     }
-- 
2.23.0

