From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Thu, 21 Nov 2019 11:59:46 +0000
Subject: [PATCH] Implement Block#getCenter


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index eeeaadbd3aa290fc916f0d465f10cad0687b900d..5113e7600f04f5e624ca0c2c2de9dbf48be53afc 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -97,6 +97,11 @@ public class CraftBlock implements Block {
         return new Location(getWorld(), position.getX(), position.getY(), position.getZ());
     }
 
+    @Override
+    public Location getCenter() {
+        return new Location(getWorld(), position.getX() + 0.5, position.getY() + 0.5, position.getZ() + 0.5);
+    }
+
     @Override
     public Location getLocation(Location loc) {
         if (loc != null) {
