From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sat, 23 Nov 2019 02:29:24 +0000
Subject: [PATCH] Case-insensitive command matching for command blocks

(Which matches the behaviour for players and console)

diff --git a/src/main/java/net/minecraft/commands/CommandDispatcher.java b/src/main/java/net/minecraft/commands/CommandDispatcher.java
index 29a395e69aab9c50a45c5fd5ab8795fd8e3ef727..f4a569f9c0858f8cf0ec502164a10bc56f88dae5 100644
--- a/src/main/java/net/minecraft/commands/CommandDispatcher.java
+++ b/src/main/java/net/minecraft/commands/CommandDispatcher.java
@@ -224,6 +224,7 @@ public class CommandDispatcher {
 
         String[] args = command.split(" ");
         if (args.length == 0) return 0; // Paper - empty commands shall not be dispatched
+        args[0] = args[0].toLowerCase();
 
         String cmd = args[0];
         if (cmd.startsWith("minecraft:")) cmd = cmd.substring("minecraft:".length());
