From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sat, 23 Nov 2019 02:29:24 +0000
Subject: [PATCH] Case-insensitive command matching for command blocks

(Which matches the behaviour for players and console)

diff --git a/src/main/java/net/minecraft/commands/CommandDispatcher.java b/src/main/java/net/minecraft/commands/CommandDispatcher.java
index 7e30ec9a08d919d2ae9218ee0a11f77719129f07..10e772d4ae370dd941d510f6449f980eef984a10 100644
--- a/src/main/java/net/minecraft/commands/CommandDispatcher.java
+++ b/src/main/java/net/minecraft/commands/CommandDispatcher.java
@@ -224,6 +224,7 @@ public class CommandDispatcher {
 
         String[] args = command.split(" ");
         if (args.length == 0) return 0; // Paper - empty commands shall not be dispatched
+        args[0] = args[0].toLowerCase();
 
         String cmd = args[0];
         if (cmd.startsWith("minecraft:")) cmd = cmd.substring("minecraft:".length());
