From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Fri, 13 Dec 2019 18:26:58 +0000
Subject: [PATCH] Add showInvalidBedMessage world config


diff --git a/src/main/java/fr/thekinrar/papyrus/PapyrusWorldConfig.java b/src/main/java/fr/thekinrar/papyrus/PapyrusWorldConfig.java
index 52274c1bb387ae8e42f3c7078328d99f5e7df1df..4cdcd51049703e12978588736fea3b9277d9aeba 100644
--- a/src/main/java/fr/thekinrar/papyrus/PapyrusWorldConfig.java
+++ b/src/main/java/fr/thekinrar/papyrus/PapyrusWorldConfig.java
@@ -59,4 +59,10 @@ public class PapyrusWorldConfig {
         config.addDefault("world-settings.default." + path, def);
         return config.getString("world-settings." + worldName + "." + path, config.getString("world-settings.default." + path));
     }
+
+    public boolean showInvalidBedMessage = true;
+    private void showInvalidBedMessage() {
+        showInvalidBedMessage = getBoolean("show-invalid-bed-message", showInvalidBedMessage);
+        log("Show invalid bed message: " + showInvalidBedMessage);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 251227402a6135de754afbaadd665a1ba6a605a1..e54f14ea90f28c71c5355840f3bce967d2971566 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -919,7 +919,9 @@ public abstract class PlayerList {
                     isBedSpawn = true;
                     location = new Location(worldserver1.getWorld(), vec3d.x, vec3d.y, vec3d.z, f1, 0.0F);
                 } else if (blockposition != null) {
-                    entityplayer1.playerConnection.sendPacket(new PacketPlayOutGameStateChange(PacketPlayOutGameStateChange.a, 0.0F));
+                    if(worldserver1.getWorld().getHandle().papyrusConfig.showInvalidBedMessage) {
+                        entityplayer1.playerConnection.sendPacket(new PacketPlayOutGameStateChange(PacketPlayOutGameStateChange.a, 0.0F));
+                    }
                 }
             }
 
