From a1ca10760a1a645c92d1a518bee4f468fe9a2072 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Fri, 13 Dec 2019 18:26:58 +0000
Subject: [PATCH] Add showInvalidBedMessage world config


diff --git a/src/main/java/fr/thekinrar/papyrus/PapyrusWorldConfig.java b/src/main/java/fr/thekinrar/papyrus/PapyrusWorldConfig.java
index 52274c1bb3..4cdcd51049 100644
--- a/src/main/java/fr/thekinrar/papyrus/PapyrusWorldConfig.java
+++ b/src/main/java/fr/thekinrar/papyrus/PapyrusWorldConfig.java
@@ -59,4 +59,10 @@ public class PapyrusWorldConfig {
         config.addDefault("world-settings.default." + path, def);
         return config.getString("world-settings." + worldName + "." + path, config.getString("world-settings.default." + path));
     }
+
+    public boolean showInvalidBedMessage = true;
+    private void showInvalidBedMessage() {
+        showInvalidBedMessage = getBoolean("show-invalid-bed-message", showInvalidBedMessage);
+        log("Show invalid bed message: " + showInvalidBedMessage);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 59f2db4cf2..100b264698 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -810,7 +810,9 @@ public abstract class PlayerList {
                     isBedSpawn = true;
                     location = new Location(worldserver1.getWorld(), vec3d.x, vec3d.y, vec3d.z);
                 } else if (blockposition != null) {
-                    entityplayer1.playerConnection.sendPacket(new PacketPlayOutGameStateChange(PacketPlayOutGameStateChange.a, 0.0F));
+                    if(worldserver1.getWorld().getHandle().papyrusConfig.showInvalidBedMessage) {
+                        entityplayer1.playerConnection.sendPacket(new PacketPlayOutGameStateChange(PacketPlayOutGameStateChange.a, 0.0F));
+                    }
                 }
             }
 
-- 
2.28.0

