From ad50c2fbd88bfe6bd28057db0723370696098bf9 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Fri, 13 Dec 2019 18:26:58 +0000
Subject: [PATCH] Add showInvalidBedMessage world config


diff --git a/src/main/java/fr/thekinrar/papyrus/PapyrusWorldConfig.java b/src/main/java/fr/thekinrar/papyrus/PapyrusWorldConfig.java
index 52274c1bb..4cdcd5104 100644
--- a/src/main/java/fr/thekinrar/papyrus/PapyrusWorldConfig.java
+++ b/src/main/java/fr/thekinrar/papyrus/PapyrusWorldConfig.java
@@ -59,4 +59,10 @@ public class PapyrusWorldConfig {
         config.addDefault("world-settings.default." + path, def);
         return config.getString("world-settings." + worldName + "." + path, config.getString("world-settings.default." + path));
     }
+
+    public boolean showInvalidBedMessage = true;
+    private void showInvalidBedMessage() {
+        showInvalidBedMessage = getBoolean("show-invalid-bed-message", showInvalidBedMessage);
+        log("Show invalid bed message: " + showInvalidBedMessage);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index d3b0ff32c..cded2a61b 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -783,7 +783,9 @@ public abstract class PlayerList {
                     entityplayer1.setRespawnPosition(worldserver1.getDimensionKey(), blockposition, flag1, false);
                     flag2 = !flag && worldserver1.getType(blockposition).getBlock() instanceof BlockRespawnAnchor;
                 } else if (blockposition != null) {
-                    entityplayer1.playerConnection.sendPacket(new PacketPlayOutGameStateChange(PacketPlayOutGameStateChange.a, 0.0F));
+                    if(worldserver1.getWorld().getHandle().papyrusConfig.showInvalidBedMessage) {
+                        entityplayer1.playerConnection.sendPacket(new PacketPlayOutGameStateChange(PacketPlayOutGameStateChange.a, 0.0F));
+                    }
                 }
             }
 
-- 
2.26.2

