From 31158bd11ada5f5b4ac5b6635e06b2b968b31574 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Mon, 6 Jan 2020 00:11:45 +0000
Subject: [PATCH] Add forceUpgradeWorlds option


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 2387f07e01..c7501b4057 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -56,6 +56,7 @@ import joptsimple.NonOptionArgumentSpec;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
 import joptsimple.OptionSpec;
+import org.apache.commons.lang3.ArrayUtils;
 import org.apache.commons.lang3.Validate;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -154,6 +155,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
     private boolean as;
     private boolean forceUpgrade;
     private boolean eraseCache;
+    private String[] forceUpgradeWorlds;
     private float av;
     public final Executor executorService;
     @Nullable
@@ -279,6 +281,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         }
 
         if (this.forceUpgrade) {
+            if(this.forceUpgradeWorlds == null || ArrayUtils.contains(this.forceUpgradeWorlds, s)) {
             MinecraftServer.LOGGER.info("Forcing world upgrade! {}", s); // CraftBukkit
             WorldData worlddata = this.getConvertable().b(s); // CraftBukkit
 
@@ -313,6 +316,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
                     }
                 }
             }
+            }
         }
 
     }
@@ -1414,6 +1418,10 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
                 dedicatedserver.setEraseCache(true);
             }
 
+            if (optionset.has("forceUpgradeWorlds")) {
+                dedicatedserver.setForceUpgradeWorlds(((String) optionset.valueOf("forceUpgradeWorlds")).split(","));
+            }
+
             dedicatedserver.serverThread.start();
             // CraftBukkit end
         } catch (Exception exception) {
@@ -1434,6 +1442,10 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         this.eraseCache = flag;
     }
 
+    protected void setForceUpgradeWorlds(String[] value) {
+        this.forceUpgradeWorlds = value;
+    }
+
     public void startServerThread() {
         /* CraftBukkit start - prevent abuse
         this.serverThread.start();
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index d3c729efb9..59d41434d4 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -114,6 +114,13 @@ public class Main {
                 acceptsAll(asList("eraseCache"), "Whether to force cache erase during world upgrade");
                 acceptsAll(asList("nogui"), "Disables the graphical console");
 
+                // Papyrus start
+                acceptsAll(asList("forceUpgradeWorlds"), "Worlds to force upgrade")
+                        .withOptionalArg()
+                        .ofType(String.class)
+                        .describedAs("Comma-separated list of world names");
+                // Papyrus end
+
                 acceptsAll(asList("nojline"), "Disables jline and emulates the vanilla console");
 
                 acceptsAll(asList("noconsole"), "Disables the console");
-- 
2.25.0

