From da61932e54f0a09fb766068d2ef1d220b49ea57b Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sun, 1 Mar 2020 01:18:01 +0000
Subject: [PATCH] More plugin-loading related APIs


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 09246e828..dadceb9ef 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -675,6 +675,11 @@ public final class CraftServer implements Server {
     }
     // End Temporary calls
 
+    @Override
+    public File getPluginFolder() {
+        return (File) console.options.valueOf("plugins");
+    }
+
     @Override
     public String getUpdateFolder() {
         return this.configuration.getString("settings.update-folder", "update");
-- 
2.28.0

