From f51d82dd8570db0aefa75ee432ed0b7ca6e3e97c Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Mon, 6 Jan 2020 00:11:45 +0000
Subject: [PATCH] Add forceUpgradeWorlds option


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 9f56b02387..8455e9b954 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -56,6 +56,7 @@ import joptsimple.NonOptionArgumentSpec;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
 import joptsimple.OptionSpec;
+import org.apache.commons.lang3.ArrayUtils;
 import org.apache.commons.lang3.Validate;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -160,6 +161,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
     private boolean as;
     private boolean forceUpgrade;
     private boolean eraseCache;
+    private String[] forceUpgradeWorlds;
     private float av;
     public final Executor executorService;
     @Nullable
@@ -285,6 +287,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         }
 
         if (this.forceUpgrade) {
+            if(this.forceUpgradeWorlds == null || ArrayUtils.contains(this.forceUpgradeWorlds, s)) {
             MinecraftServer.LOGGER.info("Forcing world upgrade! {}", s); // CraftBukkit
             WorldData worlddata = this.getConvertable().b(s); // CraftBukkit
 
@@ -319,6 +322,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
                     }
                 }
             }
+            }
         }
 
     }
@@ -1499,12 +1503,17 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
                 dedicatedserver.setForceUpgrade(true);
             }
 
+            if (optionset.has("forceUpgradeWorlds")) {
+                dedicatedserver.setForceUpgradeWorlds(((String) optionset.valueOf("forceUpgradeWorlds")).split(","));
+            }
+
             if (optionset.has("eraseCache")) {
                 dedicatedserver.setEraseCache(true);
             }
 
             Class.forName("net.minecraft.server.VillagerTrades");// Paper - load this sync so it won't fail later async
             dedicatedserver.serverThread.setPriority(Thread.NORM_PRIORITY+2); // Paper - boost priority
+
             dedicatedserver.serverThread.start();
             // CraftBukkit end
         } catch (Exception exception) {
@@ -1525,6 +1534,10 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         this.eraseCache = flag;
     }
 
+    protected void setForceUpgradeWorlds(String[] value) {
+        this.forceUpgradeWorlds = value;
+    }
+
     public void startServerThread() {
         /* CraftBukkit start - prevent abuse
         this.serverThread.start();
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 283165f0fa..59732de36f 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -115,6 +115,13 @@ public class Main {
                 acceptsAll(asList("eraseCache"), "Whether to force cache erase during world upgrade");
                 acceptsAll(asList("nogui"), "Disables the graphical console");
 
+                // Papyrus start
+                acceptsAll(asList("forceUpgradeWorlds"), "Worlds to force upgrade")
+                        .withOptionalArg()
+                        .ofType(String.class)
+                        .describedAs("Comma-separated list of world names");
+                // Papyrus end
+
                 acceptsAll(asList("nojline"), "Disables jline and emulates the vanilla console");
 
                 acceptsAll(asList("noconsole"), "Disables the console");
-- 
2.28.0

