From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Tue, 18 Aug 2020 02:32:00 +0200
Subject: [PATCH] Add forceUpgradeWorlds option


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 21543d435f9bab42580deaf450f690a1ae0337dc..7f52c6d2f80a79091207cd5380877be7e66f1cf8 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -156,6 +156,7 @@ import net.minecraft.world.phys.Vec3D;
 import net.minecraft.world.scores.PersistentScoreboard;
 import net.minecraft.world.scores.Scoreboard;
 import net.minecraft.world.scores.ScoreboardTeam;
+import org.apache.commons.lang3.ArrayUtils;
 import org.apache.commons.lang3.Validate;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -287,6 +288,8 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
     public volatile Thread shutdownThread; // Paper
     public volatile boolean abnormalExit = false; // Paper
 
+    public String[] forceUpgradeWorlds = null; // Papyrus
+
     public static <S extends MinecraftServer> S a(Function<Thread, S> function) {
         AtomicReference<S> atomicreference = new AtomicReference();
         Thread thread = new Thread(() -> {
@@ -368,6 +371,12 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         */
         // Paper end
         Runtime.getRuntime().addShutdownHook(new org.bukkit.craftbukkit.util.ServerShutdownThread(this));
+
+        // Papyrus start
+        if(options.has("forceUpgradeWorlds")) {
+            forceUpgradeWorlds = ((String) options.valueOf("forceUpgradeWorlds")).split(",");
+        }
+        // Papyrus end
     }
     // CraftBukkit end
 
@@ -517,7 +526,15 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
                 worlddata = new WorldDataServer(worldsettings, generatorsettings, Lifecycle.stable());
             }
             worlddata.checkName(name); // CraftBukkit - Migration did not rewrite the level.dat; This forces 1.8 to take the last loaded world as respawn (in this case the end)
-            // Paper - move down
+            if (options.has("forceUpgrade")) {
+                if(forceUpgradeWorlds == null || ArrayUtils.contains(this.forceUpgradeWorlds, s)) { // Papyrus
+                net.minecraft.server.Main.convertWorld(worldSession, DataConverterRegistry.a(), options.has("eraseCache"), () -> {
+                    return true;
+                }, worlddata.getGeneratorSettings().d().d().stream().map((entry1) -> {
+                    return ResourceKey.a(IRegistry.K, ((ResourceKey) entry1.getKey()).a());
+                }).collect(ImmutableSet.toImmutableSet()));
+                }
+            }
 
             IWorldDataServer iworlddataserver = worlddata;
             GeneratorSettings generatorsettings = worlddata.getGeneratorSettings();
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index d0fafe0fca4a08d1a5b2e3fac269b75297e3264d..2e3a3152eb86ac7d96340262059b468a60c01475 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -109,6 +109,8 @@ import net.minecraft.world.level.storage.WorldNBTStorage;
 import net.minecraft.world.level.storage.loot.LootTableRegistry;
 import net.minecraft.world.phys.Vec3D;
 import org.apache.commons.lang.Validate;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.StringUtils;
 import org.bukkit.BanList;
 import org.bukkit.Bukkit;
 import org.bukkit.ChatColor;
@@ -1160,6 +1162,16 @@ public final class CraftServer implements Server {
         worlddata.a(console.getServerModName(), console.getModded().isPresent());
         // Paper - move down
 
+        if (console.options.has("forceUpgrade")) {
+            if(console.forceUpgradeWorlds == null || ArrayUtils.contains(console.forceUpgradeWorlds, name)) { // Papyrus
+            net.minecraft.server.Main.convertWorld(worldSession, DataConverterRegistry.a(), console.options.has("eraseCache"), () -> {
+                return true;
+            }, worlddata.getGeneratorSettings().d().d().stream().map((entry) -> {
+                return ResourceKey.a(IRegistry.K, ((ResourceKey) entry.getKey()).a());
+            }).collect(ImmutableSet.toImmutableSet()));
+            }
+        }
+
         long j = BiomeManager.a(creator.seed());
         List<MobSpawner> list = ImmutableList.of(new MobSpawnerPhantom(), new MobSpawnerPatrol(), new MobSpawnerCat(), new VillageSiege(), new MobSpawnerTrader(worlddata));
         RegistryMaterials<WorldDimension> registrymaterials = worlddata.getGeneratorSettings().d();
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index f68746b99386060e91e1ea70862573b9a004ac3b..9762a0045bdc688b7cee92b2e83aefbd08b82848 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -123,6 +123,13 @@ public class Main {
                 acceptsAll(asList("eraseCache"), "Whether to force cache erase during world upgrade");
                 acceptsAll(asList("nogui"), "Disables the graphical console");
 
+                // Papyrus start
+                acceptsAll(asList("forceUpgradeWorlds"), "Worlds to force upgrade")
+                        .withOptionalArg()
+                        .ofType(String.class)
+                        .describedAs("Comma-separated list of world names");
+                // Papyrus end
+
                 acceptsAll(asList("nojline"), "Disables jline and emulates the vanilla console");
 
                 acceptsAll(asList("noconsole"), "Disables the console");
