From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Tue, 18 Aug 2020 02:32:00 +0200
Subject: [PATCH] Add forceUpgradeWorlds option


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 401e013989d75c8ae9c8a2212c33b53c70ad1b77..b11de6ebe7077cc4cb69c49d365557da941157b7 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -155,6 +155,7 @@ import net.minecraft.world.phys.Vec3D;
 import net.minecraft.world.scores.PersistentScoreboard;
 import net.minecraft.world.scores.Scoreboard;
 import net.minecraft.world.scores.ScoreboardTeam;
+import org.apache.commons.lang3.ArrayUtils;
 import org.apache.commons.lang3.Validate;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -285,6 +286,8 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
 
     public volatile Thread shutdownThread; // Paper
 
+    public String[] forceUpgradeWorlds = null; // Papyrus
+
     public static <S extends MinecraftServer> S a(Function<Thread, S> function) {
         AtomicReference<S> atomicreference = new AtomicReference();
         Thread thread = new Thread(() -> {
@@ -366,6 +369,12 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         */
         // Paper end
         Runtime.getRuntime().addShutdownHook(new org.bukkit.craftbukkit.util.ServerShutdownThread(this));
+
+        // Papyrus start
+        if(options.has("forceUpgradeWorlds")) {
+            forceUpgradeWorlds = ((String) options.valueOf("forceUpgradeWorlds")).split(",");
+        }
+        // Papyrus end
     }
     // CraftBukkit end
 
@@ -516,11 +525,13 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
             }
             worlddata.checkName(name); // CraftBukkit - Migration did not rewrite the level.dat; This forces 1.8 to take the last loaded world as respawn (in this case the end)
             if (options.has("forceUpgrade")) {
+                if(forceUpgradeWorlds == null || ArrayUtils.contains(this.forceUpgradeWorlds, s)) { // Papyrus
                 net.minecraft.server.Main.convertWorld(worldSession, DataConverterRegistry.a(), options.has("eraseCache"), () -> {
                     return true;
                 }, worlddata.getGeneratorSettings().d().d().stream().map((entry1) -> {
                     return ResourceKey.a(IRegistry.K, ((ResourceKey) entry1.getKey()).a());
                 }).collect(ImmutableSet.toImmutableSet()));
+                }
             }
 
             IWorldDataServer iworlddataserver = worlddata;
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 82a9e1547c5360fcb3b6f6cdf7ba15c1d5cb4005..fcb998af232dbfd6840a34d112fac2a20a06d955 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -108,6 +108,7 @@ import net.minecraft.world.level.storage.WorldNBTStorage;
 import net.minecraft.world.level.storage.loot.LootTableRegistry;
 import net.minecraft.world.phys.Vec3D;
 import org.apache.commons.lang.Validate;
+import org.apache.commons.lang3.ArrayUtils;
 import org.apache.commons.lang3.StringUtils;
 import org.bukkit.BanList;
 import org.bukkit.Bukkit;
@@ -1144,11 +1145,13 @@ public final class CraftServer implements Server {
         worlddata.a(console.getServerModName(), console.getModded().isPresent());
 
         if (console.options.has("forceUpgrade")) {
+            if(console.forceUpgradeWorlds == null || ArrayUtils.contains(console.forceUpgradeWorlds, name)) { // Papyrus
             net.minecraft.server.Main.convertWorld(worldSession, DataConverterRegistry.a(), console.options.has("eraseCache"), () -> {
                 return true;
             }, worlddata.getGeneratorSettings().d().d().stream().map((entry) -> {
                 return ResourceKey.a(IRegistry.K, ((ResourceKey) entry.getKey()).a());
             }).collect(ImmutableSet.toImmutableSet()));
+            }
         }
 
         long j = BiomeManager.a(creator.seed());
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 8037bce08eccb2eec8b931dec774868a77addb95..2d1cc5e246926b83312570884b71b06d917227fe 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -117,6 +117,13 @@ public class Main {
                 acceptsAll(asList("eraseCache"), "Whether to force cache erase during world upgrade");
                 acceptsAll(asList("nogui"), "Disables the graphical console");
 
+                // Papyrus start
+                acceptsAll(asList("forceUpgradeWorlds"), "Worlds to force upgrade")
+                        .withOptionalArg()
+                        .ofType(String.class)
+                        .describedAs("Comma-separated list of world names");
+                // Papyrus end
+
                 acceptsAll(asList("nojline"), "Disables jline and emulates the vanilla console");
 
                 acceptsAll(asList("noconsole"), "Disables the console");
