From 3cd899fb4bc04bbed9efccf2182cee5accdb4cd0 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Fri, 28 Feb 2020 21:50:26 +0000
Subject: [PATCH] Implement HybridEngine online mapping API


diff --git a/src/main/java/fr/thekinrar/papyrus/hybrid/CraftHybridEngine.java b/src/main/java/fr/thekinrar/papyrus/hybrid/CraftHybridEngine.java
index a4687e41f..e573929bb 100644
--- a/src/main/java/fr/thekinrar/papyrus/hybrid/CraftHybridEngine.java
+++ b/src/main/java/fr/thekinrar/papyrus/hybrid/CraftHybridEngine.java
@@ -243,6 +243,11 @@ public class CraftHybridEngine implements HybridEngine {
         save();
     }
 
+    @Override
+    public UUID getOnlineId(UUID id) {
+        return onlineIds.get(id);
+    }
+
     public UUID getIdForOnlinePlayer(GameProfile online) {
         if(onlineMapping.containsKey(online.getId())) {
             // We already know this online player, return cached ID
@@ -417,6 +422,21 @@ public class CraftHybridEngine implements HybridEngine {
         save();
     }
 
+    @Override
+    public UUID getOnlineMapping(UUID id) {
+        return onlineMapping.get(id);
+    }
+
+    @Override
+    public void setOnlineMapping(UUID from, UUID to) {
+        if(to == null)
+            onlineMapping.remove(to);
+        else
+            onlineMapping.put(from, to);
+
+        save();
+    }
+
     @Override
     public void registerPasswordProvider(HybridPasswordProvider provider) {
         passwordProviders.add(provider);
diff --git a/src/main/java/fr/thekinrar/papyrus/hybrid/CraftHybridProfile.java b/src/main/java/fr/thekinrar/papyrus/hybrid/CraftHybridProfile.java
index 48c04c0d9..ba9b66cbf 100644
--- a/src/main/java/fr/thekinrar/papyrus/hybrid/CraftHybridProfile.java
+++ b/src/main/java/fr/thekinrar/papyrus/hybrid/CraftHybridProfile.java
@@ -196,4 +196,14 @@ public class CraftHybridProfile implements HybridProfile {
     public void setAlwaysOffline(boolean alwaysOffline) {
         engine.setAlwaysOffline(id, alwaysOffline);
     }
+
+    @Override
+    public boolean isOnline() {
+        return getOnlineId() != null;
+    }
+
+    @Override
+    public UUID getOnlineId() {
+        return engine.getOnlineId(id);
+    }
 }
-- 
2.26.0

