From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Wed, 19 Aug 2020 06:06:57 +0200
Subject: [PATCH] Add ignoreChunkMismatch option


diff --git a/src/main/java/net/minecraft/util/worldupdate/WorldUpgrader.java b/src/main/java/net/minecraft/util/worldupdate/WorldUpgrader.java
index 1204ebc8dd7f1cad00be901c2a16d1228c56b8c8..2b9b029774329c35fa05a5ee07479d686c371b21 100644
--- a/src/main/java/net/minecraft/util/worldupdate/WorldUpgrader.java
+++ b/src/main/java/net/minecraft/util/worldupdate/WorldUpgrader.java
@@ -40,7 +40,7 @@ import net.minecraft.world.level.dimension.DimensionManager;
 
 public class WorldUpgrader {
 
-    private static final Logger LOGGER = LogManager.getLogger();
+    public static final Logger LOGGER = LogManager.getLogger(); // Papyrus - make public
     private static final ThreadFactory b = (new ThreadFactoryBuilder()).setDaemon(true).build();
     private final ImmutableSet<ResourceKey<DimensionManager>> c; // CraftBukkit
     private final boolean d;
diff --git a/src/main/java/net/minecraft/world/level/chunk/storage/IChunkLoader.java b/src/main/java/net/minecraft/world/level/chunk/storage/IChunkLoader.java
index e20b9e6c46093d48d5fa5eb3006087d4e998c205..243508074a942319b08d6f9b157a5cc9c9dbb294 100644
--- a/src/main/java/net/minecraft/world/level/chunk/storage/IChunkLoader.java
+++ b/src/main/java/net/minecraft/world/level/chunk/storage/IChunkLoader.java
@@ -13,8 +13,10 @@ import net.minecraft.SharedConstants;
 import net.minecraft.nbt.GameProfileSerializer;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.resources.ResourceKey;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.PlayerChunkMap;
 import net.minecraft.util.datafix.DataFixTypes;
+import net.minecraft.util.worldupdate.WorldUpgrader;
 import net.minecraft.world.level.ChunkCoordIntPair;
 import net.minecraft.world.level.World;
 import net.minecraft.world.level.levelgen.structure.PersistentStructureLegacy;
@@ -128,8 +130,14 @@ public class IChunkLoader implements AutoCloseable {
         // Paper start
         if (!chunkcoordintpair.equals(ChunkRegionLoader.getChunkCoordinate(nbttagcompound))) {
             String world = (this instanceof PlayerChunkMap) ? ((PlayerChunkMap)this).world.getWorld().getName() : null;
-            throw new IllegalArgumentException("Chunk coordinate and serialized data do not have matching coordinates, trying to serialize coordinate " + chunkcoordintpair.toString()
-                + " but compound says coordinate is " + ChunkRegionLoader.getChunkCoordinate(nbttagcompound).toString() + (world == null ? " for an unknown world" : (" for world: " + world)));
+            IllegalArgumentException e = new IllegalArgumentException("Chunk coordinate and serialized data do not have matching coordinates, trying to serialize coordinate " + chunkcoordintpair.toString()
+                    + " but compound says coordinate is " + ChunkRegionLoader.getChunkCoordinate(nbttagcompound).toString() + (world == null ? " for an unknown world" : (" for world: " + world)));
+
+            if(!MinecraftServer.getServer().options.has("ignoreChunkMismatch")) {
+                throw e;
+            } else {
+                WorldUpgrader.LOGGER.error("Error upgrading world - ignored", e);
+            }
         }
         // Paper end
         this.regionFileCache.write(chunkcoordintpair, nbttagcompound);
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index af9a3bad3724eaafe4826c2e1d7f30113f6c2c27..4c5e387a2eba546eeb1f6be535da949db1ce76d2 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -128,6 +128,8 @@ public class Main {
                         .withOptionalArg()
                         .ofType(String.class)
                         .describedAs("Comma-separated list of world names");
+
+                acceptsAll(asList("ignoreChunkMismatch"), "Whether to ignore chunk coordinates mismatches while upgrading a world");
                 // Papyrus end
 
                 acceptsAll(asList("nojline"), "Disables jline and emulates the vanilla console");
