From 81558dbe48e2cff2cc5881ad7e459a42b4dca2fe Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Sun, 1 Mar 2020 01:18:01 +0000
Subject: [PATCH] More plugin-loading related APIs


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 73b0b8f99..a60295d4b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -645,6 +645,11 @@ public final class CraftServer implements Server {
     }
     // End Temporary calls
 
+    @Override
+    public File getPluginFolder() {
+        return (File) console.options.valueOf("plugins");
+    }
+
     @Override
     public String getUpdateFolder() {
         return this.configuration.getString("settings.update-folder", "update");
-- 
2.25.1

