From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Indhi Rousseau <contact@thekinrar.fr>
Date: Tue, 18 Aug 2020 02:41:09 +0200
Subject: [PATCH] Add ETA to world upgrade message


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 6818f8496ab76ee6ffc747bd6848b43830ec8914..bd35c740988ad8b1023efb7c2815a42023e27677 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -43,6 +43,7 @@ import net.minecraft.util.MathHelper;
 import net.minecraft.util.datafix.DataConverterRegistry;
 import net.minecraft.util.worldupdate.WorldUpgrader;
 import net.minecraft.world.entity.npc.VillagerTrades;
+import net.minecraft.world.level.ChunkCoordIntPair;
 import net.minecraft.world.level.DataPackConfiguration;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.World;
@@ -281,6 +282,8 @@ public class Main {
         WorldUpgrader worldupgrader = new WorldUpgrader(convertable_conversionsession, datafixer, immutableset, flag);
         IChatBaseComponent ichatbasecomponent = null;
 
+        long startTime = System.currentTimeMillis();
+
         while (!worldupgrader.b()) {
             IChatBaseComponent ichatbasecomponent1 = worldupgrader.h();
 
@@ -294,7 +297,15 @@ public class Main {
             if (i > 0) {
                 int j = worldupgrader.f() + worldupgrader.g();
 
-                Main.LOGGER.info("{}% completed ({} / {} chunks)...", MathHelper.d((float) j / (float) i * 100.0F), j, i);
+                long elapsed = System.currentTimeMillis() - startTime;
+                ChunkCoordIntPair chunk = worldupgrader.currentChunk;
+
+                MinecraftServer.LOGGER.info("{}: {}% completed ({} / {} chunks), {} remaining, current: {}",
+                        convertable_conversionsession.getLevelName(),
+                        MathHelper.d((float) j / (float) i * 100.0F), j, i,
+                        org.apache.commons.lang3.time.DurationFormatUtils
+                                .formatDurationHMS((long) (((double) elapsed / (j > 0 ? j : 1)) * (i - j))),
+                        chunk == null ? "N/A" : chunk.toString() + " in " + "[" + chunk.getRegionX() + ", " + chunk.getRegionZ() + "]");
             }
 
             if (!booleansupplier.getAsBoolean()) {
diff --git a/src/main/java/net/minecraft/util/worldupdate/WorldUpgrader.java b/src/main/java/net/minecraft/util/worldupdate/WorldUpgrader.java
index a4970654496c52fcd02c5c055ff5ac551bd19da3..b1264c7cc0a2fb656af18a6315c90678f48df12e 100644
--- a/src/main/java/net/minecraft/util/worldupdate/WorldUpgrader.java
+++ b/src/main/java/net/minecraft/util/worldupdate/WorldUpgrader.java
@@ -57,6 +57,7 @@ public class WorldUpgrader {
     private volatile IChatBaseComponent o = new ChatMessage("optimizeWorld.stage.counting");
     private static final Pattern p = Pattern.compile("^r\\.(-?[0-9]+)\\.(-?[0-9]+)\\.mca$");
     private final WorldPersistentData q;
+    public ChunkCoordIntPair currentChunk = null;
 
     public WorldUpgrader(Convertable.ConversionSession convertable_conversionsession, DataFixer datafixer, ImmutableSet<ResourceKey<DimensionManager>> immutableset, boolean flag) { // CraftBukkit
         this.c = immutableset;
@@ -131,6 +132,7 @@ public class WorldUpgrader {
                     if (listiterator.hasNext()) {
                         ChunkCoordIntPair chunkcoordintpair = (ChunkCoordIntPair) listiterator.next();
                         boolean flag1 = false;
+                        currentChunk = chunkcoordintpair;
 
                         try {
                             NBTTagCompound nbttagcompound = ichunkloader.read(chunkcoordintpair);
